## <center>Download and separate the Raw Data</center>

## <center>Default Parameters</center>

```sh
# environment and location
conda activate sradownload
meta_path='/storage/PUBLIC_DATA/Metagenome'
```

## <center>NCBI source</center> 

### <center>Download avaliable datasets through study ID</center>

+ `2017_Frankel-Immunotherapy_MM`

    ```sh
    tmp_path="${meta_path}/2017_Frankel-Immunotherapy_MM"
    mkdir -p ${tmp_path}
    cd ${tmp_path}
    #===================================#
    # upload the SRR_Acc_List by manual #
    #===================================#
    touch PRJNA397906
    echo "prefetch --option-file SRR_Acc_List_PRJNA397906.csv -c -O ./RawData" > recording.sh
    nohup sh recording.sh >recording.log 2>&1 &
    ```

+ `2018_Gopalakrishnan-Immunotherapy_MM`

    ```sh
    tmp_path="${meta_path}/2018_Gopalakrishnan-Immunotherapy_MM"
    mkdir -p ${tmp_path}
    cd ${tmp_path}
    #===================================#
    # upload the SRR_Acc_List by manual #
    #===================================#
    touch PRJEB22893
    echo "prefetch --option-file SRR_Acc_List_PRJEB22893.csv -c -O ./RawData" > recording.sh
    nohup sh recording.sh >recording.log 2>&1 &
    ```

+ `2018_Matson-Immunotherapy_MM`

    ```sh
    tmp_path="${meta_path}/2018_Matson-Immunotherapy_MM"
    mkdir -p ${tmp_path}
    cd ${tmp_path}
    #===================================#
    # upload the SRR_Acc_List by manual #
    #===================================#
    touch PRJNA399742
    echo "prefetch --option-file SRR_Acc_List_SRP116709_metagenome.csv -c -O ./RawData" > recording.sh
    nohup sh recording.sh >recording.log 2>&1 &
    ```

+ `2018_Routy-Immunotherapy_NSCLC_RCC`

    ```sh
    tmp_path="${meta_path}/2018_Matson-Immunotherapy_MM"
    mkdir -p ${tmp_path}
    cd ${tmp_path}
    #===================================#
    # upload the SRR_Acc_List by manual #
    #===================================#
    touch PRJNA399742
    echo "prefetch --option-file SRR_Acc_List_SRP116709_metagenome.csv -c -O ./RawData" > recording.sh
    nohup sh recording.sh >recording.log 2>&1 &
    ```

+ `2019_Perters-Immunotherapy_MM`

    ```sh
    tmp_path="${meta_path}/2019_Perters-Immunotherapy_MM"
    mkdir -p ${tmp_path}
    cd ${tmp_path}
    #===================================#
    # upload the SRR_Acc_List by manual #
    #===================================#
    touch PRJNA541981
    echo "prefetch --option-file SRR_Acc_List_PRJNA541981_meta_non-baseline.csv -c -O ./RawData" > recording2.sh
    nohup sh recording2.sh >recording2.log 2>&1 &
    echo "prefetch --option-file SRR_Acc_List_PRJNA541981_meta_baseline.csv -c -O ./RawData" > recording.sh
    nohup sh recording.sh >recording.log 2>&1 &
    ```

+ `2019_Zheng-Immunotherapy_HCC`

    ```sh
    tmp_path="${meta_path}/2019_Zheng-Immunotherapy_HCC"
    mkdir -p ${tmp_path}
    cd ${tmp_path}
    #===================================#
    # upload the SRR_Acc_List by manual #
    #===================================#
    touch PRJNA505228
    echo "prefetch --option-file SRR_Acc_List_PRJNA505228_baseline.csv -c -O ./RawData" > recording.sh
    nohup sh recording.sh >recording.log 2>&1 &
    echo "prefetch --option-file SRR_Acc_List_PRJNA505228_non-baseline.csv -c -O ./RawData" > recording2.sh
    nohup sh recording2.sh >recording2.log 2>&1 &
    ```


+ `2021_Spencer-Immunotherapy_MM`

    ```sh
    tmp_path="${meta_path}/2021_Spencer-Immunotherapy_MM"
    mkdir -p ${tmp_path}
    cd ${tmp_path}
    #===================================#
    # upload the SRR_Acc_List by manual #
    #===================================#
    touch PRJNA770295
    echo "prefetch --option-file SRR_Acc_List_PRJNA770295.csv -c -O ./RawData" > recording.sh
    nohup sh recording.sh >recording.log 2>&1 &
    ```

+ `2021_Davar-Immunotherapy_MM`

    ```sh
    tmp_path="${meta_path}/2021_Davar-Immunotherapy_MM"
    mkdir -p ${tmp_path}
    cd ${tmp_path}
    #===================================#
    # upload the SRR_Acc_List by manual #
    #===================================#
    touch PRJNA672867
    echo "prefetch --option-file SRR_Acc_List_PRJNA672867.csv -c -O ./RawData" > recording.sh
    nohup sh recording.sh >recording.log 2>&1 &
    ```

+ `2021_Baruch-Immunotherapy_MM`

    ```sh
    tmp_path="${meta_path}/2021_Baruch-Immunotherapy_MM"
    mkdir -p ${tmp_path}
    cd ${tmp_path}
    #===================================#
    # upload the SRR_Acc_List by manual #
    #===================================#
    touch PRJNA678737
    echo "prefetch --option-file SRR_Acc_List_PRJNA678737.csv -c -O ./RawData" > recording.sh
    nohup sh recording.sh >recording.log 2>&1 &
    ```

+ `2022_McCulloch-Immunotherapy_MM`

    ```sh
    tmp_path="${meta_path}/2022_McCulloch-Immunotherapy_MM"
    mkdir -p ${tmp_path}
    cd ${tmp_path}
    #===================================#
    # upload the SRR_Acc_List by manual #
    #===================================#
    touch PRJNA762360
    echo "prefetch --option-file SRR_Acc_List_PRJNA762360_WGS.csv -c -O ./RawData" > recording.sh
    nohup sh recording.sh >recording.log 2>&1 &
    ```

### <center>Extract fastq from SRA</center>

+ `2017_Frankel-Immunotherapy_MM`
    
    ```sh
    Public_Path="/storage/PUBLIC_DATA/Metagenome/2017_Frankel-Immunotherapy_MM"
    Fastq_Path="/expt/expt_s2/Public_Data/2021-LYF-Pan_Cancer_Immunotherapy/Fastq/2017_Frankel-Immunotherapy_MM"
    SRR_Acc="SRR_Acc_List_PRJNA397906.csv"

    nohup parallel -j 3 --xapply "mkdir -p ${Fastq_Path}/{} && pfastq-dump --threads 20 --gzip -O ${Fastq_Path}/{} ${Public_Path}/RawData/{}/{}.sra" :::: ${Public_Path}/${SRR_Acc} >${Public_Path}/extract.log 2>&1 &
    ```

+ `2018_Gopalakrishnan-Immunotherapy_MM`

    ```sh
    Public_Path="/storage/PUBLIC_DATA/Metagenome/2018_Gopalakrishnan-Immunotherapy_MM"
    Fastq_Path="/expt/expt_s2/Public_Data/2021-LYF-Pan_Cancer_Immunotherapy/Fastq/2018_Gopalakrishnan-Immunotherapy_MM"
    SRR_Acc="SRR_Acc_List_PRJEB22893.csv"

    nohup parallel -j 3 --xapply "mkdir -p ${Fastq_Path}/{} && pfastq-dump --threads 20 --gzip -O ${Fastq_Path}/{} ${Public_Path}/RawData/{}/{}.sra" :::: ${Public_Path}/${SRR_Acc} >${Public_Path}/extract.log 2>&1 &
    ```

+ `2018_Matson-Immunotherapy_MM`

    ```sh
    Public_Path="/storage/PUBLIC_DATA/Metagenome/2018_Matson-Immunotherapy_MM"
    Fastq_Path="/expt/expt_s2/Public_Data/2021-LYF-Pan_Cancer_Immunotherapy/Fastq/2018_Matson-Immunotherapy_MM"
    SRR_Acc="SRR_Acc_List_SRP116709_metagenome.csv"

    nohup parallel -j 3 --xapply "mkdir -p ${Fastq_Path}/{} && pfastq-dump --threads 20 --gzip -O ${Fastq_Path}/{} ${Public_Path}/RawData/{}/{}.sra" :::: ${Public_Path}/${SRR_Acc} >${Public_Path}/extract.log 2>&1 &
    ```

+ `2018_Routy-Immunotherapy_NSCLC_RCC`

    ```sh
    Public_Path="/storage/PUBLIC_DATA/Metagenome/2018_Routy-Immunotherapy_NSCLC_RCC"
    Fastq_Path="/expt/expt_s2/Public_Data/2021-LYF-Pan_Cancer_Immunotherapy/Fastq/2018_Routy-Immunotherapy_NSCLC_RCC"
    SRR_Acc="SRR_Acc_List_PRJEB22863.csv"

    nohup parallel -j 3 --xapply "mkdir -p ${Fastq_Path}/{} && pfastq-dump --threads 20 --gzip -O ${Fastq_Path}/{} ${Public_Path}/RawData/{}/{}.sra" :::: ${Public_Path}/${SRR_Acc} >${Public_Path}/extract.log 2>&1 &
    ```

+ `2019_Perters-Immunotherapy_MM`

    ```sh
    Public_Path="/storage/PUBLIC_DATA/Metagenome/2019_Perters-Immunotherapy_MM"
    Fastq_Path="/expt/expt_s2/Public_Data/2021-LYF-Pan_Cancer_Immunotherapy/Fastq/2019_Perters-Immunotherapy_MM"
    SRR_Acc="SRR_Acc_List_PRJNA541981_metagenome.csv"

    cat ${Public_Path}/{SRR_Acc_List_PRJNA541981_meta_baseline.csv,SRR_Acc_List_PRJNA541981_meta_non-baseline.csv} > ${Public_Path}/${SRR_Acc}

    nohup parallel -j 3 --xapply "mkdir -p ${Fastq_Path}/{} && pfastq-dump --threads 20 --gzip -O ${Fastq_Path}/{} ${Public_Path}/RawData/{}/{}.sra" :::: ${Public_Path}/${SRR_Acc} >${Public_Path}/extract.log 2>&1 &
    ```

+ `2019_Zheng-Immunotherapy_HCC`

    ```sh
    Public_Path="/storage/PUBLIC_DATA/Metagenome/2019_Zheng-Immunotherapy_HCC"
    Fastq_Path="/expt/expt_s2/Public_Data/2021-LYF-Pan_Cancer_Immunotherapy/Fastq/2019_Zheng-Immunotherapy_HCC"
    SRR_Acc="SRR_Acc_List_PRJNA505228.csv"
    cat ${Public_Path}/{SRR_Acc_List_PRJNA505228_baseline.csv,SRR_Acc_List_PRJNA505228_non-baseline.csv} > ${Public_Path}/${SRR_Acc}

    nohup parallel -j 3 --xapply "mkdir -p ${Fastq_Path}/{} && pfastq-dump --threads 20 --gzip -O ${Fastq_Path}/{} ${Public_Path}/RawData/{}/{}.sra" :::: ${Public_Path}/${SRR_Acc} >${Public_Path}/extract.log 2>&1 &
    ```

+ `2021_Spencer-Immunotherapy_MM`

    ```sh
    Public_Path="/storage/PUBLIC_DATA/Metagenome/2021_Spencer-Immunotherapy_MM"
    Fastq_Path="/expt/expt_s2/Public_Data/2021-LYF-Pan_Cancer_Immunotherapy/Fastq/2021_Spencer-Immunotherapy_MM"
    SRR_Acc="SRR_Acc_List_PRJNA770295.csv"

    nohup parallel -j 5 --xapply "mkdir -p ${Fastq_Path}/{} && pfastq-dump --threads 20 --gzip -O ${Fastq_Path}/{} ${Public_Path}/RawData/{}/{}.sra" :::: ${Public_Path}/${SRR_Acc} >${Public_Path}/extract.log 2>&1 &
    ```

+ `2021_Davar-Immunotherapy_MM`

    ```sh
    Public_Path="/storage/PUBLIC_DATA/Metagenome/2021_Davar-Immunotherapy_MM"
    Fastq_Path="/expt/expt_s2/Public_Data/2021-LYF-Pan_Cancer_Immunotherapy/Fastq/2021_Davar-Immunotherapy_MM"
    prjID="PRJNA672867"
    # SRR_Acc="SRR_Acc_List_PRJNA672867_P0.csv"
    SRR_Acc="SRR_Acc_List_PRJNA672867_WGS.csv"

    nohup parallel -j 5 --xapply "mkdir -p ${Fastq_Path}/{} && pfastq-dump --threads 30 --gzip -O ${Fastq_Path}/{} ${Public_Path}/RawData/{}/{}.sra; echo \"{} finished at \`date\`\" >> ${Public_Path}/${prjID}.time" :::: ${Public_Path}/${SRR_Acc} >${Public_Path}/extract.log 2>&1 &
    ```

+ `2021_Baruch-Immunotherapy_MM`

    ```sh
    Public_Path="/storage/PUBLIC_DATA/Metagenome/2021_Baruch-Immunotherapy_MM"
    Fastq_Path="/expt/expt_s2/Public_Data/2021-LYF-Pan_Cancer_Immunotherapy/Fastq/2021_Baruch-Immunotherapy_MM"
    prjID="PRJNA678737"
    # SRR_Acc="SRR_Acc_List_PRJNA678737_P0.csv"
    SRR_Acc="SRR_Acc_List_PRJNA678737_WGS.csv"

    nohup parallel -j 5 --xapply "mkdir -p ${Fastq_Path}/{} && pfastq-dump --threads 30 --gzip -O ${Fastq_Path}/{} ${Public_Path}/RawData/{}/{}.sra; echo \"{} finished at \`date\`\" >> ${Public_Path}/${prjID}.time" :::: ${Public_Path}/${SRR_Acc} >${Public_Path}/extract.log 2>&1 &
    ```

+ `2022_McCulloch-Immunotherapy_MM`

    ```sh
    Public_Path="/storage/PUBLIC_DATA/Metagenome/2022_McCulloch-Immunotherapy_MM"
    Fastq_Path="/expt/expt_s2/Public_Data/2021-LYF-Pan_Cancer_Immunotherapy/Fastq/2022_McCulloch-Immunotherapy_MM"
    prjID="PRJNA762360"
    SRR_Acc="SRR_Acc_List_PRJNA762360_WGS.csv"

    nohup parallel -j 5 --xapply "mkdir -p ${Fastq_Path}/{} && pfastq-dump --threads 30 --gzip -O ${Fastq_Path}/{} ${Public_Path}/RawData/{}/{}.sra; echo \"{} finished at \`date\`\" >> ${Public_Path}/${prjID}.time" :::: ${Public_Path}/${SRR_Acc} >${Public_Path}/extract.log 2>&1 &
    ```

## <center>EBI source</center>

Directly download the fastq.gz files

+ `2022_Lee-Immunotherapy_MM`

    **two pair end reads**

    ```sh
    tmp_path="${meta_path}/2022_Lee-Immunotherapy_MM"
    mkdir -p ${tmp_path}
    cd ${tmp_path}
    #=====================================#
    # upload the wget_recording by manual #
    #=====================================#
    touch PRJEB43119
    $ awk -v FS='\t' -v OFS='' '{print "mkdir -p ./RawData/", $1, " && wget ftp://", $2, " -P ./RawData/", $1, " && wget ftp://", $3, " -P ./RawData/", $1}' wget_recording.tsv> recording.sh

    bash -x recording.sh 2>&1 |tee recording.log
    cp -r ./RawData/ /expt/expt_s2/Public_Data/2021-LYF-Pan_Cancer_Immunotherapy/Fastq/2022_Lee-Immunotherapy_MM
    ```

+   `wget_recording.tsv` 

    | Sample_name | Seq_1.fastq | Seq_2.fastq |
    | :--: | :--: | :--: |
    | **ERR6229432**|ftp.sra.ebi.ac.uk/vol1/fastq/ERR622/002/ERR6229432/ERR6229432_1.fastq.gz|ftp.sra.ebi.ac.uk/vol1/fastq/ERR622/002/ERR6229432/ERR6229432_2.fastq.gz|
    |**ERR6230245**|ftp.sra.ebi.ac.uk/vol1/fastq/ERR623/005/ERR6230245/ERR6230245_1.fastq.gz|ftp.sra.ebi.ac.uk/vol1/fastq/ERR623/005/ERR6230245/ERR6230245_2.fastq.gz|
    |**ERR6230278**|ftp.sra.ebi.ac.uk/vol1/fastq/ERR623/008/ERR6230278/ERR6230278_1.fastq.gz|ftp.sra.ebi.ac.uk/vol1/fastq/ERR623/008/ERR6230278/ERR6230278_2.fastq.gz|



## <center>Directly Download `*.fastq` files</center>

+   `2022_Derosa_Immunotherapy_NSCLC`

    ```sh
    # activate the environment
    # conda activate my_sradownload

    # setting the default parameters
    base_path="/storage1/Public_Data/2022_Derosa_PRJNA751792"
    fastq_path=$base_path"/Fastq"
    pfd_thrds=64
    sra_list=$base_path"/SRR_Acc_List_PRJNA751792.csv"
    temp_path="/storage1/Public_Data/temp"
    err_file=$base_path"/fastq-dump.err"

    # paralell download and pigz the fastq file 
    cat $sra_list | parallel -j 2 --tmpdir ${temp_path} --xapply " \
        mkdir -p ${fastq_path}/{1} && \
        parallel-fastq-dump --sra-id {1} \
        --threads $pfd_thrds --outdir ${fastq_path}/{1} \
        --split-3 > ${fastq_path}/{1}/{1}.log 2>&1 &&
        ([ -f ${fastq_path}/{1}/{1}.fastq ] && \
        { pigz ${fastq_path}/{1}/{1}.fastq; } || \
        { pigz ${fastq_path}/{1}/{1}_1.fastq; pigz ${fastq_path}/{1}/{1}_2.fastq;}) ||
        echo \"Something wrong in {1} when ran the parallel-fastq-dump\" > $err_file"

    # check the failed or err files
    cd ${fastq_path}
    grep 'err' */*.log|awk -F'/' '{print $1}' |uniq >> ../err.list
    cat ../err.list
    # SRR15372866
    # SRR15372882
    # SRR15372892
    # ...
    # SRR15373181
    # SRR15373182
    # SRR15373200

    # re-download the err files
    cat ${base_path}/err.list | parallel -j 4 --tmpdir ${temp_path} --xapply " \
        rm -rf ${fastq_path}/{1} && \
        mkdir -p ${fastq_path}/{1} && \
        parallel-fastq-dump --sra-id {1} \
        --threads $pfd_thrds --outdir ${fastq_path}/{1} \
        --split-3 > ${fastq_path}/{1}/{1}.log 2>&1 && \
        ([ -f ${fastq_path}/{1}/{1}.fastq ] && \
        { pigz ${fastq_path}/{1}/{1}.fastq; } || \
        { pigz ${fastq_path}/{1}/{1}_1.fastq; pigz ${fastq_path}/{1}/{1}_2.fastq;}) || \
        echo \"Something wrong in {1} when ran the parallel-fastq-dump\" > $err_file"

    # double check again the failed files
    cd ${fastq_path}
    grep 'err' */*.log|awk -F'/' '{print $1}' |uniq > ../err2.list
    cat ../err2.list

    egrep 'err|fail' */*.log|awk -F'/' '{print $1}' |uniq > ../err2.list

    ```