## <center>KneadData</center>

### Default parameters

```sh
conda activate metagenome_SOP
base_path="/storage1/yufenglin/prj/2022-LYF-Pan_Cancer_Immunotherapy/"
fastq_path="/storage1/Public_Data/2022_Derosa_PRJNA751792/Fastq"
temp_path=${base_path}"/temp" && mkdir -p ${temp_path}
meta_file=${base_path}"/00.metaInfo/20220623-Derosa-sample-path.tsv"

run_path='/expt/Yufeng/prj/2022-LYF-Pan_Cancer_Immunotherapy/'

# KneadData
knd_db='/storage1/Shared_Data/Database/Homo_sapiens/Homo_sapiens'
trim_opt='"SLIDINGWINDOW:4:20 MINLEN:50"'
thread=60
trim_path="/home/jyusers/yufenglin/anaconda3/envs/metagenome_SOP/share/trimmomatic-0.39-2/"
pigz_thrd=60

# kraken2
krn_db='/expt/Database/Kraken2_modification/GenBank/CurrentSCC'
krn_thrd=60
brn_t=0
brn_l='S'
brn_r=100

## set the meta information (sample name + sample path)
find $fastq_path -name *fastq.gz | awk -v FS='/' -v OFS='\t' '{print $(NF-1),$0 }' > $meta_file
```

### parallel run Metagenome SOP

+ 4 steps

    + move files `*.fastq`

    + kneadData

    + Kraken2 & bracken

    + storage files

```sh
mkdir -p ${temp_path}
awk -v FS='\t' -v OFS='\n' '{print $1,$2}' $meta_file | \
    parallel -j 2 --max-args=2 --tmpdir ${temp_path} --xapply " \
        rm -rf ${run_path}/{00.Recording,00.Fastq,02.KneadData,03.Kraken2_GenBank}/{1} && \
        mkdir -p ${run_path}/{00.Recording,00.Fastq,02.KneadData,03.Kraken2_GenBank}/{1} && \
        echo \"[Copy Fastq] {1} began at \`date\`\" >>${run_path}/00.Recording/{1}/{1}.time && \
        cp -f {2} ${run_path}/00.Fastq/{1}/{1}.fastq.gz && \
        echo -e \"[Copy Fastq] {1} ended at \`date\`\\n\\n[kneadData] {1} began at \`date\`\" >>${run_path}/00.Recording/{1}/{1}.time && \
        kneaddata -i ${run_path}/00.Fastq/{1}/{1}.fastq.gz \
            -o ${run_path}/02.KneadData/{1} -v -db ${knd_db} \
            --trimmomatic-options ${trim_opt} -t ${thread} \
            --remove-intermediate-output --output-prefix {1} --trimmomatic ${trim_path} \
            >${run_path}/00.Recording/{1}/02.KneadData_{1}.log 2>&1 && \
        find ${run_path}/02.KneadData/{1} -name '*fastq' |grep -v ${run_path}/02.KneadData/{1}/{1}.fastq | xargs rm -rf && \
        echo -e \"[kneadData] {1} end at\`date\`\\n\\n[Kraken2 GenBank] {1} began at \`date\`\" >>${run_path}/00.Recording/{1}/{1}.time && \
        kraken2 -db ${krn_db} --threads ${krn_thrd} --report-zero-counts \
            ${run_path}/02.KneadData/{1}/{1}.fastq \
            --report ${run_path}/03.Kraken2_GenBank/{1}/{1}_kraken.report \
            --output ${run_path}/03.Kraken2_GenBank/{1}/{1}_kraken.out \
            >${run_path}/00.Recording/{1}/03.Kraken_{1}.log 2>&1 && \
        bracken -d ${krn_db} -t ${brn_t} -l ${brn_l} -r ${brn_r} \
            -i ${run_path}/03.Kraken2_GenBank/{1}/{1}_kraken.report \
            -o ${run_path}/03.Kraken2_GenBank/{1}/{1}_bracken.report \
            -w ${run_path}/03.Kraken2_GenBank/{1}/{1}_${brn_l}.bracken.report \
            >${run_path}/00.Recording/{1}/03.Bracken_{1}.log 2>&1 && \
        rm ${run_path}/03.Kraken2_GenBank/{1}/{1}_kraken.out && \
        kreport2mpa.py -r ${run_path}/03.Kraken2_GenBank/{1}/{1}_kraken.report \
            -o ${run_path}/03.Kraken2_GenBank/{1}/{1}_kraken_mpa.report && \
        kreport2mpa.py -r ${run_path}/03.Kraken2_GenBank/{1}/{1}_${brn_l}.bracken.report \
            -o ${run_path}/03.Kraken2_GenBank/{1}/{1}_${brn_l}.bracken_mpa.report && \
        echo -e \"[Kraken2 GenBank] {1} ended at \`date\`\\n\\n[Move Data] {1} began at \`date\`\" >> ${run_path}/00.Recording/{1}/{1}.time && \
        rm -rf ${base_path}/{00.Recording,00.Fastq,02.KneadData,03.Kraken2_GenBank}/{1} && \
        rm -rf ${run_path}/00.Fastq/{1} && \
        pigz -p ${pigz_thrd} ${run_path}/02.KneadData/{1}/{1}.fastq && \
        mv ${run_path}/00.Recording/{1} ${base_path}/00.Recording/ && \
        mv ${run_path}/02.KneadData/{1} ${base_path}/02.KneadData/ && \
        mv ${run_path}/03.Kraken2_GenBank/{1} ${base_path}/03.Kraken2_GenBank/ && \
        echo -e \"[Move Data] {1} ended at \`date\`\\n\\n\" >> ${base_path}/00.Recording/{1}/{1}.time && \
        cat ${base_path}/00.Recording/{1}/{1}.time >> ${base_path}/00.Recording/summarize.time" |tee ${base_path}/00.Scripts/Derosa.log
```

### Combine the results

```sh
mkdir -p ${base_path}/Summary/{02.KneadData_{PE,SE}_log,02.KneadData_res}

cp `ls ${base_path}/02.KneadData/*/*fastq.gz | egrep -v "_paired_1.fastq.gz|_paired_2.fastq.gz" |sed 's/fastq.gz/log/g'|xargs` ${base_path}/Summary/02.KneadData_SE_log
kneaddata_read_count_table --input ${base_path}/Summary/02.KneadData_SE_log --output ${base_path}/Summary/02.KneadData_res/KneadData_merge_SE.tsv >${base_path}/Summary/02.KneadData_res/kneaddata_read_count_table_SE.log 2>&1


cp `ls ${base_path}/02.KneadData/*/*_paired_1.fastq.gz | sed 's/_paired_1.fastq.gz/.log/g'|xargs` ${base_path}/Summary/02.KneadData_PE_log
kneaddata_read_count_table --input ${base_path}/Summary/02.KneadData_PE_log --output ${base_path}/Summary/02.KneadData_res/KneadData_merge_PE.tsv >${base_path}/Summary/02.KneadData_res/kneaddata_read_count_table_PE.log 2>&1
```

### Combine the kraken2

```sh
mkdir -p ${base_path}/Summary/03.Kraken2_GenBank_merge
cp ${base_path}/03.Kraken2_GenBank/*/*_S.bracken_mpa.report ${base_path}/Summary/03.Kraken2_GenBank_merge

combine_mpa.py -i  ${base_path}/Summary/03.Kraken2_GenBank_merge/*_S.bracken_mpa.report -o ${base_path}/Summary/03.Kraken2_GenBank_merge-bracken_mpa.tsv && (find ${base_path}/Summary/03.Kraken2_GenBank_merge/*_S.bracken_mpa.report |awk 'BEGIN{FS="\\/"; ORS="\t"; OFS="\t"}{print $NF}' |sed 's/_S.bracken_mpa.report//g'| awk 'BEGIN{OFS="\t"}{print "ID", $0}' && cat ${base_path}/Summary/03.Kraken2_GenBank_merge-bracken_mpa.tsv) > ${base_path}/Summary/03.Kraken2_GenBank_merge-bracken_mpa2.tsv && mv ${base_path}/Summary/03.Kraken2_GenBank_merge-bracken_mpa2.tsv ${base_path}/Summary/03.Kraken2_GenBank_merge-bracken_mpa.tsv && sed -i '2d' ${base_path}/Summary/03.Kraken2_GenBank_merge-bracken_mpa.tsv
```