---
title: "02.Feature Selection"
author: "IvanLIN"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Packages and functions

### Packages

```{r pacakges}
if (!require(microeco)) devtools::install_github("ChiLiubio/microeco"); require(microeco) 
if (!require(magrittr)) devtools::install_github("tidyverse/magrittr"); require(magrittr) 
if (!require(ComplexHeatmap)) BiocManager::install("ComplexHeatmap"); require(ComplexHeatmap) 
if (!require(reshape2)) install.packages("reshape2"); require(reshape2) # melt
if (!require(ggplot2)) install.packages("ggplot2"); require(ggplot2) 
if (!require(ggpubr)) install.packages("ggpubr"); require(ggpubr) # as_ggplot
if (!require(gridExtra)) install.packages("gridExtra"); require(gridExtra) # grid.arrange
if (!require(tibble)) install.packages("tibble"); require(tibble) # column_to_rownames
if (!require(ALDEx2)) BiocManager::install("ALDEx2"); require(ALDEx2)
if (!require(dplyr)) install.packages("dplyr"); require(dplyr) # summarise
if (!require(ggstatsplot)) install.packages("ggstatsplot"); require(ggstatsplot) # ggscatterstats
if (!require(ggExtra)) install.packages("ggExtra"); require(ggExtra) 
if (!require(circlize)) devtools::install_github("jokergoo/circlize"); require(circlize) # colorRamp2
if (!require(plyr)) install.packages("plyr"); require(plyr) # rbind.fill
if (!require(survminer)) install.packages("survminer"); require(survminer) # surv_cutpoint

```

### Functions

#### theme_Publication

```{r theme_Publication}
theme_Publication <- function(base_size=14) {
      library(grid)
      library(ggthemes)
      (theme_foundation(base_size=base_size)
       + theme(plot.title = element_text(face = "bold",
                                         size = rel(1.2), hjust = 0.5),
               text = element_text(),
               panel.background = element_rect(colour = NA),
               plot.background = element_rect(colour = NA),
               panel.border = element_rect(colour = NA),
               axis.title = element_text(face = "bold",size = rel(1)),
               axis.title.y = element_text(angle=90,vjust =2),
               axis.title.x = element_text(vjust = -0.2),
               axis.text = element_text(), 
               axis.line = element_line(colour="black"),
               axis.ticks = element_line(),
               panel.grid.major = element_line(colour="#f0f0f0"),
               panel.grid.minor = element_blank(),
               legend.key = element_rect(colour = NA),
               legend.position = "bottom",
               legend.direction = "horizontal",
               legend.key.size= unit(0.2, "cm"),
               # legend.margin = unit(0, "cm"),
               legend.title = element_text(face="italic"),
               plot.margin=unit(c(10,5,5,5),"mm"),
               strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
               strip.text = element_text(face="bold")
          ))
      
}

```

#### FileCreate

```{r file create}
FileCreate <- function(DirPath = "./",Prefix = "ExampleTest", Suffix = "pdf", version = "0.0"){
  date=as.character(Sys.Date())
  DirPath = gsub("/$","",DirPath)
  Suffix = gsub('^[.]',"",Suffix)
  if(! dir.exists(DirPath)){
    dir.create(DirPath,recursive =TRUE)
  }
  if (version == "0.0") {
    version=10
    while(TRUE){
      version_=paste(unlist(strsplit(as.character(version),"")),collapse=".")
      DirPathName = paste0(DirPath,"/",date,'-',Prefix,'-v',version_,".",Suffix)
      if(! file.exists(DirPathName)){
        return(DirPathName)
        break
      }
      version = version + 1
    }
  }else{
    DirPathName = paste0(DirPath,"/",date,'-',Prefix,'-v',version,".",Suffix)
    if(! dir.exists(DirPathName)){
      return(DirPathName)
    }
    return(DirPathName)
  }
}
```

#### FeatureSel_Wilcox

```{r FeatureSel_Wilcox function}
FeatureSel_Wilcox <- function(ds_list, group, 
                              taxa_level = "Specise", 
                              filter_thres = 0, 
                              barplot_num = 50,
                              # plot_thd = 2,
                              p_cutoff = .05,
                              color_values,
                              group_order,
                              only_calculate = F,
                              p_adjust_method = NULL){
  new_list <- list()
  for (subgroup in names(ds_list)) {
    fit<-try(
      t1 <- trans_diff$new(dataset = ds_list[[subgroup]], 
                           method = "wilcox", group = group, 
                           filter_thres = filter_thres,
                           taxa_level = taxa_level, 
                           p_adjust_method = p_adjust_method)
    )
    if ("try-error" %in% class(fit)) {
      next
    }
    all_diff <- t1$res_diff
    all_diff %<>%  subset(Significance != "ns")
    if (only_calculate) {
      new_list[[subgroup]] <- list(trans_diff = t1, all_diff = all_diff)
    }else{
      
      for (tmp_p in c(p_cutoff, .05, .01, .005, 1e-3, 1e-4, 1e-5)) {
        
        if ( sum(all_diff$P.unadj < tmp_p) < 1) {
          next
        }
        my_barplot <- t1$plot_diff_bar(threshold = 1-tmp_p, 
                                   color_values = color_values, 
                                   group_order = group_order)+
          labs(title = paste0("p value = ", tmp_p))
      
        if (nrow(my_barplot$data) < barplot_num) {
          break
        }
      }
      
      sel_diff <- all_diff[all_diff$P.unadj < p_cutoff, ]
      sel_list <- sel_diff$Taxa
      
      abund_plot <- t1$plot_diff_abund(use_number = 1:sum(all_diff$P.unadj < tmp_p), 
                                       color_values = color_values, 
                                       group_order = group_order, coord_flip = F)+
            labs(title = paste0("p value < ", tmp_p))
      
      
      new_list[[subgroup]] <- list(trans_diff = t1, diff_bar = my_barplot,
                                   all_diff = all_diff, sel_diff = sel_diff,
                                   sel_list = sel_list, abund_plot = abund_plot) 
     
    }
    message("\nFinished ", subgroup, " at ", date(), "\n") 
  }
   message("\nFeatureSel_LEfSe has finished at ", date(), "\n")
  return(new_list)
}


```

#### qFC

Quartile Fold Change

```{r qFC}
qFC <- function(g1, g2, min = 0.01, quartile = seq(.1, .9, .1)){
  min_value <- 0.01*min(c(g1[g1!=0], g2[g2!=0]))
  g1_ <- g1; g1_[g1==0] <- min_value 
  g2_ <- g2; g2_[g2==0] <- min_value 
  fc_list <- quantile(g1_, quartile) / quantile(g2_, quartile)
  mfc <- mean(fc_list)
  return(mfc)
}

```

#### QuanFC

```{r qfc}
QuanFC <- function(g1, g2){
  mean((quantile(g1, seq(.1, .9, .1))/quantile(g2, seq(.1, .9, .1))))
}
```

#### Wil_Matrix

```{r Wil_Matrix}
Wil_Matrix <- function(sub_otu, sub_meta, sub_catolog, 
                       sub_comparison, comparison_rank,
                       rarefy_cutoff = 0.01, p_cutoff = 0.05, rarefy_or_not = T){
  if (rarefy_or_not) {
    merge_df <- merge(t(sub_otu[rowMeans(sub_otu) > rarefy_cutoff/100, ]),
                      sub_meta[, c(sub_catolog, sub_comparison)],
                      by = "row.names") %>% column_to_rownames("Row.names")
  }else{
    merge_df <- merge(t(sub_otu),
                      sub_meta[, c(sub_catolog, sub_comparison)],
                      by = "row.names") %>% column_to_rownames("Row.names")
  }
  
  
  merge_df_colname <- colnames(merge_df) 
  names(merge_df_colname) <- paste0("F", c(1:ncol(merge_df)))
  colnames(merge_df) <- paste0("F", c(1:ncol(merge_df)))
  colnames(merge_df)[c(ncol(merge_df)-c(1,0))] <- c(sub_catolog, sub_comparison)
  
  wil_df <- matrix(NA, nrow = ncol(merge_df)-2,
                   ncol = length(unique(merge_df[[sub_catolog]]))+1,
                   dimnames = list(colnames(merge_df)[-ncol(merge_df)+c(0,1)], 
                                   c(unique(merge_df[[sub_catolog]]), "All"))) %>%
    as.data.frame()
  FC_df <- wil_df
  
  for (sub_feature in colnames(merge_df)[-c(ncol(merge_df), ncol(merge_df)-1)]) {
    for (sub_subgroup in unique(merge_df[[sub_catolog]])) {
      if (length(unique(merge_df[merge_df[[sub_catolog]] == sub_subgroup,][[sub_comparison]]))==1) {
        wil_p <- NA
        tmp_fc <- NA
      }else{
        wil_p <- wilcox.test(formula = as.formula(paste0(sub_feature, " ~ ", sub_comparison)),
                             data = merge_df, subset = merge_df[[sub_catolog]] == sub_subgroup, exact=FALSE)$p.value
        tmp_fc <- QuanFC(g1 = merge_df[merge_df[[sub_catolog]] == sub_subgroup & merge_df[[sub_comparison]] == comparison_rank[1], sub_feature],
                         g2 = merge_df[merge_df[[sub_catolog]] == sub_subgroup & merge_df[[sub_comparison]] == comparison_rank[2], sub_feature])
      }
      wil_df[sub_feature, sub_subgroup] <- wil_p
      FC_df[sub_feature, sub_subgroup] <- tmp_fc
    }
    wil_p <- wilcox.test(formula = as.formula(paste0(sub_feature, " ~ ", sub_comparison)),
                         data = merge_df, exact=FALSE)$p.value
    tmp_fc <- QuanFC(g1 = merge_df[merge_df[[sub_comparison]] == comparison_rank[1], sub_feature],
                     g2 = merge_df[merge_df[[sub_comparison]] == comparison_rank[2], sub_feature])
    
    wil_df[sub_feature, "All"] <- wil_p
    FC_df[sub_feature, "All"] <- tmp_fc
  }
  wil_df[is.na(wil_df)] <- 1
  rownames(wil_df) <- merge_df_colname[rownames(wil_df)]
  rownames(FC_df) <- merge_df_colname[rownames(FC_df)]
  wil_sign_df <- ifelse(wil_df < p_cutoff, 'sign','non-sign') %>% as.data.frame()
  
  sel_wil_df <- wil_df[rowSums(wil_df[, 1:(ncol(wil_df)-1), drop = F] < p_cutoff) >= 1, ]
  sel_wil_sign_df <- ifelse(sel_wil_df < p_cutoff, 'sign','non-sign') %>% as.data.frame()
  
  return_list <- list(wil_df = wil_df, wil_sign_df = wil_sign_df,
                      sel_wil_df = sel_wil_df, sel_wil_sign_df = sel_wil_sign_df,
                      FC_df = FC_df)
  return(return_list)
}
```



#### MedianNormal

```{r MedianNormal}
MedianNormal <- function(raw_df, meta_df, rarefy_cutoff = 0.01, min_add = 0.0001){
  hRaw_df <- (raw_df[rowMeans(raw_df) >rarefy_cutoff/100, ] + min_add/100) %>% 
    t() %>% as.data.frame()
  merge_df <- merge(hRaw_df, by = "row.names",
                    meta_df[, c("Cohort_name", "Disease", "Response")]) %>%
    column_to_rownames("Row.names")
  
  mn_index_df <- merge_df %>% 
    group_by(Cohort_name, Disease, Response) %>%
    filter(Response == "R") %>%
    summarise_all('median') %>%
    subset(select = -c(Response)) %>%
    group_by(Cohort_name, Disease) %>%
    summarise_all(function(x) 1/x)
  
  nm_df <- hRaw_df
  for (subsample in rownames(nm_df)) {
    catology <- meta_df[subsample, c("Cohort_name", "Disease")] %>% as.character()
    nm_df[subsample, ] <- nm_df[subsample, ] * 
      mn_index_df[mn_index_df$Cohort_name == catology[1] & 
                    mn_index_df$Disease == catology[2], c(-1, -2)]
  }
  return(list(nm_df = nm_df, mn_index_df = mn_index_df))
}
```

#### MN_Idx_function

```{r MN_Idx_function}

MN_Idx_function <- function(otu_list, meta_list, cutoff_value = 1e-3, min_value = 1e-6, cover_value = 0.1){
  merge_otu <- lapply(otu_list, 
                      function(x) x %>% 
                        t() %>% # transfer the data.frame 
                        as.data.frame() %>% # convert the matrix to data.frame
                        rownames_to_column("rnames")) %>%  # keep the row names
    rbind.fill() %>% # combine multiple data.frame
    .[!duplicated((.$rnames)), ] %>% # only keep the first appear samples
    remove_rownames %>%  # remove the row.name 
    column_to_rownames('rnames') %>% # replace the row.name by rnames
    replace(is.na(.), 0) %>% # replace the NA by 0
    t() %>% # transfer the data.frame
    as.data.frame() # convert the matrix to data.frame
  
  
  
  sel_cand <- rownames(merge_otu)[apply(
    merge_otu, 1, function(x)
      ifelse(sum(x!=0) > cover_value*length(x), T, F) & # only keep the species appear in more than 10% samples (cover_value)
      ifelse(median(x[x!=0]) > cutoff_value, T, F))] # only select the candidates which median larger than 0.1% (1e-3) (cutoff_value)
  
  length(sel_cand)
  
  mod_otu <- (merge_otu + min_value) %>% apply(., 2, function(x) x/sum(x)) # add the min value 0.0001% (1e-6)
  
  sel_otu <- mod_otu[sel_cand, ] # select the sel_cand in modified otu table
  
  
  ## meta
  
  
  merge_meta <- lapply(meta_list, 
                       function(x) x %>% 
                         rownames_to_column("rnames")) %>%  # keep the row names
    rbind.fill() %>% # combine multiple data.frame
    .[!duplicated((.$rnames)), ] %>% # only keep the first appear samples
    remove_rownames %>%  # remove the row.name 
    column_to_rownames('rnames') # replace the row.name by rnames
  
  merge_meta$MN_idx <- ifelse(merge_meta$Period %in% c("Before_FMT", "Before_ICI") &
                                merge_meta$Response == "R", "base", "other")
  
  
  mn_index_df <- merge(sel_otu %>% t(), # transfer the otu table
                       merge_meta[, c("Cohort_name", "Disease", "MN_idx")],  # only select the useful column
                       by = 'row.names')  %>% # merge by row names
    column_to_rownames("Row.names") %>% # replace the row name by merge output
    group_by(Cohort_name, Disease, MN_idx) %>% # group by several column
    filter(MN_idx == "base") %>% # only select the MN_idx == "base", which means Response and before ICI/FMT
    summarise_all('median') %>% # calculate the median in each group
    subset(select = -c(MN_idx)) %>% # remove the useless column
    group_by(Cohort_name, Disease) %>% # group again
    summarise_all(function(x) 1/x) %>% # calculate the inverse 
    as.data.frame()
  
  freq_df <- table(merge_meta$Cohort_name, merge_meta$Disease) %>% as.data.frame() %>% filter(Freq != 0)
  if (nrow(mn_index_df) != nrow(freq_df)) {
    for (i in 1:nrow(freq_df)) {
      if (!any(((((mn_index_df[, 1] %>% unlist() %>% as.character()) == freq_df[i, 1]) & 
             ((mn_index_df[, 2] %>% unlist() %>% as.character()) == freq_df[i, 2]))))) {
        tmp_add <- cbind(freq_df[i, c(1,2)], matrix(rep(1e6, length(sel_cand)), nrow = 1))
        dimnames(tmp_add) <- list(nrow(mn_index_df)+1, colnames(mn_index_df))
        mn_index_df <- rbind(mn_index_df, tmp_add)
      }
    }
  }
  
  nm_df <- sel_otu %>% t() %>% as.data.frame()
  for (subsample in rownames(nm_df)) {
    catology <- merge_meta[subsample, c("Cohort_name", "Disease")] %>% as.character()
    nm_df[subsample, ] <- nm_df[subsample, ] * 
      mn_index_df[mn_index_df$Cohort_name == catology[1] & 
                    mn_index_df$Disease == catology[2], c(-1, -2)]
  }
  
  return(list(nm_df = nm_df,
              mn_index_df = mn_index_df,
              mod_otu = mod_otu,
              sel_otu = sel_otu,
              merge_meta = merge_meta))
}


```

#### FoldChange

```{r foldchange}

FoldChange <- function(sub_otu, sub_meta, sub_catolog, sub_comparison, comp_rank){
  merge_df <- merge(t(sub_otu),
                    sub_meta[, c(sub_catolog, sub_comparison)],
                    by = "row.names") %>% column_to_rownames("Row.names")
  merge_df_colname <- colnames(merge_df) 
  names(merge_df_colname) <- paste0("F", c(1:ncol(merge_df)))
  colnames(merge_df) <- paste0("F", c(1:ncol(merge_df)))
  colnames(merge_df)[c(ncol(merge_df)-c(1,0))] <- c(sub_catolog, sub_comparison)
  
  FC_df <- matrix(NA, nrow = ncol(merge_df)-2,
                   ncol = length(unique(merge_df[[sub_catolog]]))+1,
                   dimnames = list(colnames(merge_df)[-ncol(merge_df)+c(0,1)], 
                                   c(unique(merge_df[[sub_catolog]]), "All"))) %>%
    as.data.frame()
  
  for (sub_feature in colnames(merge_df)[-c(ncol(merge_df), ncol(merge_df)-1)]) {
    for (sub_subgroup in unique(merge_df[[sub_catolog]])) {
      FC_value <- (mean(merge_df[merge_df[[sub_catolog]] == sub_subgroup & 
                                  merge_df[[sub_comparison]] == comp_rank[1], sub_feature])/ 
        mean(merge_df[merge_df[[sub_catolog]] == sub_subgroup & 
                        merge_df[[sub_comparison]] == comp_rank[2], sub_feature])) %>% log2()
      FC_df[sub_feature, sub_subgroup] <- FC_value
    }
    FC_value <- (mean(merge_df[merge_df[[sub_comparison]] == comp_rank[1], sub_feature])/ 
        mean(merge_df[merge_df[[sub_comparison]] == comp_rank[2], sub_feature])) %>% log2()
    FC_df[sub_feature, "All"] <- FC_value
  }
  # FC_df[is.na(FC_df)] <- 1
  rownames(FC_df) <- merge_df_colname[rownames(FC_df)]
  return(FC_df)
}
```

#### Save_MN_Idx

```{r Save_MN_Idx}
Save_MN_Idx <- function(mn_list, kingdom, cutoff, min_value, cover = '10p', DirPath){
  prefix <- paste0(kingdom, 
                   "-cover_", cover, 
                   "-cutoff_", cutoff,
                   "-minValue_", min_value)
  DirPath2 <- paste0(DirPath, "01.PreProcessing/MN_Idx/", 
                     "MN_res-",prefix)
  if (!dir.exists(DirPath2)) {
    dir.create(DirPath2,recursive =TRUE)
  }
  cur_path <- getwd()
  setwd(DirPath2)
  write.csv(x = mn_list$nm_df,
            file = FileCreate(Prefix = paste0("NormalMedian-SelCandidates-Matrix"), 
                              Suffix = "csv"))
  write.csv(x = mn_list$mn_index_df, 
            file = FileCreate(Prefix = paste0("NormalMedian-Index-Matrix"), 
                              Suffix = "csv"))
  write.csv(x = mn_list$mod_otu, 
            file = FileCreate(Prefix = paste0("ModRelAbundance-Matrix"), 
                              Suffix = "csv"))
  write.csv(x = mn_list$sel_otu, 
            file = FileCreate(Prefix = paste0("ModRelAbundance-SelCandidates-Matrix"), 
                              Suffix = "csv"))
  write.csv(x = mn_list$merge_meta, 
            file = FileCreate(Prefix = paste0("Merge_meta_info-Matrix"), 
                              Suffix = "csv"))
  saveRDS(object = mn_list, 
          file = FileCreate(Prefix = paste0(kingdom, "_mn_res"),
                            Suffix = "rds"))
  setwd(cur_path)
}
```

#### ShortName

```{r shortname}
ShortName <- function(x){
  lapply(x, function(x) {
                   k <- strsplit( x = x, split = "\\|") %>%
                     unlist(); 
                   return(k[length(k)])
                   }) %>% 
                 unlist()
}
```

## Parameters

```{r parameters}
respond_col <- c(NR = '#C71E1D', R = '#015069')

tax_col <- c(k__Bacteria = "#f9b4ab", k__Eukaryota = "#FDEBD3",
             k__Archaea = "#7aa9d1", k__Viruses = "#96b6ad")

cohort_col <- c(`2017_Frankel` = "#EE6B3B", `2018_Gopalakrishnan` = "#A02C5D",
                `2018_Maston` = "#EC0F47", `2018_Routy` = "#700460",
                `2019_Peters` = "#022C7A", `2019_Zheng` = "#1A1333",
                `2021_Baruch` = "#FBBF54", `2021_Spencer` = "#045459",
                `2021_Davar` = "#ABD96D", `2022_McCulloch` = "#15C286",
                `2022_Lee` = "#262949", `2022_Derosa` = "#087353")

cancer_col <- c(MM = "#7f58af", HCC = '#64C5EB', 
                NSCLC = '#E84D8A', RCC = '#FEB326')
save_path_prefix <- "../../../cuhk/metagenome/20211108-LYF-Immunotherapy_Cancer/"
  
```


## Loading data 

```{r loading data}
# loading data
# bICI_ds_list <- readRDS(paste0(save_path_prefix, "01.PreProcessing/RDS/2022-09-22-bICI_ds_list_subKingdom-v1.0.rds"))
# bICI_rare_ds_list <- readRDS(paste0(save_path_prefix, "01.PreProcessing/RDS/2022-09-22-bICI_rare_ds_list_subKingdom-v1.0.rds"))
# bICI_ds_list <- readRDS(paste0(save_path_prefix, "01.PreProcessing/RDS/2022-09-22-bICI_rare_ds_list_subKingdom-v1.0.rds"))
bICI_ds_list <- readRDS(paste0(save_path_prefix, "01.PreProcessing/RDS/2022-10-19-bICI_rare_ds_list_subKingdom-v1.0.rds"))
pICI_ds_list <- readRDS(paste0(save_path_prefix, "01.PreProcessing/RDS/2022-12-15-pICI_rare_ds_list_subKingdom-v1.0.rds"))
iICI_ds_list <- readRDS(paste0(save_path_prefix, "01.PreProcessing/RDS/2022-12-15-iICI_rare_ds_list_subKingdom-v1.0.rds"))
# bICI_ds_list <- readRDS(paste0(save_path_prefix, "01.PreProcessing/RDS/2022-10-19-bICI_filter_ds_list_subKingdom-v1.0.rds"))

sub_otu <- bICI_ds_list$k__Bacteria$taxa_abund$Species
sub_meta <- bICI_ds_list$k__Bacteria$sample_table
sub_catolog <- "Disease"
sub_comparison <- "Response"
rarefy_cutoff <- .01
p_cutoff <- .05

```

<!--
## Pre-process data

```{r preprocess data}
cutoff_value <- 1e-3
min_value <- 1e-6

######################## Bacteria ########################

bac_otu_list <- list(bICI_ds_list$k__Bacteria$taxa_abund$Species,
                     pICI_ds_list$k__Bacteria$taxa_abund$Species,
                     iICI_ds_list$k__Bacteria$taxa_abund$Species)

bac_meta_list <- list(bICI_ds_list$k__Bacteria$sample_table,
                      pICI_ds_list$k__Bacteria$sample_table,
                      iICI_ds_list$k__Bacteria$sample_table)

bac_mn_res <- MN_Idx_function(otu_list = bac_otu_list, 
                              meta_list = bac_meta_list, 
                              cutoff_value = 1e-3,
                              min_value = 1e-6 )

Save_MN_Idx(mn_list = bac_mn_res, 
            kingdom = "Bacteria", 
            cutoff = "1e-3", 
            min_value = "1e-6", 
            cover = "10p", 
            DirPath = save_path_prefix)

######################## Eukaryota ########################

euk_otu_list <- list(bICI_ds_list$k__Eukaryota$taxa_abund$Species,
                     pICI_ds_list$k__Eukaryota$taxa_abund$Species,
                     iICI_ds_list$k__Eukaryota$taxa_abund$Species)

euk_meta_list <- list(bICI_ds_list$k__Eukaryota$sample_table,
                      pICI_ds_list$k__Eukaryota$sample_table,
                      iICI_ds_list$k__Eukaryota$sample_table)

euk_mn_res <- MN_Idx_function(otu_list = euk_otu_list, 
                              meta_list = euk_meta_list, 
                              cutoff_value = 1e-3,
                              min_value = 1e-6 )

Save_MN_Idx(mn_list = euk_mn_res, kingdom = "Eukaryota", 
            cutoff = "1e-3", 
            min_value = "1e-6", 
            cover = "10p", 
            DirPath = save_path_prefix)

######################## Archaea ########################

arc_otu_list <- list(bICI_ds_list$k__Archaea$taxa_abund$Species,
                     pICI_ds_list$k__Archaea$taxa_abund$Species,
                     iICI_ds_list$k__Archaea$taxa_abund$Species)

arc_meta_list <- list(bICI_ds_list$k__Archaea$sample_table,
                      pICI_ds_list$k__Archaea$sample_table,
                      iICI_ds_list$k__Archaea$sample_table)

arc_mn_res <- MN_Idx_function(otu_list = arc_otu_list, 
                              meta_list = arc_meta_list, 
                              cutoff_value = 1e-3,
                              min_value = 1e-6 )

Save_MN_Idx(mn_list = arc_mn_res, kingdom = "Archaea", 
            cutoff = "1e-3", 
            min_value = "1e-6", 
            cover = "10p", 
            DirPath = save_path_prefix)

######################## Viruses ########################

vir_otu_list <- list(bICI_ds_list$k__Viruses$taxa_abund$Species,
                     pICI_ds_list$k__Viruses$taxa_abund$Species,
                     iICI_ds_list$k__Viruses$taxa_abund$Species)

vir_meta_list <- list(bICI_ds_list$k__Viruses$sample_table,
                      pICI_ds_list$k__Viruses$sample_table,
                      iICI_ds_list$k__Viruses$sample_table)

vir_mn_res <- MN_Idx_function(otu_list = vir_otu_list, 
                              meta_list = vir_meta_list, 
                              cutoff_value = 1e-3,
                              min_value = 1e-6 )

Save_MN_Idx(mn_list = vir_mn_res, kingdom = "Viruses", 
            cutoff = "1e-3", 
            min_value = "1e-6", 
            cover = "10p", 
            DirPath = save_path_prefix)


dim(x = bac_mn_res$sel_otu)
# [1]   49 1356
dim(x = euk_mn_res$sel_otu)
# [1]   48 1047
dim(x = arc_mn_res$sel_otu)
# [1]  11 426
dim(x = vir_mn_res$sel_otu)
# [1]  35 929

```
-->

## Pre-process data

```{r preprocess data}
calculate_or_not <- F
cutoff_value <- 1e-4
min_value <- 1e-6

######################## Bacteria ########################
if (calculate_or_not) {
  bac_otu_list <- list(bICI_ds_list$k__Bacteria$taxa_abund$Species,
                       pICI_ds_list$k__Bacteria$taxa_abund$Species,
                       iICI_ds_list$k__Bacteria$taxa_abund$Species)
  
  bac_meta_list <- list(bICI_ds_list$k__Bacteria$sample_table,
                        pICI_ds_list$k__Bacteria$sample_table,
                        iICI_ds_list$k__Bacteria$sample_table)
  
  bac_mn_res2 <- MN_Idx_function(otu_list = bac_otu_list,
                                 meta_list = bac_meta_list, 
                                 cutoff_value = 1e-4,
                                 min_value = 1e-6 )
  
  Save_MN_Idx(mn_list = bac_mn_res2, 
              kingdom = "Bacteria", 
              cutoff = "1e-4", 
              min_value = "1e-6", 
              cover = "10p", 
              DirPath = save_path_prefix)
  
  ######################## Eukaryota ########################
  
  euk_otu_list <- list(bICI_ds_list$k__Eukaryota$taxa_abund$Species,
                       pICI_ds_list$k__Eukaryota$taxa_abund$Species,
                       iICI_ds_list$k__Eukaryota$taxa_abund$Species)
  
  euk_meta_list <- list(bICI_ds_list$k__Eukaryota$sample_table,
                        pICI_ds_list$k__Eukaryota$sample_table,
                        iICI_ds_list$k__Eukaryota$sample_table)
  
  euk_mn_res2 <- MN_Idx_function(otu_list = euk_otu_list, 
                                 meta_list = euk_meta_list, 
                                 cutoff_value = 1e-4,
                                 min_value = 1e-6 )
  
  Save_MN_Idx(mn_list = euk_mn_res2, kingdom = "Eukaryota", 
              cutoff = "1e-4", 
              min_value = "1e-6", 
              cover = "10p", 
              DirPath = save_path_prefix)
  
  ######################## Archaea ########################
  
  arc_otu_list <- list(bICI_ds_list$k__Archaea$taxa_abund$Species,
                       pICI_ds_list$k__Archaea$taxa_abund$Species,
                       iICI_ds_list$k__Archaea$taxa_abund$Species)
  
  arc_meta_list <- list(bICI_ds_list$k__Archaea$sample_table,
                        pICI_ds_list$k__Archaea$sample_table,
                        iICI_ds_list$k__Archaea$sample_table)
  
  arc_mn_res2 <- MN_Idx_function(otu_list = arc_otu_list, 
                                 meta_list = arc_meta_list, 
                                 cutoff_value = 1e-4,
                                 min_value = 1e-6 )
  
  Save_MN_Idx(mn_list = arc_mn_res2, kingdom = "Archaea", 
              cutoff = "1e-4", 
              min_value = "1e-6", 
              cover = "10p", 
              DirPath = save_path_prefix)
  
  ######################## Viruses ########################
  
  vir_otu_list <- list(bICI_ds_list$k__Viruses$taxa_abund$Species,
                       pICI_ds_list$k__Viruses$taxa_abund$Species,
                       iICI_ds_list$k__Viruses$taxa_abund$Species)
  
  vir_meta_list <- list(bICI_ds_list$k__Viruses$sample_table,
                        pICI_ds_list$k__Viruses$sample_table,
                        iICI_ds_list$k__Viruses$sample_table)
  
  vir_mn_res2 <- MN_Idx_function(otu_list = vir_otu_list, 
                                 meta_list = vir_meta_list, 
                                 cutoff_value = 1e-4,
                                 min_value = 1e-6 )
  Save_MN_Idx(mn_list = vir_mn_res2, kingdom = "Viruses", 
              cutoff = "1e-4", 
              min_value = "1e-6", 
              cover = "10p", 
              DirPath = save_path_prefix)
  
  dim(x = bac_mn_res2$sel_otu)
  # [1]  254 1356
  dim(x = euk_mn_res2$sel_otu)
  # [1] 2557 1047
  dim(x = arc_mn_res2$sel_otu)
  # [1] 179 426
  dim(x = vir_mn_res2$sel_otu)
  # [1] 127 929
}else{
  tmp_path <- getwd()
  setwd(paste0(save_path_prefix, 
               "01.PreProcessing/MN_Idx/", 
               "MN_res-Bacteria-cover_10p-cutoff_1e-4-minValue_1e-6/"))
  bac_mn_res2 <- readRDS(file = "2023-03-15-Bacteria_mn_res-v1.0.rds")
  setwd(tmp_path)
  
  setwd(paste0(save_path_prefix, 
               "01.PreProcessing/MN_Idx/", 
               "MN_res-Eukaryota-cover_10p-cutoff_1e-4-minValue_1e-6/"))
  euk_mn_res2 <- readRDS(file = "2023-03-15-Eukaryota_mn_res-v1.0.rds")
  setwd(tmp_path)
  
  setwd(paste0(save_path_prefix, 
               "01.PreProcessing/MN_Idx/", 
               "MN_res-Archaea-cover_10p-cutoff_1e-4-minValue_1e-6/"))
  arc_mn_res2 <- readRDS("2023-03-15-Archaea_mn_res-v1.0.rds")
  setwd(tmp_path)
  
  setwd(paste0(save_path_prefix, 
               "01.PreProcessing/MN_Idx/", 
               "MN_res-Viruses-cover_10p-cutoff_1e-4-minValue_1e-6/"))
  vir_mn_res2 <- readRDS("2023-03-15-Viruses_mn_res-v1.0.rds")
  setwd(tmp_path)
  
}


```


## Feature Selection

- MN_Idx

- before ICI

### Wilcoxon test (Cancer type)

```{r Wilcox CancerType}
calculate_or_not <- F

if (calculate_or_not) {
  bICI_sel_list <- list() 
  # store the selected candidates (10% present, mean of non-zero abundance larger than 1e-4.)
  # Bacteria 254/892; Eukaryota 2557/703; Archaea 179/279; Viruses 127/596
  ###########################################################################################
  
  # bacteria
  bICI_sel_list$k__Bacteria <- t(bac_mn_res2$nm_df)[,rownames(bICI_ds_list$k__Bacteria$sample_table)]
  wil_bac_sp_mn <- Wil_Matrix(sub_otu = bICI_sel_list$k__Bacteria, p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = bICI_ds_list$k__Bacteria$sample_table, comparison_rank = c("R", "NR"),
                              sub_catolog = "Disease", sub_comparison = "Response")
  # calculate the Wilcox different between Response vs non-Response, separate via cancer type (MM, NSCLC, RCC, HCC) in Bacteria
  #############################################################################################################################
  mn_bac_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                           Prefix = "Wilcox-MN_Idx-Bacteria-bICI", Suffix = "rds")
  saveRDS(object = wil_bac_sp_mn, file = mn_bac_rds)
  
  # Eukaryota
  bICI_sel_list$k__Eukaryota <- t(euk_mn_res2$nm_df)[,rownames(bICI_ds_list$k__Eukaryota$sample_table)]
  wil_euk_sp_mn <- Wil_Matrix(sub_otu = bICI_sel_list$k__Eukaryota, p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = bICI_ds_list$k__Eukaryota$sample_table, comparison_rank = c("R", "NR"),
                              sub_catolog = "Disease", sub_comparison = "Response")
  # calculate the Wilcox different between Response vs non-Response, separate via cancer type (MM, NSCLC, RCC, HCC) in Eukaryota
  ##############################################################################################################################
  mn_euk_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                           Prefix = "Wilcox-MN_Idx-Eukaryota-bICI", Suffix = "rds")
  saveRDS(object = wil_euk_sp_mn, file = mn_euk_rds)
  
  
  # Archaea
  bICI_sel_list$k__Archaea <- t(arc_mn_res2$nm_df)[,rownames(bICI_ds_list$k__Archaea$sample_table)]
  wil_arc_sp_mn <- Wil_Matrix(sub_otu = bICI_sel_list$k__Archaea, p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = bICI_ds_list$k__Archaea$sample_table, comparison_rank = c("R", "NR"),
                              sub_catolog = "Disease", sub_comparison = "Response")
  # calculate the Wilcox different between Response vs non-Response, separate via cancer type (MM, NSCLC, RCC, HCC) in Archaea
  ############################################################################################################################
  mn_arc_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                           Prefix = "Wilcox-MN_Idx-Archaea-bICI", Suffix = "rds")
  saveRDS(object = wil_arc_sp_mn, file = mn_arc_rds)
  
  
  # Viruses
  bICI_sel_list$k__Viruses <- t(vir_mn_res2$nm_df)[,rownames(bICI_ds_list$k__Viruses$sample_table)]
  wil_vir_sp_mn <- Wil_Matrix(sub_otu = bICI_sel_list$k__Viruses, p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = bICI_ds_list$k__Viruses$sample_table, comparison_rank = c("R", "NR"),
                              sub_catolog = "Disease", sub_comparison = "Response")
  # calculate the Wilcox different between Response vs non-Response, separate via cancer type (MM, NSCLC, RCC, HCC) in Viruses
  ############################################################################################################################
  mn_vir_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                           Prefix = "Wilcox-MN_Idx-Viruses-bICI", Suffix = "rds")
  saveRDS(object = wil_vir_sp_mn, file = mn_vir_rds)
  
  # bICI_sel_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/"),
  #                          Prefix = "bICI-MN_Idx", Suffix = "rds")
  # bICI_sel_list
  
  
  
  
}else{
  wil_bac_sp_mn <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-03-15-Wilcox-MN_Idx-",
                                         "Bacteria", "-bICI-v1.0.rds"))
  wil_euk_sp_mn <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-03-15-Wilcox-MN_Idx-",
                                         "Eukaryota", "-bICI-v1.0.rds"))
  wil_arc_sp_mn <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-03-15-Wilcox-MN_Idx-",
                                         "Archaea", "-bICI-v1.0.rds"))
  wil_vir_sp_mn <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-03-15-Wilcox-MN_Idx-",
                                         "Viruses", "-bICI-v1.0.rds"))
}

# check how many species show significant in different kingdom
nrow(wil_bac_sp_mn$sel_wil_df) # 84  
nrow(wil_euk_sp_mn$sel_wil_df) # 722
nrow(wil_arc_sp_mn$sel_wil_df) # 80
nrow(wil_vir_sp_mn$sel_wil_df) # 34

# check how many species show significant in at least 2 cancer types across different kingdom
sum(rowSums(wil_bac_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 8   
sum(rowSums(wil_euk_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 39
sum(rowSums(wil_arc_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 4
sum(rowSums(wil_vir_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 0

sum(rowSums(wil_bac_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 0  
sum(rowSums(wil_euk_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 1
sum(rowSums(wil_arc_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 0
sum(rowSums(wil_vir_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 0

################################
# transform the matrix to melt
# Feature   Cancer Type   Select
# Bac 1     MM            sign
# Bac 2     MM            sign
################################
melt_wil_mn_list <- list()

melt_wil_mn_list$k__Bacteria <- melt(as.matrix(wil_bac_sp_mn$wil_sign_df[, 1:4]),
                                  varnames = c("Feature", "CancerType"), 
                                  value.name = "Select") %>% subset(Select == "sign")
melt_wil_mn_list$k__Eukaryota <- melt(as.matrix(wil_euk_sp_mn$wil_sign_df[, 1:4]),
                                   varnames = c("Feature", "CancerType"), 
                                   value.name = "Select") %>% subset(Select == "sign")
melt_wil_mn_list$k__Archaea <- melt(as.matrix(wil_arc_sp_mn$wil_sign_df[, 1:4]),
                                 varnames = c("Feature", "CancerType"), 
                                 value.name = "Select") %>% subset(Select == "sign")
melt_wil_mn_list$k__Viruses <- melt(as.matrix(wil_vir_sp_mn$wil_sign_df[, 1:4]),
                                 varnames = c("Feature", "CancerType"), 
                                 value.name = "Select") %>% subset(Select == "sign")

```

### Wilcoxon test (CohDis)

```{r wilcox CohDis}

calculate_or_not <- F
if (calculate_or_not) {
  # Temp meta info (Bacteria)
  tmp_meta <- bICI_ds_list$k__Bacteria$sample_table
  # Add the new group (CohDis)
  tmp_meta$CohDis <- paste0(tmp_meta$Cohort_name, "_", tmp_meta$Disease)
  # Calculate the Wilcox between Response and sub-group in CohDis
  wil_bac_sp_cohort <- Wil_Matrix(sub_otu = bICI_sel_list$k__Bacteria, p_cutoff = 0.05, rarefy_or_not = F,
                                  sub_meta = tmp_meta,
                                  comparison_rank = c("R", "NR"),
                                  sub_catolog = "CohDis", sub_comparison = "Response")
  # Save the rds
  co_bac_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds"),
                           Prefix = "Wilcox-MN_Idx-Bacteria-CohDis", Suffix = "rds")
  saveRDS(object = wil_bac_sp_cohort, file = co_bac_rds)
  
  # Temp meta info (Eukaryota)
  tmp_meta <- bICI_ds_list$k__Eukaryota$sample_table
  # Add the new group (CohDis)
  tmp_meta$CohDis <- paste0(tmp_meta$Cohort_name, "_", tmp_meta$Disease)
  # Calculate the Wilcox between Response and sub-group in CohDis
  wil_euk_sp_cohort <- Wil_Matrix(sub_otu = bICI_sel_list$k__Eukaryota, p_cutoff = 0.05, rarefy_or_not = F,
                                  sub_meta = tmp_meta,
                                  comparison_rank = c("R", "NR"),
                                  sub_catolog = "CohDis", sub_comparison = "Response")
  co_euk_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds"),
                           Prefix = "Wilcox-MN_Idx-Eukaryota-CohDis", Suffix = "rds")
  saveRDS(object = wil_euk_sp_cohort, file = co_euk_rds)
  
  # Temp meta info (Archaea)
  tmp_meta <- bICI_ds_list$k__Archaea$sample_table
  # Add the new group (CohDis)
  tmp_meta$CohDis <- paste0(tmp_meta$Cohort_name, "_", tmp_meta$Disease)
  # Calculate the Wilcox between Response and sub-group in CohDis
  wil_arc_sp_cohort <- Wil_Matrix(sub_otu = bICI_sel_list$k__Archaea, p_cutoff = 0.05, rarefy_or_not = F,
                           sub_meta = tmp_meta,
                              comparison_rank = c("R", "NR"),
                           sub_catolog = "CohDis", sub_comparison = "Response")
  co_arc_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds"),
                           Prefix = "Wilcox-MN_Idx-Archaea-CohDis", Suffix = "rds")
  saveRDS(object = wil_arc_sp_cohort, file = co_arc_rds)
  
  # Temp meta info (Viruses)
  tmp_meta <- bICI_ds_list$k__Viruses$sample_table
  # Add the new group (CohDis)
  tmp_meta$CohDis <- paste0(tmp_meta$Cohort_name, "_", tmp_meta$Disease)
  # Calculate the Wilcox between Response and sub-group in CohDis
  wil_vir_sp_cohort <- Wil_Matrix(sub_otu = bICI_sel_list$k__Viruses, p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = tmp_meta,
                              comparison_rank = c("R", "NR"),
                              sub_catolog = "CohDis", sub_comparison = "Response")
  co_vir_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds"),
                           Prefix = "Wilcox-MN_Idx-Viruses-CohDis", Suffix = "rds")
  saveRDS(object = wil_vir_sp_cohort, file = co_vir_rds)
}else{
  wil_bac_sp_cohort <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds/2023-03-15-Wilcox-MN_Idx-",
                                         "Bacteria-CohDis", "-v1.0.rds"))
  wil_euk_sp_cohort <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds/2023-03-15-Wilcox-MN_Idx-",
                                         "Eukaryota-CohDis", "-v1.0.rds"))
  wil_arc_sp_cohort <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds/2023-03-15-Wilcox-MN_Idx-",
                                         "Archaea-CohDis", "-v1.0.rds"))
  wil_vir_sp_cohort <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds/2023-03-15-Wilcox-MN_Idx-",
                                         "Viruses-CohDis", "-v1.0.rds"))
}



nrow(wil_bac_sp_cohort$sel_wil_df) # 130
nrow(wil_euk_sp_cohort$sel_wil_df) # 1117
nrow(wil_arc_sp_cohort$sel_wil_df) # 51
nrow(wil_vir_sp_cohort$sel_wil_df) # 25


sum(rowSums(wil_bac_sp_cohort$sel_wil_df[, 1:11] <.05) >= 2) # 45
sum(rowSums(wil_euk_sp_cohort$sel_wil_df[, 1:11] <.05) >= 2) # 198
sum(rowSums(wil_arc_sp_cohort$sel_wil_df[, 1:11] <.05) >= 2) # 8
sum(rowSums(wil_vir_sp_cohort$sel_wil_df[, 1:11] <.05) >= 2) # 2

sum(rowSums(wil_bac_sp_cohort$sel_wil_df[, 1:11] <.05) >= 3) # 12
sum(rowSums(wil_euk_sp_cohort$sel_wil_df[, 1:11] <.05) >= 3) # 29
sum(rowSums(wil_arc_sp_cohort$sel_wil_df[, 1:11] <.05) >= 3) # 0
sum(rowSums(wil_vir_sp_cohort$sel_wil_df[, 1:11] <.05) >= 3) # 0

sum(rowSums(wil_bac_sp_cohort$sel_wil_df[, 1:11] <.05) >= 4) # 1 
sum(rowSums(wil_euk_sp_cohort$sel_wil_df[, 1:11] <.05) >= 4) # 3 

rownames(wil_bac_sp_cohort$sel_wil_df)[rowSums(wil_bac_sp_cohort$sel_wil_df[, 1:11] <.05) >= 4]
# s__Faecalibacterium_prausnitzii

rownames(wil_euk_sp_cohort$sel_wil_df)[rowSums(wil_euk_sp_cohort$sel_wil_df[, 1:11] <.05) >= 4]
# s__Tetrahymena_elliotti
# s__Endocarpon_pusillum
# s__Microthyrium_microscopicum


################################
# transform the matrix to melt
# Feature   CohDis        Select
# Bac 1     xx_xx         sign
# Bac 2     xx_xx         sign
################################
melt_wil_cohort_list <- list()

melt_wil_cohort_list$k__Bacteria <- melt(as.matrix(wil_bac_sp_cohort$wil_sign_df[, 1:11]),
                                  varnames = c("Feature", "CohDis"), 
                                  value.name = "Select") %>% subset(Select == "sign")
melt_wil_cohort_list$k__Eukaryota <- melt(as.matrix(wil_euk_sp_cohort$wil_sign_df[, 1:11]),
                                   varnames = c("Feature", "CohDis"), 
                                   value.name = "Select") %>% subset(Select == "sign")
melt_wil_cohort_list$k__Archaea <- melt(as.matrix(wil_arc_sp_cohort$wil_sign_df[, 1:11]),
                                 varnames = c("Feature", "CohDis"), 
                                 value.name = "Select") %>% subset(Select == "sign")
melt_wil_cohort_list$k__Viruses <- melt(as.matrix(wil_vir_sp_cohort$wil_sign_df[, 1:11]),
                                 varnames = c("Feature", "CohDis"), 
                                 value.name = "Select") %>% subset(Select == "sign")
```


## UpSet Plot (Wilcox MN)

### All the significant features

```{r upset all}
##############################################################################
# plot all the combination, including the characters only appeared in one set.
##############################################################################
lt_list2 <- list()
# store all the significant candidates
# format: lt_list2$kingdom$cancertype <- c(cand1, cand2, ...)
m_list2 <- list()
# store the make_comb_mat 
# format: m_list2$kingdom <- make_comb_mat(lt_list2$kingdom)
lt2 <- list()
# as same as lt_list2
# format: lt$cancertype <- c(cand1, cand2, ...) [all kingdoms combine]

save_or_not <- F


for (sel_kingdoms in names(melt_wil_mn_list)) {
  lt_list2[[sel_kingdoms]] <- list()
  for (sel_disease in levels(unique(melt_wil_mn_list[[sel_kingdoms]]$CancerType))) {
    lt_list2[[sel_kingdoms]][[sel_disease]] <- melt_wil_mn_list[[sel_kingdoms]] %>%
      subset(CancerType == sel_disease, select = "Feature") %>% unlist() %>% as.character()
    lt2[[sel_disease]] <- c(lt2[[sel_disease]], lt_list2[[sel_kingdoms]][[sel_disease]])
  }
  m_list2[[sel_kingdoms]] <- make_comb_mat(lt_list2[[sel_kingdoms]])
}
m2 = make_comb_mat(lt2)# as same as m2

# store the set_size, the num of sign. cand. in sub-cancer-type and sub-kingdom
rowAnn_mt2 <- data.frame(k__Bacteria = set_size(m_list2$k__Bacteria),
                         k__Eukaryota = set_size(m_list2$k__Eukaryota),
                         k__Archaea = set_size(m_list2$k__Archaea),
                         k__Viruses = set_size(m_list2$k__Viruses)) %>%
  as.matrix()

# store the combination, the num of sign. cand. in combinations
colAnn_mt2 <- matrix(data = 0, 
                    nrow = length(comb_name(m2)), 
                    ncol = 4,
                    dimnames = list(comb_name(m2),
                                    names(tax_col)))

for (sub_k in names(m_list2)) {
  for (InSize in names(comb_size(m_list2[[sub_k]]))) {
    colAnn_mt2[InSize, sub_k] <- comb_size(m_list2[[sub_k]])[InSize]
  }
}

# store the detail of candidates
# format: MM    NSCLC RCC   HCC   Num Label           Group_Num
#         0/1   0/1   0/1   0/1   n   species name    [1:4]
label_df <- matrix(data = NA, nrow = 0, 
                   ncol = c(length(x = set_name(m2)) + 2),
                   dimnames = list(c(), c( set_name(m2), "Num", "Label"))) %>%
  as.data.frame()
label_colname <- colnames(label_df)

for (sel_cname in comb_name(m2)) {
  tmp_lab <- extract_comb(m2, sel_cname)
  tmp_df <- matrix(rep(c(strsplit(x = sel_cname, split = "") %>% 
                         unlist() %>% as.numeric(), length(tmp_lab)), 
                       each = length(tmp_lab)),
                   nrow = length(tmp_lab)) %>%
    cbind(., tmp_lab) 
  colnames(tmp_df) <- label_colname
  label_df <- rbind(label_df, tmp_df)
}



label_df$Group_Num <-  apply(label_df[, c(1:length(set_name(m2)))], 1, function(x) as.numeric(x) %>% sum)

if (save_or_not) {
  label_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Upset"),
                          Prefix = "Melt-label-Wilcox_MN", Suffix = "csv")
  write.csv(x = label_df, file = label_csv, row.names = F)
}


# show the upset plot (colorful version)
tmp_ht2 <- UpSet(m2, # import make_comb_mat file
                 set_order = order(set_size(m2), decreasing = T), 
                 # order from biggest to smallest via set_size (row)
                 heatmap_height = unit(length(set_name(m2))*1.5, "cm"),
                 heatmap_width = unit(length(comb_name(m2))*1.5, "cm"),
                 pt_size = unit(5, "mm"), lwd = 3, 
                 # set the point size
                 bg_col = c("#A9C0C9", "#D8D7CD"),
                 # set the background color
                 comb_order = order(lapply(comb_name(m2), function(x) {
                   strsplit(x = x, split = "") %>% 
                     unlist() %>%
                     as.numeric() %>% 
                     sum()
                   }) %>% unlist() * -1, 
                   comb_size(m2), 
                   decreasing = T),
                 # order from least overlap to most (column)
                 top_annotation = HeatmapAnnotation(
                   "Intersection\nsize" = anno_barplot(colAnn_mt2, annotation_name_side = "left",
                                                       gp = gpar(fill = tax_col, 
                                                                 col = NA),
                                                       height = unit(3, "cm"),
                                                       border = F),
                   annotation_name_rot = 90,
                   annotation_name_side = "left"),
                 right_annotation = rowAnnotation(
                   "Set size" = anno_barplot(rowAnn_mt2,
                                             gp = gpar(fill = tax_col,
                                                       col = NA),
                                             # add_numbers = TRUE,
                                             width = unit(3, "cm"),
                                             border = F)),
                 row_names_gp = gpar(fontface ='bold'))

# show the upset plot (balck & white version)
sim_ht2 <- UpSet(m2, set_order = order(set_size(m2), decreasing = T), 
                heatmap_height = unit(length(set_name(m2))*1.5, "cm"),
                heatmap_width = unit(length(comb_name(m2))*1.5, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#A9C0C9", "#D8D7CD"),
                comb_order = order(lapply(comb_name(m2), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m2), 
                                   decreasing = T),
                top_annotation = upset_top_annotation(m2, add_numbers = TRUE,
                                                      numbers_gp = gpar(fontsize = 8),
                                                      numbers_rot = 0,
                                                      annotation_name_rot = 90,
                                                      height = unit(3, "cm")),
                right_annotation = upset_right_annotation(m2, add_numbers = TRUE,
                                                          width = unit(3, "cm")), 
                row_names_gp = gpar(fontface ='bold'))




if (save_or_not) {
  
  rowAnn_order_mt2 <- rowAnn_mt2[order(rowSums(rowAnn_mt2), decreasing = T),]
  
  colAnn_order_mt2 <- colAnn_mt2[comb_name(m2)[order(lapply(comb_name(m2),
                                                         function(x) {
                                                           strsplit(x = x, split = "") %>%
                                                             unlist() %>%
                                                             as.numeric() %>%
                                                             sum()
                                                           }) %>% unlist() * -1,
                                                  comb_size(m2),
                                                  decreasing = T)],] %>% t()

  bICI_hm_pdf2 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Upset"),
                            Prefix = "Heatmap-bICI-Wilcox_MN", Suffix = "pdf")

  pdf(file = bICI_hm_pdf2, width = 10, height = 5)
  draw(tmp_ht2)
  draw(sim_ht2)
  dev.off()
  
  colAnn_csv2 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Upset"),
                           Prefix = "Matrix-ColumnAnnotation-Wilcox_MN", Suffix = "csv")
  write.csv(x = colAnn_order_mt2, file = colAnn_csv2)
  rowAnn_csv2 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Upset"),
                           Prefix = "Matrix-RowAnnotation-Wilcox_MN", Suffix = "csv")
  write.csv(x = rowAnn_order_mt2, file = rowAnn_csv2)

}

```



### Union significant features

```{r upset union}
#################################################################################
# plot selected combination, including the characters appeared in at least 2 set.
# as same as above #
#################################################################################
lt_list3 <- list()
m_list3 <- list()
lt3 <- list()
save_or_not <- F


union_melt_wil_mn_list <- melt_wil_mn_list

for (sel_kingdoms in names(union_melt_wil_mn_list)) {
  tmp_freq <- table(union_melt_wil_mn_list[[sel_kingdoms]][["Feature"]])
  sel_tmp <- names(tmp_freq)[tmp_freq >1]
  union_melt_wil_mn_list[[sel_kingdoms]] <- melt_wil_mn_list[[sel_kingdoms]][melt_wil_mn_list[[sel_kingdoms]]$Feature %in% sel_tmp, ]
}

for (sel_kingdoms in names(union_melt_wil_mn_list)) {
  if (nrow(union_melt_wil_mn_list[[sel_kingdoms]]) == 0) {
    union_melt_wil_mn_list[[sel_kingdoms]]<- NULL
  }
}


for (sel_kingdoms in names(union_melt_wil_mn_list)) {
  lt_list3[[sel_kingdoms]] <- list()
  for (sel_disease in levels(unique(union_melt_wil_mn_list[[sel_kingdoms]]$CancerType))) {
    lt_list3[[sel_kingdoms]][[sel_disease]] <- union_melt_wil_mn_list[[sel_kingdoms]] %>%
      subset(CancerType == sel_disease, select = "Feature") %>% unlist() %>% as.character()
    lt3[[sel_disease]] <- c(lt3[[sel_disease]], lt_list3[[sel_kingdoms]][[sel_disease]])
  }
  m_list3[[sel_kingdoms]] <- make_comb_mat(lt_list3[[sel_kingdoms]])
}
m3 = make_comb_mat(lt3)

rowAnn_mt3 <- data.frame(k__Bacteria = set_size(m_list3$k__Bacteria),
                        k__Eukaryota = set_size(m_list3$k__Eukaryota),
                        k__Archaea = set_size(m_list3$k__Archaea)) %>%
  as.matrix()

colAnn_mt3 <- matrix(data = 0, 
                    nrow = length(comb_name(m3)), 
                    ncol = 4,
                    dimnames = list(comb_name(m3),
                                    names(tax_col)))

for (sub_k in names(m_list3)) {
  for (InSize in names(comb_size(m_list3[[sub_k]]))) {
    colAnn_mt3[InSize, sub_k] <- comb_size(m_list3[[sub_k]])[InSize]
  }
}

label_df3 <- matrix(data = NA, nrow = 0, 
                   ncol = c(length(x = set_name(m3)) + 2),
                   dimnames = list(c(), c( set_name(m3), "Num", "Label"))) %>%
  as.data.frame()
label_colname <- colnames(label_df3)

for (sel_cname in comb_name(m3)) {
  tmp_lab <- extract_comb(m3, sel_cname)
  tmp_df <- matrix(rep(c(strsplit(x = sel_cname, split = "") %>% 
                         unlist() %>% as.numeric(), length(tmp_lab)), 
                       each = length(tmp_lab)),
                   nrow = length(tmp_lab)) %>%
    cbind(., tmp_lab) 
  colnames(tmp_df) <- label_colname
  label_df3 <- rbind(label_df3, tmp_df)
}



label_df3$Group_Num <-  apply(label_df3[, c(1:length(set_name(m3)))], 1, function(x) as.numeric(x) %>% sum)

if (save_or_not) {
  label_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Upset"),
                          Prefix = "Union-Melt-label-Wilcox_MN", Suffix = "csv")
  write.csv(x = label_df3, file = label_csv, row.names = F)
}


tmp_ht3 <- UpSet(m3, set_order = order(set_size(m3), decreasing = T), 
                heatmap_height = unit(length(set_name(m3))*1.5, "cm"),
                heatmap_width = unit(length(comb_name(m3))*1.5, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#A9C0C9", "#D8D7CD"),
                comb_order = order(lapply(comb_name(m3), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m3), 
                                   decreasing = T),
                top_annotation = HeatmapAnnotation(
                  "Intersection\nsize" = anno_barplot(colAnn_mt3, annotation_name_side = "left",
                                                     gp = gpar(fill = tax_col, 
                                                               col = NA),
                                                     height = unit(3, "cm"),
                                                     border = F),
                  annotation_name_rot = 90,
                  annotation_name_side = "left"),
                right_annotation = rowAnnotation(
                  "Set size" = anno_barplot(rowAnn_mt3,
                                            gp = gpar(fill = tax_col,
                                                      col = NA),
                                            # add_numbers = TRUE,
                                            width = unit(3, "cm"),
                                            border = F)),
                row_names_gp = gpar(fontface ='bold'))


sim_ht3 <- UpSet(m3, set_order = order(set_size(m3), decreasing = T), 
                heatmap_height = unit(length(set_name(m3))*1.5, "cm"),
                heatmap_width = unit(length(comb_name(m3))*1.5, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#A9C0C9", "#D8D7CD"),
                comb_order = order(lapply(comb_name(m3), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m3), 
                                   decreasing = T),
                top_annotation = upset_top_annotation(m3, add_numbers = TRUE,
                                                      numbers_gp = gpar(fontsize = 8),
                                                      numbers_rot = 0,
                                                      annotation_name_rot = 90,
                                                      height = unit(3, "cm")),
                right_annotation = upset_right_annotation(m3, add_numbers = TRUE,
                                                          width = unit(3, "cm")), 
                row_names_gp = gpar(fontface ='bold'))




if (save_or_not) {
  
  rowAnn_order_mt3 <- rowAnn_mt3[order(rowSums(rowAnn_mt3), decreasing = T),]
  
  colAnn_order_mt3 <- colAnn_mt3[comb_name(m3)[order(lapply(comb_name(m3),
                                                         function(x) {
                                                           strsplit(x = x, split = "") %>%
                                                             unlist() %>%
                                                             as.numeric() %>%
                                                             sum()
                                                           }) %>% unlist() * -1,
                                                  comb_size(m3),
                                                  decreasing = T)],] %>% t()
  tmp_path <- getwd()
  setwd(paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Upset"))
  bICI_hm_pdf3 <- FileCreate(DirPath = './',
                            Prefix = "Union-Heatmap-bICI-Wilcox_MN", Suffix = "pdf")

  pdf(file = bICI_hm_pdf3, width = 10, height = 5)
  draw(tmp_ht3)
  draw(sim_ht3)
  dev.off()
  
  colAnn_csv3 <- FileCreate(DirPath = './',
                           Prefix = "Union-Matrix-ColumnAnnotation-Wilcox_MN", Suffix = "csv")
  write.csv(x = colAnn_order_mt3, file = colAnn_csv3)
  rowAnn_csv3 <- FileCreate(DirPath = './',
                           Prefix = "Union-Matrix-RowAnnotation-Wilcox_MN", Suffix = "csv")
  write.csv(x = rowAnn_order_mt3, file = rowAnn_csv3)
  setwd(tmp_path)
}

```


## Heatmap (cancer type + Cohort disease)

- Fold change --> `color`

- Candidates --> `row`

- Cohort --> `column`

- p value --> `label`

```{r hm combine 2}
calculate_or_not <- F
save_or_not <- F



# filter the candidates, whom appreaded in at least 2 cancer types or in at least 4 cohorts.

sel_candi_df2 <- c(rownames(wil_bac_sp_cohort$sel_wil_df)[rowSums(wil_bac_sp_cohort$sel_wil_df[, 1:11] <.05) >= 4],
                   rownames(wil_euk_sp_cohort$sel_wil_df)[rowSums(wil_euk_sp_cohort$sel_wil_df[, 1:11] <.05) >= 4],
                   rownames(wil_bac_sp_mn$sel_wil_df)[rowSums(wil_bac_sp_mn$sel_wil_df[, 1:4] < 0.05) >= 2],
                   rownames(wil_euk_sp_mn$sel_wil_df)[rowSums(wil_euk_sp_mn$sel_wil_df[, 1:4] < 0.05) >= 2],
                   rownames(wil_arc_sp_mn$sel_wil_df)[rowSums(wil_arc_sp_mn$sel_wil_df[, 1:4] < 0.05) >= 2]) %>%
  unique() %>% 
  as.data.frame(ncol = 1) %>% 
  set_colnames(c("FullName")) %>%
  set_rownames( .$FullName) %>%
  add_column(Kingdom = lapply(.$FullName, 
                              function(x) {
                                k <- strsplit( x = x, split = "\\|") %>% 
                                  unlist(); 
                                return(k[1])
                                }) %>% unlist(),
             ShortName = lapply(.$FullName,
                                function(x) {
                                  k <- strsplit( x = x, split = "\\|") %>% 
                                    unlist(); 
                                  return(k[length(k)])
                                  }) %>% unlist())

if (save_or_not) {
  sel_cand_csv2 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN"),
  Prefix = "Union-Wilcox_MN-Kingdom-ShortName", Suffix = "csv")
  write.csv(x = sel_candi_df2, file = sel_cand_csv2, row.names = F)
}




if (calculate_or_not) {
  # selected the union candidates in median normalization abundance matrix
  com_mn_bac_df <- bICI_sel_list$k__Bacteria[rownames(bICI_sel_list$k__Bacteria) %in% sel_candi_df2$FullName, ]
  # obtain the meta info
  bac_meta2 <- bICI_ds_list$k__Bacteria$sample_table
  # paste the cohort name and cancer type 
  bac_meta2$CohDis <- paste0(bac_meta2$Cohort_name, "_", bac_meta2$Disease)
  # calculate the Wilcox res by new sub-group CohDis with selected candidates
  wil_bac_cohort2 <- Wil_Matrix(sub_otu = com_mn_bac_df, p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = bac_meta2, comparison_rank = c("R", "NR"),
                              sub_catolog = "CohDis", sub_comparison = "Response")
  # calculate the Wilcox res by previous group cancer type with selected candidates
  wil_bac_disease2 <- Wil_Matrix(sub_otu = com_mn_bac_df, p_cutoff = 0.05, rarefy_or_not = F,
                                sub_meta = bICI_ds_list$k__Bacteria$sample_table,
                                comparison_rank = c("R", "NR"),
                                sub_catolog = "Disease", sub_comparison = "Response")
  # save the rds
  mn_bac_cohort_rds2 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Comb-Wilcox-MN_Idx-Bacteria-Cohort", Suffix = "rds")
  saveRDS(object = wil_bac_cohort2, file = mn_bac_cohort_rds2)
  mn_bac_disease_rds2 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Comb-Wilcox-MN_Idx-Bacteria-disease", Suffix = "rds")
  saveRDS(object = wil_bac_disease2, file = mn_bac_disease_rds2)
  
  
  # as same as above (euk instead of bac)
  com_mn_euk_df <- bICI_sel_list$k__Eukaryota[rownames(bICI_sel_list$k__Eukaryota) %in% sel_candi_df2$FullName, ]
  euk_meta2 <- bICI_ds_list$k__Eukaryota$sample_table
  euk_meta2$CohDis <- paste0(euk_meta2$Cohort_name, "_", euk_meta2$Disease)
  wil_euk_cohort2 <- Wil_Matrix(sub_otu = com_mn_euk_df, p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = euk_meta2, comparison_rank = c("R", "NR"),
                              sub_catolog = "CohDis", sub_comparison = "Response")
  wil_euk_disease2 <- Wil_Matrix(sub_otu = com_mn_euk_df, p_cutoff = 0.05, rarefy_or_not = F,
                                sub_meta = bICI_ds_list$k__Eukaryota$sample_table,
                                comparison_rank = c("R", "NR"),
                                sub_catolog = "Disease", sub_comparison = "Response")
  mn_euk_cohort_rds2 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Comb-Wilcox-MN_Idx-Eukaryota-Cohort", Suffix = "rds")
  saveRDS(object = wil_euk_cohort2, file = mn_euk_cohort_rds2)
  mn_euk_disease_rds2 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Comb-Wilcox-MN_Idx-Eukaryota-disease", Suffix = "rds")
  saveRDS(object = wil_euk_disease2, file = mn_euk_disease_rds2)
  
  
  # as same as above (arc instead of bac)
  com_mn_arc_df <- bICI_sel_list$k__Archaea[rownames(bICI_sel_list$k__Archaea) %in% sel_candi_df2$FullName, ]
  arc_meta2 <- bICI_ds_list$k__Archaea$sample_table
  arc_meta2$CohDis <- paste0(arc_meta2$Cohort_name, "_", arc_meta2$Disease)
  wil_arc_cohort2 <- Wil_Matrix(sub_otu = com_mn_arc_df, p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = tmp_meta, comparison_rank = c("R", "NR"),
                              sub_catolog = "CohDis", sub_comparison = "Response")
  wil_arc_disease2 <- Wil_Matrix(sub_otu = com_mn_arc_df, p_cutoff = 0.05, rarefy_or_not = F,
                                sub_meta = bICI_ds_list$k__Archaea$sample_table,
                                comparison_rank = c("R", "NR"),
                                sub_catolog = "Disease", sub_comparison = "Response")
  mn_arc_cohort_rds2 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Comb-Wilcox-MN_Idx-Archaea-Cohort", Suffix = "rds")
  saveRDS(object = wil_arc_cohort2, file = mn_arc_cohort_rds2)
  mn_arc_disease_rds2 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Comb-Wilcox-MN_Idx-Archaea-disease", Suffix = "rds")
  saveRDS(object = wil_arc_disease2, file = mn_arc_disease_rds2)
  
  
  
}else{
  wil_bac_cohort2 <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-03-15-Comb-Wilcox-MN_Idx-",
                                         "Bacteria", "-Cohort-v1.0.rds"))
  wil_euk_cohort2 <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-03-15-Comb-Wilcox-MN_Idx-",
                                         "Eukaryota", "-Cohort-v1.0.rds"))
  wil_arc_cohort2 <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-03-15-Comb-Wilcox-MN_Idx-",
                                         "Archaea", "-Cohort-v1.0.rds"))
  
  wil_bac_disease2 <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-03-15-Comb-Wilcox-MN_Idx-",
                                         "Bacteria", "-disease-v1.0.rds"))
  wil_euk_disease2 <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-03-15-Comb-Wilcox-MN_Idx-",
                                         "Eukaryota", "-disease-v1.0.rds"))
  wil_arc_disease2 <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-03-15-Comb-Wilcox-MN_Idx-",
                                         "Archaea", "-disease-v1.0.rds"))
}




# merge the Fold change results via Wilcox_res. (CohDis + CancerType + All)
hm_col_df2 <- log2(do.call("rbind",
                          list(cbind(wil_bac_cohort2$FC_df[, -ncol(wil_bac_cohort2$FC_df)], wil_bac_disease2$FC_df),
                               cbind(wil_euk_cohort2$FC_df[, -ncol(wil_euk_cohort2$FC_df)], wil_euk_disease2$FC_df),
                               cbind(wil_arc_cohort2$FC_df[, -ncol(wil_arc_cohort2$FC_df)], wil_arc_disease2$FC_df))))

# hm_col_df2[, "All", drop = F] %>% 
#   rownames_to_column("rname") %>%
#   add_column(Kingdom = lapply(.$rname, 
#                               function(x) {
#                                 k <- strsplit( x = x, split = "\\|") %>% 
#                                   unlist(); 
#                                 return(k[1])
#                                 }) %>% unlist(),
#              ShortName = lapply(.$rname,
#                                 function(x) {
#                                   k <- strsplit( x = x, split = "\\|") %>% 
#                                     unlist(); 
#                                   return(k[length(k)])
#                                   }) %>% unlist()) %>%
#   arrange(Kingdom, All) %>%
#   
#   View()
  
# replace the max/min value 
hm_col_df2[hm_col_df2 >5] <- 5
hm_col_df2[hm_col_df2 < -5] <- -5

# merge the Wilcox results via Wilcox_res. (CohDis + CancerType + All)
tmp_lab_df2 <- do.call("rbind",
                      list(cbind(wil_bac_cohort2$wil_df[, -ncol(wil_bac_cohort2$wil_df)], wil_bac_disease2$wil_df),
                           cbind(wil_euk_cohort2$wil_df[, -ncol(wil_euk_cohort2$wil_df)], wil_euk_disease2$wil_df),
                           cbind(wil_arc_cohort2$wil_df[, -ncol(wil_arc_cohort2$wil_df)], wil_arc_disease2$wil_df)))
# p symbols instead of p value
hm_lab_df2 <- ifelse(tmp_lab_df2 > .05,
                    "",
                    ifelse(tmp_lab_df2 >.01, "*",
                           ifelse(tmp_lab_df2 >.01, "**", "***")))

# define the color
# col_fun = colorRamp2(c(-5, 0, 5), c("#00468bcc", "#ffffff", "#ed0000cc"))
col_fun = colorRamp2(c(-5, 0, 5), c("#ed0000cc", "#ffffff", "#00468bcc"))


# set the top annotations with cancer type
top_anno2 <- HeatmapAnnotation(Disease = colnames(hm_col_df2) %>% gsub("^.+_", "", .), 
                              col = list(Disease = c(cancer_col, All = "black")),
                              gp = gpar(col = 'black'))
# set the right annotation with sub-kingdom
# right_anno2 <- rowAnnotation(Kingdom = sel_candi_df2[rownames(hm_col_df2), ]$Kingdom,
#                             col = list(Kingdom = c(tax_col)), 
#                             gp = gpar(col = 'black'))

############################
bac_sel_otu <- bICI_ds_list$k__Bacteria$taxa_abund$Species[rownames(wil_bac_cohort2$wil_df), ] %>% 
  t() %>% 
  as.data.frame() %>% 
  set_colnames(ShortName(colnames(.))) %>% 
  mutate(Age = bICI_ds_list$k__Bacteria$sample_table$Age,
         Gender = bICI_ds_list$k__Bacteria$sample_table$Gender,
         Continent = bICI_ds_list$k__Bacteria$sample_table$Continent,
         Response = bICI_ds_list$k__Bacteria$sample_table$Response,
         Disease = bICI_ds_list$k__Bacteria$sample_table$Disease)
colnames(bac_sel_otu) <- colnames(bac_sel_otu) %>% 
  gsub("\\[", "", .) %>% 
  gsub("\\-", "_", .) %>% 
  gsub("\\]", "", .) %>% 
  gsub("\\.", "", .)

euk_sel_otu <- bICI_ds_list$k__Eukaryota$taxa_abund$Species[rownames(wil_euk_cohort2$wil_df), ] %>% 
  t() %>% 
  as.data.frame() %>% 
  set_colnames(ShortName(colnames(.))) %>% 
  mutate(Age = bICI_ds_list$k__Eukaryota$sample_table$Age,
         Gender = bICI_ds_list$k__Eukaryota$sample_table$Gender,
         Continent = bICI_ds_list$k__Eukaryota$sample_table$Continent,
         Response = bICI_ds_list$k__Eukaryota$sample_table$Response,
         Disease = bICI_ds_list$k__Eukaryota$sample_table$Disease)
colnames(euk_sel_otu) <- colnames(euk_sel_otu) %>% 
  gsub("\\[", "", .) %>% 
  gsub("\\-", "_", .) %>% 
  gsub("\\]", "", .) %>% 
  gsub("\\.", "", .)


arc_sel_otu <- bICI_ds_list$k__Archaea$taxa_abund$Species[rownames(wil_arc_cohort2$wil_df), ] %>% 
  t() %>% 
  as.data.frame() %>% 
  set_colnames(ShortName(colnames(.))) %>% 
  mutate(Age = bICI_ds_list$k__Archaea$sample_table$Age,
         Gender = bICI_ds_list$k__Archaea$sample_table$Gender,
         Continent = bICI_ds_list$k__Archaea$sample_table$Continent,
         Response = bICI_ds_list$k__Archaea$sample_table$Response,
         Disease = bICI_ds_list$k__Archaea$sample_table$Disease)
colnames(arc_sel_otu) <- colnames(arc_sel_otu) %>% 
  gsub("\\[", "", .) %>% 
  gsub("\\-", "_", .) %>% 
  gsub("\\]", "", .) %>% 
  gsub("\\.", "", .)


aov_df <- matrix(
  data = NA,
  nrow = 52,
  ncol = 4) %>% 
  as.data.frame() %>% 
  set_rownames(
    c(rownames(wil_bac_cohort2$wil_df), 
      rownames(wil_euk_cohort2$wil_df),
      rownames(wil_arc_cohort2$wil_df)) %>% 
      gsub("\\[", "", .) %>% 
      gsub("\\-", "_", .) %>% 
      gsub("\\]", "", .) %>% 
      gsub("\\.", "", .) %>% 
      ShortName()) %>% 
  set_colnames(c("Response", "Age", "Gender", "Continent"))


for (i in rownames(wil_bac_cohort2$wil_df)) {
  i1 <- i %>% 
    gsub("\\[", "", .) %>% 
    gsub("\\-", "_", .) %>% 
    gsub("\\]", "", .) %>% 
    gsub("\\.", "", .) %>% 
    ShortName()
  tmp_res <- summary(aov(as.formula(paste0(i1,  "~ Response + Age + Gender + Continent")), data = bac_sel_otu))[[1]]
  tmp_p <- p.adjust(tmp_res$`Pr(>F)`[1:4], method =  "bonferroni")
  aov_df[i1,] <- tmp_p
}

for (i in rownames(wil_euk_cohort2$wil_df)) {
  i1 <- i %>% 
    gsub("\\[", "", .) %>% 
    gsub("\\-", "_", .) %>% 
    gsub("\\]", "", .) %>% 
    gsub("\\.", "", .) %>% 
    ShortName()
  tmp_res <- summary(aov(as.formula(paste0(i1,  "~ Response + Age + Gender + Continent")), data = euk_sel_otu))[[1]]
  tmp_p <- p.adjust(tmp_res$`Pr(>F)`[1:4], method =  "bonferroni")
  aov_df[i1,] <- tmp_p
}

for (i in rownames(wil_arc_cohort2$wil_df)) {
  i1 <- i %>% 
    gsub("\\[", "", .) %>% 
    gsub("\\-", "_", .) %>% 
    gsub("\\]", "", .) %>% 
    gsub("\\.", "", .) %>% 
    ShortName()
  tmp_res <- summary(aov(as.formula(paste0(i1,  "~ Response + Age + Gender + Continent")), data = arc_sel_otu))[[1]]
  tmp_p <- p.adjust(tmp_res$`Pr(>F)`[1:4], method =  "bonferroni")
  aov_df[i1,] <- tmp_p
}

# 
# tmp_res <- summary(aov(s__Enterocloster_bolteae ~ Response + Age + Gender + Continent, data = bac_sel_otu))[[1]]
# # tmp_res <- summary(aov(s__Faecalibacterium_prausnitzii ~ Response + Age + Gender, data = bac_sel_otu))[[1]]
# 
# tmp_res



(aov_df< 0.05) %>% colSums()

aov_df2 <- aov_df<0.05

aov_df2 <- aov_df2[, -1]

rownames(aov_df2) <- c(rownames(wil_bac_cohort2$wil_df), 
                       rownames(wil_euk_cohort2$wil_df),
                       rownames(wil_arc_cohort2$wil_df))


hm_right2 <- aov_df2[rownames(hm_col_df2), ]

hm_right2[hm_right2==T] <- 1
hm_right2[hm_right2 == F] <- 0

######################################

right_anno2 <- rowAnnotation(Kingdom = sel_candi_df2[rownames(hm_col_df2), ]$Kingdom,
                             # PrePost = hm_right3_,
                             AOV = hm_right2,
                             col = list(Kingdom = c(tax_col),
                                        AOV = c("1" = "black", "0" = "white"),
                                        # PrePost = c("1" = "#C71E1D", "0" = "white", "-1" = "#015069")
                                        ), 
                            gp = gpar(col = 'black'))
# set the left annotation color palette
left_col_fun <- function(x){
  ifelse(x == 0, "white", ifelse(x == 1, "#84bd007f", 
                                 ifelse(x == 2, "#8ca252cc",
                                        ifelse(x == 3, "#637939cc", "#20854ecc"))))
}

# set the left annotation with freq appeared in cohorts
left_anno2 <- rowAnnotation(Num = anno_barplot(rowSums(hm_lab_df2[, 1:11] != ""),
                                              axis_param = list(direction = "reverse"),
                                              gp = gpar(fill = do.call(left_col_fun, 
                                                                       list(rowSums(hm_lab_df2[, 1:11] != "")))),
                                              width = unit(1.5, "cm")))

# set the row Separate
rowSplit <- ifelse(rowSums(hm_col_df2[, 1:11] > 0, na.rm = T) >= 7, "enriched", 
                   ifelse(rowSums(hm_col_df2[, 1:11] < 0, na.rm = T) >= 7, "depleted", "mess"))

# plot the complex heat map
comb_hm2 <- Heatmap(hm_col_df2 %>% as.matrix(),
                    name = "log2(R/NR)",
                    width = unit(ncol(hm_col_df2)/2, "cm"),
                    height = unit(nrow(hm_col_df2)/2, "cm"),
                    # separate the column by manual group, Cohort + Disease + All
                    column_split = factor(c(rep("Cohort", ncol(wil_bac_cohort2$FC_df)-1), 
                                            rep("Disease", ncol(wil_bac_disease2$FC_df)-1),
                                            "All"), levels = c("Cohort", "Disease", "All")),
                    # forced to set the column sub-group rank with factor levels
                    cluster_column_slices = F,
                    cluster_rows = T,
                    # row_order = order(sel_attend %>% sort() %>% names()),
                    # separate the row by manual group, enriched + mess + depleted
                    row_split = factor(rowSplit,
                                       levels = c("enriched", "mess","depleted")),
                    # forced to set the row sub-group rank with factor levels
                    cluster_row_slices = F,
                    # set the symbols p value test in each cell
                    cell_fun = function(j, i, x, y, width, height, fill) {
                      grid.text(hm_lab_df2[i, j], x, y, 
                                rot = 90,
                                gp = gpar(fontsize = 10))},
                    # replace the rowname with matrix's rownames
                    row_labels = sel_candi_df2[rownames(hm_col_df2) ,'ShortName'],
                    col = col_fun, top_annotation = top_anno2,
                    right_annotation = right_anno2, left_annotation = left_anno2,
                    # set the border color and legend parameters
                    border = "grey", heatmap_legend_param = list(direction = "horizontal")) 



if (save_or_not) {
  tmp_path <- getwd()
  setwd(paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Comb-Heatmap"))
  
  comb_hm_pdf2 <- FileCreate(DirPath = './',
                            Prefix = "comb-Heatmap-combine-cohort-disease-foldchange", Suffix = "pdf")
  pdf(file = comb_hm_pdf2, width = ncol(hm_col_df2)/25.4 + 15, height = nrow(hm_col_df2)*5/25.4 + 5)
  draw(comb_hm2, merge_legend = TRUE, 
       heatmap_legend_side = "left",
       annotation_legend_side = "left")
  dev.off()
  setwd(tmp_path)
}


```

## Survival models

### Survival Curve

```{r survival curve}
bICI_meta <- list(k__Bacteria = bICI_ds_list$k__Bacteria$sample_table %>% 
                    dplyr::select(Antibiotics, Age, Gender, Respond_detail, RECIST, 
                                  Sequencing_platform, Cohort_name, Disease, Response, PFS, PFS_status) %>%
                    filter(!is.na(.data[["PFS"]]),
                           !(.data[["PFS"]] %in% c("", "yes", "no"))),
                  k__Eukaryota = bICI_ds_list$k__Eukaryota$sample_table %>% 
                    dplyr::select(Antibiotics, Age, Gender, Respond_detail, RECIST, 
                                  Sequencing_platform, Cohort_name, Disease, Response, PFS, PFS_status) %>%
                    filter(!is.na(.data[["PFS"]]),
                           !(.data[["PFS"]] %in% c("", "yes", "no"))),
                  k__Archaea = bICI_ds_list$k__Archaea$sample_table %>% 
                    dplyr::select(Antibiotics, Age, Gender, Respond_detail, RECIST, 
                                  Sequencing_platform, Cohort_name, Disease, Response, PFS, PFS_status) %>%
                    filter(!is.na(.data[["PFS"]]),
                           !(.data[["PFS"]] %in% c("", "yes", "no"))),
                  k__Viruses = bICI_ds_list$k__Viruses$sample_table %>% 
                    dplyr::select(Antibiotics, Age, Gender, Respond_detail, RECIST, 
                                  Sequencing_platform, Cohort_name, Disease, Response, PFS, PFS_status) %>%
                    filter(!is.na(.data[["PFS"]]),
                           !(.data[["PFS"]] %in% c("", "yes", "no"))))

sign_sel_cand <- hm_lab_df2 %>%
  as.data.frame() %>%
  rownames_to_column(var = 'rname') %>%
  mutate(CohSum = rowSums((.[2:12] != ""))) %>%
  cbind(EnrDep = rowSplit) %>%
  mutate(Sel = ifelse((.$CohSum >= 3 & .$EnrDep != "mess"), "sel", "fil"),
         Kingdom = lapply(.$rname,
                          function(x) {
                            k <- strsplit( x = x, split = "\\|") %>% 
                              unlist(); 
                            return(k[1])
                            }) %>% unlist(),
         ShortName = lapply(.$rname,
                            function(x) {
                              k <- strsplit( x = x, split = "\\|") %>% 
                                unlist(); 
                              return(k[length(k)])
                              }) %>% unlist())



### Eukaryota ###

euk_cand_list <- sign_sel_cand %>% # selected the candidates form Euk and sel
  filter(Kingdom == "k__Eukaryota",
         Sel == "sel") %>% 
  rownames()

euk_surv_df <- cbind(bICI_meta$k__Eukaryota,
                     euk_mn_res2$nm_df %>%
                       dplyr::select(all_of(euk_cand_list)) %>%
                       subset(rownames(.) %in% # selected the samples
                                rownames(bICI_meta$k__Eukaryota))) %>%
  set_colnames(., colnames(.) %>% # replace the column name as the short tax
                 ShortName()) %>%
  as.data.frame() %>%
  mutate(PFS_status = ifelse(PFS_status == "Dead", 1, 0),
         PFS = as.numeric(PFS))




euk.res.cut <- surv_cutpoint(euk_surv_df, # data set
                             time = "PFS", # survival period
                             event = "PFS_status", # survival stat
                             variables = ShortName(euk_cand_list)) # candidate
euk.res.cat <- surv_categorize(euk.res.cut) # catology of cutoff



if (save_or_not) {
  euk_survival_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Survival"),
                                 Prefix = "CombineAll-SurvivalCurve-Euk", Suffix = "pdf")
  pdf(file = euk_survival_pdf, width = 10, height = 6)
  for (sel_cand in ShortName(euk_cand_list)) {
    fit <- survfit(as.formula(paste0('Surv(PFS, PFS_status) ~ ', sel_cand)), data = euk.res.cat)#
    g <- ggsurvplot(fit,
                    palette = c("#ed0000cc", "#00468bcc"),
                    data = euk.res.cat,
                    risk.table = TRUE,
                    pval = T, 
                    title = sel_cand,
                    xscale = "d_y",
                    break.x.by = 365, 
                    xlim=c(0,365*5),
                    tables.y.text = F)
    print(g)
  }
  
  dev.off()
}



                         


### Bacteria ###

bac_cand_list <- sign_sel_cand %>% # selected the candidates form Bac and sel
  filter(Kingdom == "k__Bacteria",
         Sel == "sel") %>% 
  rownames()

bac_surv_df <- cbind(bICI_meta$k__Bacteria,
                     bac_mn_res2$nm_df %>%
                       dplyr::select(all_of(bac_cand_list)) %>%
                       subset(rownames(.) %in% # selected the samples
                                rownames(bICI_meta$k__Bacteria))) %>%
  set_colnames(., colnames(.) %>% # replace the column name as the short tax
                 ShortName()) %>%
  as.data.frame() %>%
  mutate(PFS_status = ifelse(PFS_status == "Dead", 1, 0),
         PFS = as.numeric(PFS))




bac.res.cut <- surv_cutpoint(bac_surv_df, # data set
                             time = "PFS", # survival period
                             event = "PFS_status", # survival stat
                             variables = ShortName(bac_cand_list)) # candidate
bac.res.cat <- surv_categorize(bac.res.cut) # catology of cutoff

if (calculate_or_not) {
  bac_survival_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Survival"),
                                 Prefix = "CombineAll-SurvivalCurve-Bac", Suffix = "pdf")
  pdf(file = bac_survival_pdf, width = 10, height = 6)
  for (sel_cand in ShortName(bac_cand_list)) {
    fit <- survfit(as.formula(paste0('Surv(PFS, PFS_status) ~ ', sel_cand)), data = bac.res.cat)#
    g <- ggsurvplot(fit,
                    palette = c("#ed0000cc", "#00468bcc"),
                    data = bac.res.cat,
                    risk.table = TRUE,
                    pval = T, 
                    title = sel_cand,
                    xscale = "d_y",
                    break.x.by = 365, 
                    xlim=c(0,365*5),
                    tables.y.text = F)
    print(g)
  }
  
  dev.off()
}




```

### multivariate analysis

```{r forest plot}
forest_list <- list()

# Eukaryota
for (sel_cand in ShortName(euk_cand_list)) {
  euk_surv_df %>%
    select(c("PFS", "PFS_status", sel_cand,
             "Age", "Gender",
             "Disease")) %>%
    mutate(cand = euk.res.cat[[sel_cand]]) %>%
    select( -`sel_cand`) %>% 
    na.omit() %>%
    analyse_multivariate(vars(PFS, PFS_status),
                         covariates = vars(Age, Gender,
                                           cand, Disease)) ->
    forest_list[["Euk"]][[sel_cand]]
}

# Bacteria
for (sel_cand in ShortName(bac_cand_list)) {
  bac_surv_df %>%
    select(c("PFS", "PFS_status", sel_cand, 
             "Age", "Gender", 
             "Disease")) %>%
    mutate(cand = bac.res.cat[[sel_cand]]) %>%
    select( -`sel_cand`) %>% 
    na.omit() %>%
    analyse_multivariate(vars(PFS, PFS_status),
                         covariates = vars(Age, Gender,
                                           cand, Disease)) ->
    forest_list[["Bac"]][[sel_cand]]
}

survival_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/multivariate"),
                           Prefix = "MultVar-Analysis", Suffix = "pdf")
pdf(file = survival_pdf, width = 10, height = 3)

for (sel_k in names(forest_list)) {
  for (sel_cand in names(forest_list[[sel_k]])) {
    my_forest <- forest_plot(forest_list[[sel_k]][[sel_cand]], 
                             orderer = ~ order(HR),
                             labels_displayed = c("factor", "n"),
                             relative_widths = c(2, 4, 2), 
                             title = sel_cand)
    plot(my_forest)
  }
}
dev.off()


```



### Feature Validation

```{r sel features validation}
# paired ICI samples

pICI_sel_list <- list()
## bacteria
pICI_sel_list$k__Bacteria <- t(bac_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Bacteria$sample_table)]
# eukaryota
pICI_sel_list$k__Eukaryota <- t(euk_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Eukaryota$sample_table)]
# archaea
pICI_sel_list$k__Archaea <- t(arc_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Archaea$sample_table)]
# viruses
pICI_sel_list$k__Viruses <- t(vir_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Viruses$sample_table)]



pICI_comparisons <- list(c("NR-Before_ICI", "NR-ICI_treat"),
                         c("NR-Before_ICI", "R-Before_ICI"),
                         c("NR-Before_ICI", "R-ICI_treat"),
                         c("R-Before_ICI", "R-ICI_treat"))



pcomp_col <- c(`NR-Before_ICI` = "tomato",
               `NR-ICI_treat` = "red",
               `R-Before_ICI` = "steelblue1",
               `R-ICI_treat` = "royalblue")

# paired ICI
## EUK
pICI_euk_df <- cbind(pICI_ds_list$k__Eukaryota$sample_table,
                     euk_mn_res2$nm_df[rownames(pICI_ds_list$k__Eukaryota$sample_table),] %>% 
                       select(ends_with(c("Nemania_serpens", 
                                          "Fusarium_transvaalense",
                                          "Hyphopichia_pseudoburtonii")))) 
colnames(pICI_euk_df) <- lapply(colnames(pICI_euk_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
colnames(pICI_euk_df)[(ncol(pICI_euk_df)-2):ncol(pICI_euk_df)] <- c("NS", "FT", "HP")
pICI_euk_df$Period2 <- factor(paste0(pICI_euk_df$Response, "-", 
                                     ifelse(pICI_euk_df$Period == "Before_ICI", 
                                            "Before_ICI", "ICI_treat")), 
                              levels = c("NR-Before_ICI", 
                                         "NR-ICI_treat",
                                         "R-Before_ICI", 
                                         "R-ICI_treat"))

g1_NS <- ggplot(pICI_euk_df, aes(x = Period2, y = NS)) +
  theme_Publication() +
  geom_violin(alpha = 0, aes(col = Period2)) +
  geom_boxplot(width = 0.1, outlier.colour = NA) +
  # scale_y_log10() +
  geom_jitter(aes(col = Period2), alpha = 0.5) +
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)+
  theme(legend.position = "none")



violin_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Boxplot"),
                               Prefix = "CombineAll-MultVar", Suffix = "pdf")
pdf(file = violin_pdf, width = 7, height = 4.5)

plot(g1_NS)
dev.off()
```



### Feature Validation (revision)

```{r sel features validation revision}
# paired ICI samples

pICI_sel_list <- list()
## bacteria
pICI_sel_list$k__Bacteria <- t(bac_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Bacteria$sample_table)]
# eukaryota
pICI_sel_list$k__Eukaryota <- t(euk_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Eukaryota$sample_table)]
# archaea
pICI_sel_list$k__Archaea <- t(arc_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Archaea$sample_table)]
# viruses
pICI_sel_list$k__Viruses <- t(vir_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Viruses$sample_table)]



pICI_comparisons <- list(c("NR-Before_ICI", "NR-ICI_treat"),
                         c("NR-Before_ICI", "R-Before_ICI"),
                         c("NR-Before_ICI", "R-ICI_treat"),
                         c("R-Before_ICI", "R-ICI_treat"))



pcomp_col <- c(`NR-Before_ICI` = "tomato",
               `NR-ICI_treat` = "red",
               `R-Before_ICI` = "steelblue1",
               `R-ICI_treat` = "royalblue")

# paired ICI
## EUK
euk_tmp_list <- rownames(sel_candi_df2)[sel_candi_df2$Kingdom == "k__Eukaryota"]

pICI_euk_df <- cbind(pICI_ds_list$k__Eukaryota$sample_table,
                     euk_mn_res2$nm_df[rownames(pICI_ds_list$k__Eukaryota$sample_table),] %>% 
                       select(ends_with(c(euk_tmp_list)))) 
colnames(pICI_euk_df) <- lapply(colnames(pICI_euk_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
pICI_euk_df$Period2 <- factor(paste0(pICI_euk_df$Response, "-", 
                                     ifelse(pICI_euk_df$Period == "Before_ICI", 
                                            "Before_ICI", "ICI_treat")), 
                              levels = c("NR-Before_ICI", 
                                         "NR-ICI_treat",
                                         "R-Before_ICI", 
                                         "R-ICI_treat"))

violin_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Boxplot"),
                               Prefix = "CombineAll-MultVar-Euk", Suffix = "pdf")
pdf(file = violin_pdf, width = 7, height = 4.5)

for (i in 27:(ncol(pICI_euk_df)-1)) {
  tmp_violin_df <- pICI_euk_df[, c(ncol(pICI_euk_df), i)]
  tmp_name <- colnames(tmp_violin_df)[2]
  colnames(tmp_violin_df)[2] <- "cand"
  tmp_violin_df <- tmp_violin_df %>%
   mutate_at(2, funs(c(scale(.))))
  tmp_vio_plot <- ggplot(tmp_violin_df, aes(x = Period2, y = cand)) +
    theme_Publication() +
    geom_violin(alpha = 0, aes(col = Period2)) +
    geom_boxplot(width = 0.1, outlier.colour = NA) +
    # scale_y_log10() +
    geom_jitter(aes(col = Period2), alpha = 0.5) +
    stat_compare_means(comparisons = pICI_comparisons) +
    scale_color_manual(values = pcomp_col) +
    theme(legend.position = "none") + 
    ylab(tmp_name)
  plot(tmp_vio_plot)
}

dev.off()


violin_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Boxplot"),
                               Prefix = "CombineAll-MultVar-sel-Euk", Suffix = "pdf")
pdf(file = violin_pdf, width = 3, height = 4.5)

for (i in c(33, 34)) {
  tmp_violin_df <- pICI_euk_df[, c(ncol(pICI_euk_df), i)]
  tmp_name <- colnames(tmp_violin_df)[2]
  colnames(tmp_violin_df)[2] <- "cand"
  tmp_violin_df <- tmp_violin_df %>%
   mutate_at(2, funs(c(scale(.))))
  tmp_vio_plot <- ggplot(tmp_violin_df, aes(x = Period2, y = cand)) +
    theme_Publication() +
    # geom_violin(alpha = 0, aes(col = Period2)) +
    geom_boxplot(width = 0.1, outlier.colour = NA) +
    # scale_y_log10() +
    geom_jitter(aes(col = Period2), alpha = 0.5) +
    stat_compare_means(comparisons = pICI_comparisons) +
    scale_color_manual(values = pcomp_col) +
    theme(legend.position = "none") + 
    ylab(tmp_name)
  plot(tmp_vio_plot)
}

dev.off()

# paired ICI
# BAC
bac_tmp_list <- rownames(sel_candi_df2)[sel_candi_df2$Kingdom == "k__Bacteria"]

pICI_bac_df <- cbind(pICI_ds_list$k__Bacteria$sample_table,
                     bac_mn_res2$nm_df[rownames(pICI_ds_list$k__Bacteria$sample_table),] %>% 
                       select(ends_with(c(bac_tmp_list)))) 
colnames(pICI_bac_df) <- lapply(colnames(pICI_bac_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
colnames(pICI_bac_df) %<>% gsub("\\[", "", .) %>% 
  gsub("\\]", "", .) %>% 
  gsub("\\.", "", .)

pICI_bac_df$Period2 <- factor(paste0(pICI_bac_df$Response, "-", 
                                     ifelse(pICI_bac_df$Period == "Before_ICI", 
                                            "Before_ICI", "ICI_treat")), 
                              levels = c("NR-Before_ICI", 
                                         "NR-ICI_treat",
                                         "R-Before_ICI", 
                                         "R-ICI_treat"))

violin_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Boxplot"),
                               Prefix = "CombineAll-MultVar-Bac", Suffix = "pdf")
pdf(file = violin_pdf, width = 7, height = 4.5)

for (i in 27:(ncol(pICI_bac_df)-1)) {
  tmp_violin_df <- pICI_bac_df[, c(ncol(pICI_bac_df), i)]
  tmp_name <- colnames(tmp_violin_df)[2]
  colnames(tmp_violin_df)[2] <- "cand"
  tmp_vio_plot <- ggplot(tmp_violin_df, aes(x = Period2, y = cand)) +
    theme_Publication() +
    geom_violin(alpha = 0, aes(col = Period2)) +
    geom_boxplot(width = 0.1, outlier.colour = NA) +
    # scale_y_log10() +
    geom_jitter(aes(col = Period2), alpha = 0.5) +
    stat_compare_means(comparisons = pICI_comparisons) +
    scale_color_manual(values = pcomp_col) +
    theme(legend.position = "none") + 
    ylab(tmp_name)
  plot(tmp_vio_plot)
}

dev.off()










```
### Feautre Validation2 (revision)


```{r sel features validation revision 2}
# paired ICI samples

pICI_sel_list <- list()
## bacteria
pICI_sel_list$k__Bacteria <- t(bac_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Bacteria$sample_table)]
# eukaryota
pICI_sel_list$k__Eukaryota <- t(euk_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Eukaryota$sample_table)]
# archaea
pICI_sel_list$k__Archaea <- t(arc_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Archaea$sample_table)]
# viruses
pICI_sel_list$k__Viruses <- t(vir_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Viruses$sample_table)]



pICI_comparisons <- list(c("NR-Before_ICI", "NR-ICI_treat"),
                         c("R-Before_ICI", "R-ICI_treat"))



pcomp_col <- c(`NR-Before_ICI` = "tomato",
               `NR-ICI_treat` = "red",
               `R-Before_ICI` = "steelblue1",
               `R-ICI_treat` = "royalblue")

# paired ICI
## EUK
euk_tmp_list <- rownames(sel_candi_df2)[sel_candi_df2$Kingdom == "k__Eukaryota"]

pICI_euk_df <- cbind(pICI_ds_list$k__Eukaryota$sample_table,
                     euk_mn_res2$nm_df[rownames(pICI_ds_list$k__Eukaryota$sample_table),] %>% 
                       select(ends_with(c(euk_tmp_list)))) 
colnames(pICI_euk_df) <- lapply(colnames(pICI_euk_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
pICI_euk_df$Period2 <- factor(paste0(pICI_euk_df$Response, "-", 
                                     ifelse(pICI_euk_df$Period == "Before_ICI", 
                                            "Before_ICI", "ICI_treat")), 
                              levels = c("NR-Before_ICI", 
                                         "NR-ICI_treat",
                                         "R-Before_ICI", 
                                         "R-ICI_treat"))

violin_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Boxplot"),
                               Prefix = "CombineAll-MultVar-Euk", Suffix = "pdf")
pdf(file = violin_pdf, width = 7, height = 4.5)

for (i in 27:(ncol(pICI_euk_df)-1)) {
  tmp_violin_df <- pICI_euk_df[, c(ncol(pICI_euk_df), i)]
  tmp_name <- colnames(tmp_violin_df)[2]
  colnames(tmp_violin_df)[2] <- "cand"
  tmp_violin_df <- tmp_violin_df %>%
   mutate_at(2, funs(c(scale(.))))
  tmp_vio_plot <- ggplot(tmp_violin_df, aes(x = Period2, y = cand)) +
    theme_Publication() +
    geom_violin(alpha = 0, aes(col = Period2)) +
    geom_boxplot(width = 0.1, outlier.colour = NA) +
    # scale_y_log10() +
    geom_jitter(aes(col = Period2), alpha = 0.5) +
    stat_compare_means(comparisons = pICI_comparisons) +
    scale_color_manual(values = pcomp_col) +
    theme(legend.position = "none") + 
    ylab(tmp_name)
  plot(tmp_vio_plot)
}

dev.off()


violin_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Boxplot"),
                               Prefix = "CombineAll-MultVar-sel-Euk", Suffix = "pdf")
pdf(file = violin_pdf, width = 3, height = 4.5)

for (i in c(33, 34)) {
  tmp_violin_df <- pICI_euk_df[, c(ncol(pICI_euk_df), i)]
  tmp_name <- colnames(tmp_violin_df)[2]
  colnames(tmp_violin_df)[2] <- "cand"
  tmp_violin_df <- tmp_violin_df %>%
   mutate_at(2, funs(c(scale(.))))
  tmp_vio_plot <- ggplot(tmp_violin_df, aes(x = Period2, y = cand)) +
    theme_Publication() +
    # geom_violin(alpha = 0, aes(col = Period2)) +
    geom_boxplot(width = 0.1, outlier.colour = NA) +
    # scale_y_log10() +
    geom_jitter(aes(col = Period2), alpha = 0.5) +
    stat_compare_means(comparisons = pICI_comparisons) +
    scale_color_manual(values = pcomp_col) +
    theme(legend.position = "none") + 
    ylab(tmp_name)
  plot(tmp_vio_plot)
}

dev.off()

# paired ICI
# BAC
bac_tmp_list <- rownames(sel_candi_df2)[sel_candi_df2$Kingdom == "k__Bacteria"]

pICI_bac_df <- cbind(pICI_ds_list$k__Bacteria$sample_table,
                     bac_mn_res2$nm_df[rownames(pICI_ds_list$k__Bacteria$sample_table),] %>% 
                       select(ends_with(c(bac_tmp_list)))) 
colnames(pICI_bac_df) <- lapply(colnames(pICI_bac_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
colnames(pICI_bac_df) %<>% gsub("\\[", "", .) %>% 
  gsub("\\]", "", .) %>% 
  gsub("\\.", "", .)

pICI_bac_df$Period2 <- factor(paste0(pICI_bac_df$Response, "-", 
                                     ifelse(pICI_bac_df$Period == "Before_ICI", 
                                            "Before_ICI", "ICI_treat")), 
                              levels = c("NR-Before_ICI", 
                                         "NR-ICI_treat",
                                         "R-Before_ICI", 
                                         "R-ICI_treat"))

violin_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Boxplot"),
                               Prefix = "CombineAll-MultVar-Bac", Suffix = "pdf")
pdf(file = violin_pdf, width = 7, height = 4.5)

for (i in 27:(ncol(pICI_bac_df)-1)) {
  tmp_violin_df <- pICI_bac_df[, c(ncol(pICI_bac_df), i)]
  tmp_name <- colnames(tmp_violin_df)[2]
  colnames(tmp_violin_df)[2] <- "cand"
  tmp_vio_plot <- ggplot(tmp_violin_df, aes(x = Period2, y = cand)) +
    theme_Publication() +
    geom_violin(alpha = 0, aes(col = Period2)) +
    geom_boxplot(width = 0.1, outlier.colour = NA) +
    # scale_y_log10() +
    geom_jitter(aes(col = Period2), alpha = 0.5) +
    stat_compare_means(comparisons = pICI_comparisons) +
    scale_color_manual(values = pcomp_col) +
    theme(legend.position = "none") + 
    ylab(tmp_name)
  plot(tmp_vio_plot)
}

dev.off()










```


## Feature Selection (cancer type)

- selected the candidate in sub-cancer types MM/NSCLC

### UpSet Plot (cancer type)


#### All

```{r upset all}

############### MM ############### 

lt_list4 <- list()
m_list4 <- list()
lt4 <- list()
save_or_not <- T


for (sel_kingdoms in names(melt_wil_cohort_list)) {
  lt_list4[[sel_kingdoms]] <- list()
  for (sel_disease in levels(unique(melt_wil_cohort_list[[sel_kingdoms]]$CohDis)) %>% grep(., pattern = "_MM$", value = T)) {
    lt_list4[[sel_kingdoms]][[sel_disease]] <- melt_wil_cohort_list[[sel_kingdoms]] %>%
      subset(CohDis == sel_disease, select = "Feature") %>% unlist() %>% as.character()
    lt4[[sel_disease]] <- c(lt4[[sel_disease]], lt_list4[[sel_kingdoms]][[sel_disease]])
  }
  m_list4[[sel_kingdoms]] <- make_comb_mat(lt_list4[[sel_kingdoms]])
}
m4 = make_comb_mat(lt4)

rowAnn_mt4 <- data.frame(k__Bacteria = set_size(m_list4$k__Bacteria),
                         k__Eukaryota = set_size(m_list4$k__Eukaryota),
                         k__Archaea = set_size(m_list4$k__Archaea),
                         k__Viruses = set_size(m_list4$k__Viruses)) %>%
  as.matrix()

colAnn_mt4 <- matrix(data = 0, 
                    nrow = length(comb_name(m4)), 
                    ncol = 4,
                    dimnames = list(comb_name(m4),
                                    names(tax_col)))

for (sub_k in names(m_list4)) {
  for (InSize in names(comb_size(m_list4[[sub_k]]))) {
    colAnn_mt4[InSize, sub_k] <- comb_size(m_list4[[sub_k]])[InSize]
  }
}

label_df4 <- matrix(data = NA, nrow = 0, 
                   ncol = c(length(x = set_name(m4)) + 2),
                   dimnames = list(c(), c( set_name(m4), "Num", "Label"))) %>%
  as.data.frame()
label_colname <- colnames(label_df4)

for (sel_cname in comb_name(m4)) {
  tmp_lab <- extract_comb(m4, sel_cname)
  tmp_df <- matrix(rep(c(strsplit(x = sel_cname, split = "") %>% 
                         unlist() %>% as.numeric(), length(tmp_lab)), 
                       each = length(tmp_lab)),
                   nrow = length(tmp_lab)) %>%
    cbind(., tmp_lab) 
  colnames(tmp_df) <- label_colname
  label_df4 <- rbind(label_df4, tmp_df)
}



label_df4$Group_Num <-  apply(label_df4[, c(1:length(set_name(m4)))], 1, function(x) as.numeric(x) %>% sum)

if (save_or_not) {
  label_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                          Prefix = "Melt-label-Wilcox_MN-MM-cohort", Suffix = "csv")
  write.csv(x = label_df4, file = label_csv, row.names = F)
}


tmp_ht4 <- UpSet(m4, set_order = order(set_size(m4), decreasing = T), 
                heatmap_height = unit(length(set_name(m4))*1.5, "cm"),
                heatmap_width = unit(length(comb_name(m4))*1.5, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#A9C0C9", "#D8D7CD"),
                comb_order = order(lapply(comb_name(m4), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m4), 
                                   decreasing = T),
                top_annotation = HeatmapAnnotation(
                  "Intersection\nsize" = anno_barplot(colAnn_mt4, annotation_name_side = "left",
                                                     gp = gpar(fill = tax_col, 
                                                               col = NA),
                                                     height = unit(3, "cm"),
                                                     border = F),
                  annotation_name_rot = 90,
                  annotation_name_side = "left"),
                right_annotation = rowAnnotation(
                  "Set size" = anno_barplot(rowAnn_mt4,
                                            gp = gpar(fill = tax_col,
                                                      col = NA),
                                            # add_numbers = TRUE,
                                            width = unit(3, "cm"),
                                            border = F)),
                row_names_gp = gpar(fontface ='bold'))


sim_ht4 <- UpSet(m4, set_order = order(set_size(m4), decreasing = T), 
                heatmap_height = unit(length(set_name(m4))*1.5, "cm"),
                heatmap_width = unit(length(comb_name(m4))*1.5, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#A9C0C9", "#D8D7CD"),
                comb_order = order(lapply(comb_name(m4), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m4), 
                                   decreasing = T),
                top_annotation = upset_top_annotation(m4, add_numbers = TRUE,
                                                      numbers_gp = gpar(fontsize = 8),
                                                      numbers_rot = 0,
                                                      annotation_name_rot = 90,
                                                      height = unit(3, "cm")),
                right_annotation = upset_right_annotation(m4, add_numbers = TRUE,
                                                          width = unit(3, "cm")), 
                row_names_gp = gpar(fontface ='bold'))


if (save_or_not) {
  
  rowAnn_order_mt4 <- rowAnn_mt4[order(rowSums(rowAnn_mt4), decreasing = T),]
  
  colAnn_order_mt4 <- colAnn_mt4[comb_name(m4)[order(lapply(comb_name(m4),
                                                         function(x) {
                                                           strsplit(x = x, split = "") %>%
                                                             unlist() %>%
                                                             as.numeric() %>%
                                                             sum()
                                                           }) %>% unlist() * -1,
                                                  comb_size(m4),
                                                  decreasing = T)],] %>% t()

  bICI_hm_pdf4 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                            Prefix = "Heatmap-bICI-Wilcox_MN-MM", Suffix = "pdf")
    
  pdf(file = bICI_hm_pdf4, width = 30, height = 5)
  draw(tmp_ht4)
  draw(sim_ht4)
  dev.off()
  
  colAnn_csv4 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                           Prefix = "Matrix-ColumnAnnotation-Wilcox_MN-MM", Suffix = "csv")
  write.csv(x = colAnn_order_mt4, file = colAnn_csv4)
  rowAnn_csv4 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                           Prefix = "Matrix-RowAnnotation-Wilcox_MN-MM", Suffix = "csv")
  write.csv(x = rowAnn_order_mt4, file = rowAnn_csv4)

}

############### NSCLC ############### 
lt_list5 <- list()
m_list5 <- list()
lt5 <- list()
save_or_not <- T


for (sel_kingdoms in names(melt_wil_cohort_list)) {
  lt_list5[[sel_kingdoms]] <- list()
  for (sel_disease in levels(unique(melt_wil_cohort_list[[sel_kingdoms]]$CohDis)) %>% grep(., pattern = "_NSCLC$", value = T)) {
    lt_list5[[sel_kingdoms]][[sel_disease]] <- melt_wil_cohort_list[[sel_kingdoms]] %>%
      subset(CohDis == sel_disease, select = "Feature") %>% unlist() %>% as.character()
    lt5[[sel_disease]] <- c(lt5[[sel_disease]], lt_list5[[sel_kingdoms]][[sel_disease]])
  }
  m_list5[[sel_kingdoms]] <- make_comb_mat(lt_list5[[sel_kingdoms]])
}
m5 = make_comb_mat(lt5)

rowAnn_mt5 <- data.frame(k__Bacteria = set_size(m_list5$k__Bacteria),
                        k__Eukaryota = set_size(m_list5$k__Eukaryota),
                        k__Archaea = set_size(m_list5$k__Archaea),
                        k__Viruses = set_size(m_list5$k__Viruses)) %>%
  as.matrix()

colAnn_mt5 <- matrix(data = 0, 
                    nrow = length(comb_name(m5)), 
                    ncol = 4,
                    dimnames = list(comb_name(m5),
                                    names(tax_col)))

for (sub_k in names(m_list5)) {
  for (InSize in names(comb_size(m_list5[[sub_k]]))) {
    colAnn_mt5[InSize, sub_k] <- comb_size(m_list5[[sub_k]])[InSize]
  }
}

label_df5 <- matrix(data = NA, nrow = 0, 
                   ncol = c(length(x = set_name(m5)) + 2),
                   dimnames = list(c(), c( set_name(m5), "Num", "Label"))) %>%
  as.data.frame()
label_colname <- colnames(label_df5)

for (sel_cname in comb_name(m5)) {
  tmp_lab <- extract_comb(m5, sel_cname)
  tmp_df <- matrix(rep(c(strsplit(x = sel_cname, split = "") %>% 
                         unlist() %>% as.numeric(), length(tmp_lab)), 
                       each = length(tmp_lab)),
                   nrow = length(tmp_lab)) %>%
    cbind(., tmp_lab) 
  colnames(tmp_df) <- label_colname
  label_df5 <- rbind(label_df5, tmp_df)
}



label_df5$Group_Num <-  apply(label_df5[, c(1:length(set_name(m5)))], 1, function(x) as.numeric(x) %>% sum)

if (save_or_not) {
  label_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                          Prefix = "Melt-label-Wilcox_MN-NSCLC-cohort", Suffix = "csv")
  write.csv(x = label_df5, file = label_csv, row.names = F)
}


tmp_ht5 <- UpSet(m5, set_order = order(set_size(m5), decreasing = T), 
                heatmap_height = unit(length(set_name(m5))*1.5 + 3, "cm"),
                heatmap_width = unit(length(comb_name(m5))*3 + 3, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#A9C0C9", "#D8D7CD"),
                comb_order = order(lapply(comb_name(m5), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m5), 
                                   decreasing = T),
                top_annotation = HeatmapAnnotation(
                  "Intersection\nsize" = anno_barplot(colAnn_mt5, annotation_name_side = "left",
                                                     gp = gpar(fill = tax_col, 
                                                               col = NA),
                                                     height = unit(3, "cm"),
                                                     border = F),
                  annotation_name_rot = 90,
                  annotation_name_side = "left"),
                right_annotation = rowAnnotation(
                  "Set size" = anno_barplot(rowAnn_mt5,
                                            gp = gpar(fill = tax_col,
                                                      col = NA),
                                            # add_numbers = TRUE,
                                            width = unit(3, "cm"),
                                            border = F)),
                row_names_gp = gpar(fontface ='bold'))


sim_ht5 <- UpSet(m5, set_order = order(set_size(m5), decreasing = T), 
                heatmap_height = unit(length(set_name(m5))*1.5 + 3, "cm"),
                heatmap_width = unit(length(comb_name(m5))*3 + 3, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#A9C0C9", "#D8D7CD"),
                comb_order = order(lapply(comb_name(m5), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m5), 
                                   decreasing = T),
                top_annotation = upset_top_annotation(m5, add_numbers = TRUE,
                                                      numbers_gp = gpar(fontsize = 8),
                                                      numbers_rot = 0,
                                                      annotation_name_rot = 90,
                                                      height = unit(3, "cm")),
                right_annotation = upset_right_annotation(m5, add_numbers = TRUE,
                                                          width = unit(3, "cm")), 
                row_names_gp = gpar(fontface ='bold'))


if (save_or_not) {
  
  rowAnn_order_mt5 <- rowAnn_mt5[order(rowSums(rowAnn_mt5), decreasing = T),]
  
  colAnn_order_mt5 <- colAnn_mt5[comb_name(m5)[order(lapply(comb_name(m5),
                                                         function(x) {
                                                           strsplit(x = x, split = "") %>%
                                                             unlist() %>%
                                                             as.numeric() %>%
                                                             sum()
                                                           }) %>% unlist() * -1,
                                                  comb_size(m5),
                                                  decreasing = T)],] %>% t()

  bICI_hm_pdf5 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                            Prefix = "Heatmap-bICI-Wilcox_MN-NSCLC", Suffix = "pdf")
    
  pdf(file = bICI_hm_pdf5, width = 10, height = 5)
  draw(tmp_ht5)
  draw(sim_ht5)
  dev.off()
  
  colAnn_csv5 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                           Prefix = "Matrix-ColAnnotation-Wilcox_MN-NSCLC", Suffix = "csv")
  write.csv(x = colAnn_order_mt5, file = colAnn_csv5)
  rowAnn_csv5 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                           Prefix = "Matrix-RowAnnotation-Wilcox_MN-NSCLC", Suffix = "csv")
  write.csv(x = rowAnn_order_mt5, file = rowAnn_csv5)

}



```




#### union

```{r upset union}
calculate_or_not <- T

union_melt_wil_cohort_list <- melt_wil_cohort_list



############### MM ############### 

mm_union_list <- union_melt_wil_cohort_list

mm_union_list <- lapply(union_melt_wil_cohort_list, function(x){
  x1 <- x[grep(x$CohDis, pattern = "_MM$"),]
  x1[x1$Feature %in% (table(x1$Feature)[table(x1$Feature)>1] %>% names()), ]
})


lt_list6 <- list()
m_list6 <- list()
lt6 <- list()
save_or_not <- T


for (sel_kingdoms in names(mm_union_list)[-4]) {
  lt_list6[[sel_kingdoms]] <- list()
  for (sel_disease in levels(unique(mm_union_list[[sel_kingdoms]]$CohDis)) %>% grep(., pattern = "_MM$", value = T)) {
    lt_list6[[sel_kingdoms]][[sel_disease]] <- mm_union_list[[sel_kingdoms]] %>%
      subset(CohDis == sel_disease, select = "Feature") %>% unlist() %>% as.character()
    lt6[[sel_disease]] <- c(lt6[[sel_disease]], lt_list6[[sel_kingdoms]][[sel_disease]])
  }
  m_list6[[sel_kingdoms]] <- make_comb_mat(lt_list6[[sel_kingdoms]])
}
m6 = make_comb_mat(lt6)

rowAnn_mt6 <- data.frame(k__Bacteria = set_size(m_list6$k__Bacteria),
                         k__Eukaryota = set_size(m_list6$k__Eukaryota),
                         k__Archaea = set_size(m_list6$k__Archaea)) %>%
  as.matrix()

colAnn_mt6 <- matrix(data = 0, 
                    nrow = length(comb_name(m6)), 
                    ncol = 4,
                    dimnames = list(comb_name(m6),
                                    names(tax_col)))

for (sub_k in names(m_list6)) {
  for (InSize in names(comb_size(m_list6[[sub_k]]))) {
    colAnn_mt6[InSize, sub_k] <- comb_size(m_list6[[sub_k]])[InSize]
  }
}

label_df6 <- matrix(data = NA, nrow = 0, 
                   ncol = c(length(x = set_name(m6)) + 2),
                   dimnames = list(c(), c( set_name(m6), "Num", "Label"))) %>%
  as.data.frame()
label_colname <- colnames(label_df6)

for (sel_cname in comb_name(m6)) {
  tmp_lab <- extract_comb(m6, sel_cname)
  tmp_df <- matrix(rep(c(strsplit(x = sel_cname, split = "") %>% 
                         unlist() %>% as.numeric(), length(tmp_lab)), 
                       each = length(tmp_lab)),
                   nrow = length(tmp_lab)) %>%
    cbind(., tmp_lab) 
  colnames(tmp_df) <- label_colname
  label_df6 <- rbind(label_df6, tmp_df)
}



label_df6$Group_Num <-  apply(label_df6[, c(1:length(set_name(m6)))], 1, function(x) as.numeric(x) %>% sum)

if (save_or_not) {
  label_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                          Prefix = "Union-Melt-label-Wilcox_MN-MM-cohort", Suffix = "csv")
  write.csv(x = label_df6, file = label_csv, row.names = F)
}


tmp_ht6 <- UpSet(m6, set_order = order(set_size(m6), decreasing = T), 
                heatmap_height = unit(length(set_name(m6))*1.5, "cm"),
                heatmap_width = unit(length(comb_name(m6))*1.5, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#A9C0C9", "#D8D7CD"),
                comb_order = order(lapply(comb_name(m6), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m6), 
                                   decreasing = T),
                top_annotation = HeatmapAnnotation(
                  "Intersection\nsize" = anno_barplot(colAnn_mt6, annotation_name_side = "left",
                                                     gp = gpar(fill = tax_col, 
                                                               col = NA),
                                                     height = unit(3, "cm"),
                                                     border = F),
                  annotation_name_rot = 90,
                  annotation_name_side = "left"),
                right_annotation = rowAnnotation(
                  "Set size" = anno_barplot(rowAnn_mt6,
                                            gp = gpar(fill = tax_col,
                                                      col = NA),
                                            # add_numbers = TRUE,
                                            width = unit(3, "cm"),
                                            border = F)),
                row_names_gp = gpar(fontface ='bold'))


sim_ht6 <- UpSet(m6, set_order = order(set_size(m6), decreasing = T), 
                heatmap_height = unit(length(set_name(m6))*1.5, "cm"),
                heatmap_width = unit(length(comb_name(m6))*1.5, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#A9C0C9", "#D8D7CD"),
                comb_order = order(lapply(comb_name(m6), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m6), 
                                   decreasing = T),
                top_annotation = upset_top_annotation(m6, add_numbers = TRUE,
                                                      numbers_gp = gpar(fontsize = 8),
                                                      numbers_rot = 0,
                                                      annotation_name_rot = 90,
                                                      height = unit(3, "cm")),
                right_annotation = upset_right_annotation(m6, add_numbers = TRUE,
                                                          width = unit(3, "cm")), 
                row_names_gp = gpar(fontface ='bold'))


if (save_or_not) {
  
  rowAnn_order_mt6 <- rowAnn_mt6[order(rowSums(rowAnn_mt6), decreasing = T),]
  
  colAnn_order_mt6 <- colAnn_mt6[comb_name(m6)[order(lapply(comb_name(m6),
                                                         function(x) {
                                                           strsplit(x = x, split = "") %>%
                                                             unlist() %>%
                                                             as.numeric() %>%
                                                             sum()
                                                           }) %>% unlist() * -1,
                                                  comb_size(m6),
                                                  decreasing = T)],] %>% t()

  bICI_hm_pdf6 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                            Prefix = "Union-Heatmap-bICI-Wilcox_MN-MM", Suffix = "pdf")
    
  pdf(file = bICI_hm_pdf6, width = 30, height = 5)
  draw(tmp_ht6)
  draw(sim_ht6)
  dev.off()
  colAnn_csv6 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                           Prefix = "Union-Matrix-ColAnno-Wilcox_MN-MM", Suffix = "csv")
  write.csv(x = colAnn_order_mt6, file = colAnn_csv6)
  rowAnn_csv6 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                           Prefix = "Union-Matrix-RowAnno-Wilcox_MN-MM", Suffix = "csv")
  write.csv(x = rowAnn_order_mt6, file = rowAnn_csv6)

}

############### NSCLC ############### 

nsclc_union_list <- union_melt_wil_cohort_list

nsclc_union_list <- lapply(union_melt_wil_cohort_list, function(x){
  x1 <- x[grep(x$CohDis, pattern = "_NSCLC$"),]
  x1[x1$Feature %in% (table(x1$Feature)[table(x1$Feature)>1] %>% names()), ]
})


lt_list7 <- list()
m_list7 <- list()
lt7 <- list()
save_or_not <- T


for (sel_kingdoms in names(nsclc_union_list)[c(-3)]) {
  lt_list7[[sel_kingdoms]] <- list()
  for (sel_disease in levels(unique(nsclc_union_list[[sel_kingdoms]]$CohDis)) %>% grep(., pattern = "_NSCLC$", value = T)) {
    lt_list7[[sel_kingdoms]][[sel_disease]] <- nsclc_union_list[[sel_kingdoms]] %>%
      subset(CohDis == sel_disease, select = "Feature") %>% unlist() %>% as.character()
    lt7[[sel_disease]] <- c(lt7[[sel_disease]], lt_list7[[sel_kingdoms]][[sel_disease]])
  }
  m_list7[[sel_kingdoms]] <- make_comb_mat(lt_list7[[sel_kingdoms]])
}
m7 = make_comb_mat(lt7)

rowAnn_mt7 <- data.frame(k__Bacteria = set_size(m_list7$k__Bacteria),
                         k__Eukaryota = set_size(m_list7$k__Eukaryota),
                         k__Archaea = 0,
                         k__Viruses = set_size(m_list7$k__Viruses)) %>%
  as.matrix()

colAnn_mt7 <- matrix(data = 0, 
                    nrow = length(comb_name(m7)), 
                    ncol = 4,
                    dimnames = list(comb_name(m7),
                                    names(tax_col)))

for (sub_k in names(m_list7)) {
  for (InSize in names(comb_size(m_list7[[sub_k]]))) {
    colAnn_mt7[InSize, sub_k] <- comb_size(m_list7[[sub_k]])[InSize]
  }
}

label_df7 <- matrix(data = NA, nrow = 0, 
                   ncol = c(length(x = set_name(m7)) + 2),
                   dimnames = list(c(), c( set_name(m7), "Num", "Label"))) %>%
  as.data.frame()
label_colname <- colnames(label_df7)

for (sel_cname in comb_name(m7)) {
  tmp_lab <- extract_comb(m7, sel_cname)
  tmp_df <- matrix(rep(c(strsplit(x = sel_cname, split = "") %>% 
                         unlist() %>% as.numeric(), length(tmp_lab)), 
                       each = length(tmp_lab)),
                   nrow = length(tmp_lab)) %>%
    cbind(., tmp_lab) 
  colnames(tmp_df) <- label_colname
  label_df7 <- rbind(label_df7, tmp_df)
}



label_df7$Group_Num <-  apply(label_df7[, c(1:length(set_name(m7)))], 1, function(x) as.numeric(x) %>% sum)

if (save_or_not) {
  label_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                          Prefix = "Union-Melt-label-Wilcox_MN-NSCLC-coh", Suffix = "csv")
  write.csv(x = label_df7, file = label_csv, row.names = F)
}


tmp_ht7 <- UpSet(m7, set_order = order(set_size(m7), decreasing = T), 
                heatmap_height = unit(length(set_name(m7))*1.5+4, "cm"),
                heatmap_width = unit(length(comb_name(m7))*1.5+8, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#A9C0C9", "#D8D7CD"),
                comb_order = order(lapply(comb_name(m7), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m7), 
                                   decreasing = T),
                top_annotation = HeatmapAnnotation(
                  "Intersection\nsize" = anno_barplot(colAnn_mt7, annotation_name_side = "left",
                                                     gp = gpar(fill = tax_col, 
                                                               col = NA),
                                                     height = unit(3, "cm"),
                                                     border = F),
                  annotation_name_rot = 90,
                  annotation_name_side = "left"),
                right_annotation = rowAnnotation(
                  "Set size" = anno_barplot(rowAnn_mt7,
                                            gp = gpar(fill = tax_col,
                                                      col = NA),
                                            # add_numbers = TRUE,
                                            width = unit(3, "cm"),
                                            border = F)),
                row_names_gp = gpar(fontface ='bold'))


sim_ht7 <- UpSet(m7, set_order = order(set_size(m7), decreasing = T), 
                heatmap_height = unit(length(set_name(m7))*1.5+4, "cm"),
                heatmap_width = unit(length(comb_name(m7))*1.5+8, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#A9C0C9", "#D8D7CD"),
                comb_order = order(lapply(comb_name(m7), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m7), 
                                   decreasing = T),
                top_annotation = upset_top_annotation(m7, add_numbers = TRUE,
                                                      numbers_gp = gpar(fontsize = 8),
                                                      numbers_rot = 0,
                                                      annotation_name_rot = 90,
                                                      height = unit(3, "cm")),
                right_annotation = upset_right_annotation(m7, add_numbers = TRUE,
                                                          width = unit(3, "cm")), 
                row_names_gp = gpar(fontface ='bold'))


if (save_or_not) {
  
  rowAnn_order_mt7 <- rowAnn_mt7[order(rowSums(rowAnn_mt7), decreasing = T),]
  
  colAnn_order_mt7 <- colAnn_mt7[comb_name(m7)[order(lapply(comb_name(m7),
                                                         function(x) {
                                                           strsplit(x = x, split = "") %>%
                                                             unlist() %>%
                                                             as.numeric() %>%
                                                             sum()
                                                           }) %>% unlist() * -1,
                                                  comb_size(m7),
                                                  decreasing = T)],] %>% t()

  bICI_hm_pdf7 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                            Prefix = "Union-Heatmap-bICI-Wilcox_MN-NSCLC", Suffix = "pdf")
    
  pdf(file = bICI_hm_pdf7, width = 30, height = 5)
  draw(tmp_ht7)
  draw(sim_ht7)
  dev.off()
  
  colAnn_csv7 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                           Prefix = "Union-Matrix-ColAnno-Wilcox_MN-NSCLC", Suffix = "csv")
  write.csv(x = colAnn_order_mt7, file = colAnn_csv7)
  rowAnn_csv7 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                           Prefix = "Union-Matrix-RowAnno-Wilcox_MN-NSCLC", Suffix = "csv")
  write.csv(x = rowAnn_order_mt7, file = rowAnn_csv7)

}



```



### Heatmap (MM NSCLC)

- Fold change --> `color`

- Candidates --> `row`

- Cohort --> `column`

- p value --> `label`


```{r hm MM NSCLC}
calculate_or_not <- T
save_or_not <- T

sel_cand_df4 <- label_df4[label_df4$Group_Num >= 2, ]
rownames(sel_cand_df4) <- sel_cand_df4$Label
sel_cand_df4$Kingdom <- lapply(sel_cand_df4$Label, function(x) {
  k <- strsplit( x = x, split = "\\|") %>% unlist(); return(k[1])
}) %>% unlist()

sel_cand_df4$ShortName <- lapply(sel_cand_df4$Label, function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()

sel_cand_csv4 <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis"),
                          Prefix = "Union-Wilcox_MN-MM-Kingdom-ShortName", Suffix = "csv")
write.csv(x = sel_cand_df4, file = sel_cand_csv4, row.names = F)



sel_mn_bac_df4 <- bICI_sel_list$k__Bacteria[rownames(bICI_sel_list$k__Bacteria) %in% sel_cand_df4$Label, ]

if (calculate_or_not) {
  tmp_path <- getwd()
  dir.create(paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds"), showWarnings = F, recursive = T)
  setwd(paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds"))
  tmp_meta4 <- bICI_ds_list$k__Bacteria$sample_table
  sel_mn_bac_df4 <- bICI_sel_list$k__Bacteria[rownames(bICI_sel_list$k__Bacteria) %in% sel_cand_df4$Label, tmp_meta4$Disease == "MM"]
  tmp_meta4$CohDis <- paste0(tmp_meta4$Cohort_name, "_", tmp_meta4$Disease)
  wil_bac_cohort4 <- Wil_Matrix(sub_otu = sel_mn_bac_df4, p_cutoff = 0.05, 
                                rarefy_or_not = F,
                                sub_meta = tmp_meta4, comparison_rank = c("R", "NR"),
                                sub_catolog = "CohDis", sub_comparison = "Response")
  mn_bac_cohort_rds4 <- FileCreate(DirPath = "./",
                                  Prefix = "Wilcox-MN_Idx-Bacteria-Cohort-MM", Suffix = "rds")
  saveRDS(object = wil_bac_cohort4, file = mn_bac_cohort_rds4)
  
  tmp_meta4 <- bICI_ds_list$k__Eukaryota$sample_table
  sel_mn_euk_df4 <- bICI_sel_list$k__Eukaryota[rownames(bICI_sel_list$k__Eukaryota) %in% sel_cand_df4$Label, tmp_meta4$Disease == "MM"]
  tmp_meta4$CohDis <- paste0(tmp_meta4$Cohort_name, "_", tmp_meta4$Disease)
  wil_euk_cohort4 <- Wil_Matrix(sub_otu = sel_mn_euk_df4, p_cutoff = 0.05, 
                                rarefy_or_not = F, 
                                sub_meta = tmp_meta4, comparison_rank = c("R", "NR"),
                                sub_catolog = "CohDis", sub_comparison = "Response")
  mn_euk_cohort_rds4 <- FileCreate(DirPath = "./",
                                  Prefix = "Wilcox-MN_Idx-Eukaryota-Cohort-MM", Suffix = "rds")
  saveRDS(object = wil_euk_cohort4, file = mn_euk_cohort_rds4)
  
  tmp_meta4 <- bICI_ds_list$k__Archaea$sample_table
  sel_mn_arc_df4 <- bICI_sel_list$k__Archaea[rownames(bICI_sel_list$k__Archaea) %in% sel_cand_df4$Label, tmp_meta4$Disease == "MM"]
  tmp_meta4$CohDis <- paste0(tmp_meta4$Cohort_name, "_", tmp_meta4$Disease)
  wil_arc_cohort4 <- Wil_Matrix(sub_otu = sel_mn_arc_df4, p_cutoff = 0.05, 
                                rarefy_or_not = F, 
                                sub_meta = tmp_meta4, comparison_rank = c("R", "NR"),
                                sub_catolog = "CohDis", sub_comparison = "Response")
  mn_arc_cohort_rds4 <- FileCreate(DirPath = "./",
                                  Prefix = "Wilcox-MN_Idx-Archaea-Cohort-MM", Suffix = "rds")
  saveRDS(object = wil_arc_cohort4, file = mn_arc_cohort_rds4)
  setwd(tmp_path)
  # sel_mn_vir_df4 <- bICI_sel_list$k__Viruses[rownames(bICI_sel_list$k__Viruses) %in% sel_cand_df4$Label, ]
  # tmp_meta <- bICI_ds_list$k__Viruses$sample_table
  # tmp_meta$CohDis <- paste0(tmp_meta$Cohort_name, "_", tmp_meta$Disease)
  # wil_vir_cohort <- Wil_Matrix(sub_otu = t(sel_mn_vir_df), p_cutoff = 0.05, rarefy_or_not = F,
  #                             sub_meta = tmp_meta, comparison_rank = c("R", "NR"),
  #                             sub_catolog = "CohDis", sub_comparison = "Response")
  # wil_vir_disease <- Wil_Matrix(sub_otu = t(sel_mn_vir_df), p_cutoff = 0.05, rarefy_or_not = F,
  #                               sub_meta = bICI_ds_list$k__Viruses$sample_table,
  #                               comparison_rank = c("R", "NR"),
  #                               sub_catolog = "Disease", sub_comparison = "Response")
  # mn_vir_cohort_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
  #                                 Prefix = "Wilcox-MN-Viruses-Cohort", Suffix = "rds")
  # saveRDS(object = wil_vir_cohort, file = mn_vir_cohort_rds)
  # mn_vir_disease_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
  #                                 Prefix = "Wilcox-MN-Viruses-disease", Suffix = "rds")
  # saveRDS(object = wil_vir_disease, file = mn_vir_disease_rds)
  
}else{
  wil_bac_cohort4 <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds/2023-02-23-Wilcox-MN_Idx-",
                                         "Bacteria", "-Cohort-MM-v1.0.rds"))
  wil_euk_cohort4 <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds/2023-02-23-Wilcox-MN_Idx-",
                                         "Eukaryota", "-Cohort-MM-v1.0.rds"))
  wil_arc_cohort4 <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/rds/2023-02-23-Wilcox-MN_Idx-",
                                         "Archaea", "-Cohort-MM-v1.0.rds"))
  
}

hm_col_df4 <- log2(do.call("rbind",
                          list(wil_bac_cohort4$FC_df[, -ncol(wil_bac_cohort4$FC_df)],
                               wil_euk_cohort4$FC_df[, -ncol(wil_euk_cohort4$FC_df)],
                               wil_arc_cohort4$FC_df[, -ncol(wil_arc_cohort4$FC_df)])))
hm_col_df4[hm_col_df4 >5] <- 5
hm_col_df4[hm_col_df4 < -5] <- -5
tmp_lab_df4 <- do.call("rbind",
                      list(wil_bac_cohort4$wil_df[, -ncol(wil_bac_cohort4$wil_df)],
                           wil_euk_cohort4$wil_df[, -ncol(wil_euk_cohort4$wil_df)],
                           wil_arc_cohort4$wil_df[, -ncol(wil_arc_cohort4$wil_df)]))
hm_lab_df4 <- ifelse(tmp_lab_df4 > .05, "",
                    ifelse(tmp_lab_df4 >.01, "*",
                           ifelse(tmp_lab_df4 >.01, "**", "***")))


col_fun = colorRamp2(c(-5, 0, 5), c("#00468bcc", "#ffffff", "#ed0000cc"))



top_anno4 <- HeatmapAnnotation(Disease = colnames(hm_col_df4) %>% gsub("^.+_", "", .), 
                              col = list(Disease = c(cancer_col, All = "black")),
                              gp = gpar(col = 'black'))
right_anno4 <- rowAnnotation(Kingdom = sel_cand_df4[rownames(hm_col_df4), ]$Kingdom,
                            col = list(Kingdom = c(tax_col)), 
                            gp = gpar(col = 'black'))

left_col_fun4 <- function(x){
  ifelse(x == 0, "white", ifelse(x == 1, "#84bd007f", 
                                 ifelse(x == 2, "#8ca252cc",
                                        ifelse(x == 3, "#637939cc", "#20854ecc"))))
}

left_anno4 <- rowAnnotation(Num = anno_barplot(rowSums(hm_lab_df4[, 1:7] != ""),
                                              axis_param = list(direction = "reverse"),
                                              gp = gpar(fill = do.call(left_col_fun4, 
                                                                       list(rowSums(hm_lab_df4[, 1:7] != "")))),
                                              width = unit(1.5, "cm")))


comb_hm4 <- Heatmap(hm_col_df4 %>% as.matrix(),
                   name = "log2(R/NR)",
                   width = unit(ncol(hm_col_df4)/2, "cm"),
                   height = unit(nrow(hm_col_df4)/2, "cm"),
                   cluster_column_slices = F,
                   cell_fun = function(j, i, x, y, width, height, fill) {
                     grid.text(hm_lab_df4[i, j], x, y, gp = gpar(fontsize = 10))},
                   row_labels = sel_cand_df4[rownames(hm_col_df4) ,'ShortName'],
                   col = col_fun,
                   right_annotation = right_anno4, 
                   border = "grey", heatmap_legend_param = list(direction = "horizontal")) 

if (save_or_not) {
  tmp_path <- getwd()
  target_path <- paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/Comb-Heatmap")
  dir.create(path = target_path, showWarnings = F, recursive = T)
  setwd(target_path)
  
  comb_hm_pdf4 <- FileCreate(DirPath = './',
                            Prefix = "Heatmap-combine-cohort-disease-foldchange-MM", Suffix = "pdf")
  pdf(file = comb_hm_pdf4, width = ncol(hm_col_df4)/25.4 + 15, height = nrow(hm_col_df4)*5/25.4 + 5)
  draw(comb_hm4, merge_legend = TRUE, 
       heatmap_legend_side = "left",
       annotation_legend_side = "left")
  dev.off()
  setwd(tmp_path)
}


```



### Survival Curve

#### Calculate 

```{r survival curve calculate}
bICI_meta <- list(k__Bacteria = bICI_ds_list$k__Bacteria$sample_table %>% 
                    select(Antibiotics, Age, Gender, Respond_detail, RECIST, Sequencing_platform,
                           Cohort_name, Disease, Response, PFS, PFS_status) %>%
                    filter(!is.na(.data[["PFS"]]),
                           !(.data[["PFS"]] %in% c("", "yes", "no"))),
                  k__Eukaryota = bICI_ds_list$k__Eukaryota$sample_table %>% 
                    select(Antibiotics, Age, Gender, Respond_detail, RECIST, Sequencing_platform,
                           Cohort_name, Disease, Response, PFS, PFS_status) %>%
                    filter(!is.na(.data[["PFS"]]),
                           !(.data[["PFS"]] %in% c("", "yes", "no"))),
                  k__Archaea = bICI_ds_list$k__Archaea$sample_table %>% 
                    select(Antibiotics, Age, Gender, Respond_detail, RECIST, Sequencing_platform,
                           Cohort_name, Disease, Response, PFS, PFS_status) %>%
                    filter(!is.na(.data[["PFS"]]),
                           !(.data[["PFS"]] %in% c("", "yes", "no"))),
                  k__Viruses = bICI_ds_list$k__Viruses$sample_table %>% 
                    select(Antibiotics, Age, Gender, Respond_detail, RECIST, Sequencing_platform,
                           Cohort_name, Disease, Response, PFS, PFS_status) %>%
                    filter(!is.na(.data[["PFS"]]),
                           !(.data[["PFS"]] %in% c("", "yes", "no"))))

sub_survival_list <- list()
### NSCLC ###
nsclc_sel_list <- lapply(nsclc_union_list, 
                         function(x){
                           x1 <-table(x$Feature)
                           names(x1)[x1>=2]
                           }
                         ) %>% 
  unlist() %>% 
  unname()
### MM ###
mm_sel_list <- lapply(mm_union_list, 
                      function(x){
                        x1 <-table(x$Feature)
                        names(x1)[x1>=3]
                        }
                      ) %>% 
  unlist() %>% 
  unname()


```

#### Euk, NSCLC

```{r survival Euk NSCLC}
### Eukaryota ###
#### NSCLC ####
euk_nsclc_surv_df <- cbind(bICI_meta$k__Eukaryota,
                           euk_mn_res2$nm_df %>%
                             filter(row.names(.) %in% rownames(bICI_meta$k__Eukaryota)) %>%
                             select(grep("k__Eukaryota", nsclc_sel_list, value = T))) %>%
  filter(Disease == "NSCLC") %>%
  mutate(PFS_status = ifelse(PFS_status == "Dead", 1, 0),
         PFS = as.numeric(PFS))

colnames(euk_nsclc_surv_df) <- lapply(colnames(euk_nsclc_surv_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()

euk_nsclc_var <- euk_nsclc_surv_df %>%
  colnames() %>%
  grep("^s__", ., value = T)

euk_nsclc_res.cut <- surv_cutpoint(euk_nsclc_surv_df, # data set
                             time = "PFS", # survival period
                             event = "PFS_status", # survival stat
                             variables = euk_nsclc_var) # candidate
euk_nsclc_res.cat <- surv_categorize(euk_nsclc_res.cut)


sub_survival_list$NSCLC$Euk <- euk_nsclc_surv_df
sub_survival_list$NSCLC$Euk[, euk_nsclc_var] <- euk_nsclc_res.cat[, euk_nsclc_var]
```

#### Bac, NSCLC

```{r survival bac NSCLC}
### Bacteria ###
#### NSCLC ####
bac_nsclc_surv_df <- cbind(bICI_meta$k__Bacteria,
                           bac_mn_res2$nm_df %>% 
                             filter(row.names(.) %in% rownames(bICI_meta$k__Bacteria)) %>%
                       select(grep("k__Bacteria", nsclc_sel_list, value = T))) %>%
  filter(Disease == "NSCLC") %>%
  mutate(PFS_status = ifelse(PFS_status == "Dead", 1, 0),
         PFS = as.numeric(PFS))

colnames(bac_nsclc_surv_df) <- lapply(colnames(bac_nsclc_surv_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()

bac_nsclc_var <- bac_nsclc_surv_df %>%
  colnames() %>%
  grep("^s__", ., value = T)

bac_nsclc_res.cut <- surv_cutpoint(bac_nsclc_surv_df, # data set
                             time = "PFS", # survival period
                             event = "PFS_status", # survival stat
                             variables = bac_nsclc_var)
bac_nsclc_res.cat <- surv_categorize(bac_nsclc_res.cut)

sub_survival_list$NSCLC$Bac <- bac_nsclc_surv_df
sub_survival_list$NSCLC$Bac[, bac_nsclc_var] <- bac_nsclc_res.cat[, bac_nsclc_var]
```

#### Vir, NSCLC

```{r survival Vir NSCLC}
### Viruses  ###
#### NSCLC ####
vir_nsclc_surv_df <- cbind(bICI_meta$k__Viruses,
                     vir_mn_res2$nm_df %>% 
                       filter(row.names(.) %in% rownames(bICI_meta$k__Viruses)) %>%
                       select(grep("k__Viruses", nsclc_sel_list, value = T))) %>%
  filter(Disease == "NSCLC") %>%
  mutate(PFS_status = ifelse(PFS_status == "Dead", 1, 0),
         PFS = as.numeric(PFS))

colnames(vir_nsclc_surv_df) <- lapply(colnames(vir_nsclc_surv_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()

vir_nsclc_var <- vir_nsclc_surv_df %>%
  colnames() %>%
  grep("^s__", ., value = T)

vir_nsclc_res.cut <- surv_cutpoint(vir_nsclc_surv_df, # data set
                                   time = "PFS", # survival period
                                   event = "PFS_status", # survival stat
                                   variables = vir_nsclc_var)

vir_nsclc_res.cat <- surv_categorize(vir_nsclc_res.cut)

sub_survival_list$NSCLC$Vir <- vir_nsclc_surv_df
sub_survival_list$NSCLC$Vir[, vir_nsclc_var] <- vir_nsclc_res.cat[, vir_nsclc_var]
```

#### Bac, MM

```{r survival bac MM}
### Bacteria ###
#### MM ####
bac_mm_surv_df <- cbind(bICI_meta$k__Bacteria,
                     bac_mn_res2$nm_df %>% 
                       filter(row.names(.) %in% rownames(bICI_meta$k__Bacteria)) %>%
                       select(grep("k__Bacteria", mm_sel_list, value = T))) %>%
  filter(Disease == "MM") %>%
  mutate(PFS_status = ifelse(PFS_status == "Dead", 1, 0),
         PFS = as.numeric(PFS))

colnames(bac_mm_surv_df) <- lapply(colnames(bac_mm_surv_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()

bac_mm_var <- bac_mm_surv_df %>%
  colnames() %>%
  grep("^s__", ., value = T)


bac_mm_res.cut <- surv_cutpoint(bac_mm_surv_df, # data set
                             time = "PFS", # survival period
                             event = "PFS_status", # survival stat
                             variables = bac_mm_var)
bac_mm_res.cat <- surv_categorize(bac_mm_res.cut)

sub_survival_list$MM$Bac <- bac_mm_surv_df
sub_survival_list$MM$Bac[, bac_mm_var] <- bac_mm_res.cat[, bac_mm_var]
```

#### Survival Curve Plot

```{r survival plot}

sel_sur_list <- list(bac_nsclc_res.cat,
                     bac_nsclc_res.cat,
                     euk_nsclc_res.cat,
                     euk_nsclc_res.cat,
                     euk_nsclc_res.cat,
                     vir_nsclc_res.cat,
                     bac_mm_res.cat)
names(sel_sur_list) <- lapply(c(nsclc_sel_list, mm_sel_list), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()

survival_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/Survival"),
                               Prefix = "SurCurve-CohDis", Suffix = "pdf")
pdf(file = survival_pdf, width = 12, height = 6)

for (sel_cand in names(sel_sur_list)) {
  fit <- survfit(as.formula(paste0("Surv(PFS, PFS_status) ~ ", sel_cand)), 
                 data = sel_sur_list[[sel_cand]])
  g <- ggsurvplot(fit,
                  palette = c("#ed0000cc", "#00468bcc"),
                  data = sel_sur_list[[sel_cand]],
                  risk.table = TRUE,
                  pval = T, 
                  title = sel_cand,
                  xscale = "d_y",
                  break.x.by = 365, 
                  xlim=c(0,365*4),
                  tables.y.text = F)
  print(g)
}

dev.off()

```

#### multivariate analysis

```{r forest plot}
forest_list <- list()

## nsclc euk
sub_survival_list$NSCLC$Euk %>% 
  select(c("PFS", "PFS_status", "s__Eimeria_brunetti", "Age", "Gender")) %>%
  na.omit() %>%
  analyse_multivariate(vars(PFS, PFS_status),
                       covariates = vars(Age, Gender, s__Eimeria_brunetti)) ->
  forest_list$NSCLC$Euk$s__Eimeria_brunetti

sub_survival_list$NSCLC$Euk %>% 
  select(c("PFS", "PFS_status", "s__Aspergillus_tamarii", "Age", "Gender")) %>%
  na.omit() %>%
  analyse_multivariate(vars(PFS, PFS_status),
                       covariates = vars(Age, Gender, s__Aspergillus_tamarii)) ->
  forest_list$NSCLC$Euk$s__Aspergillus_tamarii

sub_survival_list$NSCLC$Euk %>% 
  select(c("PFS", "PFS_status", "s__Fusarium_anguioides", "Age", "Gender")) %>%
  na.omit() %>%
  analyse_multivariate(vars(PFS, PFS_status),
                       covariates = vars(Age, Gender, s__Fusarium_anguioides)) ->
  forest_list$NSCLC$Euk$s__Fusarium_anguioides

## nsclc bac
sub_survival_list$NSCLC$Bac %>% 
  select(c("PFS", "PFS_status", "s__Clostridia_bacterium", "Age", "Gender")) %>%
  na.omit() %>%
  analyse_multivariate(vars(PFS, PFS_status),
                       covariates = vars(Age, Gender, s__Clostridia_bacterium)) ->
  forest_list$NSCLC$Bac$s__Clostridia_bacterium

sub_survival_list$NSCLC$Bac %>% 
  select(c("PFS", "PFS_status", "s__Candidatus_Gastranaerophilales_bacterium_MH_37", 
           "Age", "Gender")) %>%
  na.omit() %>%
  analyse_multivariate(vars(PFS, PFS_status),
                       covariates = vars(Age, Gender, s__Candidatus_Gastranaerophilales_bacterium_MH_37)) ->
  forest_list$NSCLC$Bac$s__Candidatus_Gastranaerophilales_bacterium_MH_37


## nsclc vir
sub_survival_list$NSCLC$Vir %>% 
  select(c("PFS", "PFS_status", "s__crAssphage_cr127_1", "Age", "Gender")) %>%
  na.omit() %>%
  analyse_multivariate(vars(PFS, PFS_status),
                       covariates = vars(Age, Gender, s__crAssphage_cr127_1)) ->
  forest_list$NSCLC$Vir$s__crAssphage_cr127_1


## mm bac
sub_survival_list$MM$Bac %>% 
  select(c("PFS", "PFS_status", "s__Faecalibacterium_prausnitzii", "Age", "Gender")) %>%
  na.omit() %>%
  analyse_multivariate(vars(PFS, PFS_status),
                       covariates = vars(Age, Gender, s__Faecalibacterium_prausnitzii)) ->
  forest_list$MM$Bac$s__Faecalibacterium_prausnitzii

f_plot <- list()

f_plot$s__Eimeria_brunetti <- forest_plot(forest_list$NSCLC$Euk$s__Eimeria_brunetti, 
                                          orderer = ~ order(HR),
                                          labels_displayed = c("factor", "n"),
                                          relative_widths = c(2, 4, 2))
f_plot$s__Aspergillus_tamarii <- forest_plot(forest_list$NSCLC$Euk$s__Aspergillus_tamarii, 
                                          orderer = ~ order(HR),
                                          labels_displayed = c("factor", "n"),
                                          relative_widths = c(2, 4, 2))
f_plot$s__Fusarium_anguioides <- f_plot$s__Eimeria_brunetti <- forest_plot(forest_list$NSCLC$Euk$s__Fusarium_anguioides, 
                                                                           orderer = ~ order(HR),
                                                                           labels_displayed = c("factor", "n"),
                                                                           relative_widths = c(2, 4, 2))
f_plot$s__Candidatus_Gastranaerophilales_bacterium_MH_37 <- forest_plot(forest_list$NSCLC$Bac$s__Candidatus_Gastranaerophilales_bacterium_MH_37, 
                                                                        orderer = ~ order(HR),
                                                                        labels_displayed = c("factor", "n"),
                                                                        relative_widths = c(2, 4, 2))
f_plot$s__Clostridia_bacterium <- forest_plot(forest_list$NSCLC$Bac$s__Clostridia_bacterium, 
                                              orderer = ~ order(HR),
                                              labels_displayed = c("factor", "n"),
                                              relative_widths = c(2, 4, 2))
f_plot$s__crAssphage_cr127_1 <- forest_plot(forest_list$NSCLC$Vir$s__crAssphage_cr127_1, 
                                            orderer = ~ order(HR),
                                            labels_displayed = c("factor", "n"),
                                            relative_widths = c(2, 4, 2))
f_plot$s__Faecalibacterium_prausnitzii <- forest_plot(forest_list$MM$Bac$s__Faecalibacterium_prausnitzii, 
                                                      orderer = ~ order(HR),
                                                      labels_displayed = c("factor", "n"),
                                                      relative_widths = c(2, 4, 2))


cohDis_survival_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN-CohDis/Survival"),
                               Prefix = "MultVar-CohDis", Suffix = "pdf")
pdf(file = cohDis_survival_pdf, width = 15, height = 5)
for (sel_cand in  names(sel_sur_list)) {
  print(f_plot[[sel_cand]])
}
dev.off()






```


<!--
### Other validation

```{r other validation }
# paired ICI
## EUK NSCLC
pICI_euk_NSCLC <- cbind(pICI_ds_list$k__Eukaryota$sample_table,
                     euk_mn_res2$nm_df[rownames(pICI_ds_list$k__Eukaryota$sample_table),] %>% 
                       select(ends_with(c("Eimeria_brunetti", 
                                          "Aspergillus_tamarii", 
                                          "Fusarium_anguioides")))) %>%
  filter(Disease == "NSCLC")
colnames(pICI_euk_NSCLC) <- lapply(colnames(pICI_euk_NSCLC), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
colnames(pICI_euk_NSCLC)[(ncol(pICI_euk_NSCLC)-2):ncol(pICI_euk_NSCLC)] <- c("EB", "AT", "FA")
pICI_euk_NSCLC$Period2 <- factor(paste0(pICI_euk_NSCLC$Response, "-", 
                                     ifelse(pICI_euk_NSCLC$Period == "Before_ICI", 
                                            "Before_ICI", "ICI_treat")), 
                              levels = c("NR-Before_ICI", 
                                         "NR-ICI_treat",
                                         "R-Before_ICI", 
                                         "R-ICI_treat"))
## BAC MM
pICI_bac_MM <- cbind(pICI_ds_list$k__Bacteria$sample_table,
                     bac_mn_res2$nm_df[rownames(pICI_ds_list$k__Bacteria$sample_table),] %>% 
                       select(ends_with(c("Faecalibacterium_prausnitzii")))) %>%
  filter(Disease == "MM")
colnames(pICI_bac_MM) <- lapply(colnames(pICI_bac_MM), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
colnames(pICI_bac_MM)[(ncol(pICI_bac_MM)-0):ncol(pICI_bac_MM)] <- c("FP")
pICI_bac_MM$Period2 <- factor(paste0(pICI_bac_MM$Response, "-", 
                                     ifelse(pICI_bac_MM$Period == "Before_ICI", 
                                            "Before_ICI", "ICI_treat")), 
                              levels = c("NR-Before_ICI", 
                                         "NR-ICI_treat",
                                         "R-Before_ICI", 
                                         "R-ICI_treat"))

## BAC NSCLC
pICI_bac_NSCLC <- cbind(pICI_ds_list$k__Bacteria$sample_table,
                     bac_mn_res2$nm_df[rownames(pICI_ds_list$k__Bacteria$sample_table),] %>% 
                       select(ends_with(c("Candidatus_Gastranaerophilales_bacterium_MH_37",
                                          "Clostridia_bacterium")))) %>%
  filter(Disease == "NSCLC")
colnames(pICI_bac_NSCLC) <- lapply(colnames(pICI_bac_NSCLC), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
colnames(pICI_bac_NSCLC)[(ncol(pICI_bac_NSCLC)-1):ncol(pICI_bac_NSCLC)] <- c("CG", "CB")
pICI_bac_NSCLC$Period2 <- factor(paste0(pICI_bac_NSCLC$Response, "-", 
                                     ifelse(pICI_bac_NSCLC$Period == "Before_ICI", 
                                            "Before_ICI", "ICI_treat")), 
                              levels = c("NR-Before_ICI", 
                                         "NR-ICI_treat",
                                         "R-Before_ICI", 
                                         "R-ICI_treat"))

## VIR NSCLC
pICI_vir_NSCLC <- cbind(pICI_ds_list$k__Viruses$sample_table,
                     vir_mn_res2$nm_df[rownames(pICI_ds_list$k__Viruses$sample_table),] %>% 
                       select(ends_with(c("crAssphage_cr127_1")))) %>%
  filter(Disease == "NSCLC")
colnames(pICI_vir_NSCLC) <- lapply(colnames(pICI_vir_NSCLC), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
colnames(pICI_vir_NSCLC)[(ncol(pICI_vir_NSCLC)-0):ncol(pICI_vir_NSCLC)] <- c("crA")
pICI_vir_NSCLC$Period2 <- factor(paste0(pICI_vir_NSCLC$Response, "-", 
                                     ifelse(pICI_vir_NSCLC$Period == "Before_ICI", 
                                            "Before_ICI", "ICI_treat")), 
                              levels = c("NR-Before_ICI", 
                                         "NR-ICI_treat",
                                         "R-Before_ICI", 
                                         "R-ICI_treat"))


sub_EB <- ggplot(pICI_euk_NSCLC, aes(x = Period2, y = EB)) +
  theme_Publication() +
  geom_violin(alpha = 0, aes(col = Period2)) +
  geom_boxplot(width = 0.1, outlier.colour = NA) +
  scale_y_log10() +
  geom_jitter(aes(col = Period2), alpha = 0.5) +
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)+
  theme(legend.position = "none")
  
sub_AT <- ggplot(pICI_euk_NSCLC, aes(x = Period2, y = AT)) +
  theme_Publication() +
  geom_boxplot(width = 0.3, outlier.colour = NA, alpha = 0) +
  scale_y_log10() +
  geom_jitter(aes(col = Period2), alpha = 0.5) +
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)+
  theme(legend.position = "none") 

sub_FA <- ggplot(pICI_euk_NSCLC, aes(x = Period2, y = FA)) +
  theme_Publication() +
  geom_boxplot(width = 0.3, outlier.colour = NA, alpha = 0) +
  scale_y_log10() +
  geom_jitter(aes(col = Period2), alpha = 0.5) +
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)+
  theme(legend.position = "none")

sub_FP <- ggplot(pICI_bac_MM, aes(x = Period2, y = FP)) +
  theme_Publication() +
  geom_boxplot(width = 0.3, outlier.colour = NA, alpha = 0) +
  scale_y_log10() +
  geom_jitter(aes(col = Period2), alpha = 0.5) +
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)+
  theme(legend.position = "none")

sub_CG <- ggplot(pICI_bac_NSCLC, aes(x = Period2, y = CG)) +
  theme_Publication() +
  geom_boxplot(width = 0.3, outlier.colour = NA, alpha = 0) +
  scale_y_log10() +
  geom_jitter(aes(col = Period2), alpha = 0.5) +
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)+
  theme(legend.position = "none")

sub_CB <- ggplot(pICI_bac_NSCLC, aes(x = Period2, y = CB)) +
  theme_Publication() +
  geom_boxplot(width = 0.3, outlier.colour = NA, alpha = 0) +
  scale_y_log10() +
  geom_jitter(aes(col = Period2), alpha = 0.5) +
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)+
  theme(legend.position = "none")

sub_crA <- ggplot(pICI_vir_NSCLC, aes(x = Period2, y = crA)) +
  theme_Publication() +
  geom_boxplot(width = 0.3, outlier.colour = NA, alpha = 0) +
  scale_y_log10() +
  geom_jitter(aes(col = Period2), alpha = 0.5) +
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)+
  theme(legend.position = "none")






```

### Survival Curve (all combined selected)

```{r survival curve 2}


### Eukaryota ###

euk_surv_df <- cbind(bICI_meta$k__Eukaryota,
                     euk_mn_res2$nm_df[rownames(bICI_meta$k__Eukaryota),] %>% 
                       select(ends_with(c("Nemania_serpens", 
                                          "Fusarium_transvaalense",
                                          "Trichophyton_verrucosum", 
                                          "Hyphopichia_pseudoburtonii"))))
euk_surv_df$PFS_status <- ifelse(euk_surv_df$PFS_status == "Dead", 1, 0)

euk_surv_df$PFS <- as.numeric(euk_surv_df$PFS)
colnames(euk_surv_df) <- lapply(colnames(euk_surv_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()

colnames(euk_surv_df)[(ncol(euk_surv_df)-3):ncol(euk_surv_df)] <- c("NS", "FT", "TV", "HP")

euk.res.cut <- surv_cutpoint(euk_surv_df, # data set
                             time = "PFS", # survival period
                             event = "PFS_status", # survival stat
                             variables = c("NS", "FT", "TV", "HP")) # candidate
                         

euk.res.cat <- surv_categorize(euk.res.cut)

euk_surv_df2 <- euk_surv_df
euk_surv_df2[, (ncol(euk_surv_df2)-3):ncol(euk_surv_df2)] <- euk.res.cat[, (ncol(euk.res.cat)-3):ncol(euk.res.cat)]

euk_survival_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Survival"),
                               Prefix = "CombineAll-SurvivalCurve-Euk", Suffix = "pdf")
pdf(file = euk_survival_pdf, width = 10, height = 6)
fit <- survfit(Surv(PFS, PFS_status) ~ NS, data = euk.res.cat)#

ggsurvplot(fit,
           palette = c("#ed0000cc", "#00468bcc"),
           data = euk.res.cat,
           risk.table = TRUE,
           pval = T, 
           title = "Nemania serpens",
           xscale = "d_y",
           break.x.by = 365, 
           xlim=c(0,365*5))


fit <- survfit(Surv(PFS, PFS_status) ~ FT, data = euk.res.cat)#

ggsurvplot(fit,
           palette = c("#ed0000cc", "#00468bcc"),
           data = euk.res.cat,
           risk.table = TRUE,
           pval = T, 
           title = "Fusarium transvaalense",
           xscale = "d_y",
           break.x.by = 365, 
           xlim=c(0,365*5))


fit <- survfit(Surv(PFS, PFS_status) ~ TV, data = euk.res.cat)#

ggsurvplot(fit,
           palette = c("#ed0000cc", "#00468bcc"),
           data = euk.res.cat,
           risk.table = TRUE,
           pval = T, 
           title = "Trichophyton verrucosum",
           xscale = "d_y",
           break.x.by = 365, 
           xlim=c(0,365*5))


fit <- survfit(Surv(PFS, PFS_status) ~ HP, data = euk.res.cat)#

ggsurvplot(fit,
           palette = c("#ed0000cc", "#00468bcc"),
           data = euk.res.cat,
           risk.table = TRUE,
           pval = T, 
           title = "Hyphopichia pseudoburtonii",
           xscale = "d_y",
           break.x.by = 365, 
           xlim=c(0,365*5))

dev.off()


### Bacteria ###
bac_surv_df <- cbind(bICI_meta$k__Bacteria,
                     bac_mn_res2$nm_df[rownames(bICI_meta$k__Bacteria),] %>% 
                       select(ends_with(c("Coprococcus_comes", 
                                          "Hungatella_hathewayi"))))
bac_surv_df$PFS_status <- ifelse(bac_surv_df$PFS_status == "Dead", 1, 0)

bac_surv_df$PFS <- as.numeric(bac_surv_df$PFS)
colnames(bac_surv_df) <- lapply(colnames(bac_surv_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
colnames(bac_surv_df)[(ncol(bac_surv_df)-1):ncol(bac_surv_df)] <- c("CC", "HH")


bac.res.cut <- surv_cutpoint(bac_surv_df, # data set
                         time = "PFS", # survival period
                         event = "PFS_status", # survival stat
                         variables = c("CC",
                                       "HH"))

bac.res.cat <- surv_categorize(bac.res.cut)

bac_surv_df2 <- bac_surv_df
bac_surv_df2[, (ncol(bac_surv_df2)-1):ncol(bac_surv_df2)] <- bac.res.cat[, (ncol(bac.res.cat)-1):ncol(bac.res.cat)]


bac_survival_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Survival"),
                               Prefix = "CombineAll-SurvivalCurve-Bac", Suffix = "pdf")
pdf(file = bac_survival_pdf, width = 10, height = 6)


fit <- survfit(Surv(PFS, PFS_status) ~ CC, data = bac.res.cat)#

ggsurvplot(fit,
           palette = c("#ed0000cc", "#00468bcc"),
           data = bac.res.cat,
           risk.table = TRUE,
           pval = T, 
           title = "Coprococcus comes",
           xscale = "d_y",
           break.x.by = 365, 
           xlim=c(0,365*5))

fit <- survfit(Surv(PFS, PFS_status) ~ HH, data = bac.res.cat)#

ggsurvplot(fit,
           palette = c("#ed0000cc", "#00468bcc"),
           data = bac.res.cat,
           risk.table = TRUE,
           pval = T, 
           title = "Hungatella hathewayi",
           xscale = "d_y",
           break.x.by = 365, 
           xlim=c(0,365*5))

dev.off()



```


### multivariate analysis (all combined selected)

```{r forest plot 2}

bac_surv_df2 %>% 
  select(c("PFS", "PFS_status", "HH", "Antibiotics", "Age", "Gender", "Disease")) %>%
  na.omit() %>%
  analyse_multivariate(vars(PFS, PFS_status),
                       covariates = vars(Age, Gender, Disease, HH)) ->
  HH_bac_mulvar_result

forest_plot(HH_bac_mulvar_result, orderer = ~ order(HR))


bac_surv_df2 %>% 
  select(c("PFS", "PFS_status", "CC", "Antibiotics", "Age", "Gender", "Disease", "HH")) %>%
  na.omit() %>%
  analyse_multivariate(vars(PFS, PFS_status),
                       covariates = vars(Age, Gender, Disease, CC)) ->
  CC_bac_mulvar_result


bac_survival_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/"),
                               Prefix = "CombineAll-MultVar-Bac", Suffix = "pdf")
pdf(file = bac_survival_pdf, width = 10, height = 3)


forest_plot(CC_bac_mulvar_result, 
            orderer = ~ order(HR),
            labels_displayed = c("factor", "n"),
            relative_widths = c(1, 1.5, 1))

forest_plot(HH_bac_mulvar_result, 
            orderer = ~ order(HR),
            labels_displayed = c("factor", "n"),
            relative_widths = c(1, 1.5, 1))
dev.off()


```

### Performance in validation cohorts (all combined selected)

```{r sel features in validation cohorts 2}
# paired ICI samples

pICI_sel_list <- list()
iICI_sel_list <- list()
## bacteria
pICI_sel_list$k__Bacteria <- t(bac_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Bacteria$sample_table)]
iICI_sel_list$k__Bacteria <- t(bac_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Bacteria$sample_table)]
# eukaryota
pICI_sel_list$k__Eukaryota <- t(euk_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Eukaryota$sample_table)]
iICI_sel_list$k__Eukaryota <- t(euk_mn_res2$nm_df)[,rownames(iICI_ds_list$k__Eukaryota$sample_table)]
# archaea
pICI_sel_list$k__Archaea <- t(arc_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Archaea$sample_table)]
iICI_sel_list$k__Archaea <- t(arc_mn_res2$nm_df)[,rownames(iICI_ds_list$k__Archaea$sample_table)]
# viruses
pICI_sel_list$k__Viruses <- t(vir_mn_res2$nm_df)[,rownames(pICI_ds_list$k__Viruses$sample_table)]
iICI_sel_list$k__Viruses <- t(vir_mn_res2$nm_df)[,rownames(iICI_ds_list$k__Viruses$sample_table)]


# paired ICI
## EUK
pICI_euk_df <- cbind(pICI_ds_list$k__Eukaryota$sample_table,
                     euk_mn_res2$nm_df[rownames(pICI_ds_list$k__Eukaryota$sample_table),] %>% 
                       select(ends_with(c("Nemania_serpens", 
                                          "Fusarium_transvaalense",
                                          "Hyphopichia_pseudoburtonii")))) 
colnames(pICI_euk_df) <- lapply(colnames(pICI_euk_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
colnames(pICI_euk_df)[(ncol(pICI_euk_df)-2):ncol(pICI_euk_df)] <- c("NS", "FT", "HP")
pICI_euk_df$Period2 <- factor(paste0(pICI_euk_df$Response, "-", 
                                     ifelse(pICI_euk_df$Period == "Before_ICI", 
                                            "Before_ICI", "ICI_treat")), 
                              levels = c("NR-Before_ICI", 
                                         "NR-ICI_treat",
                                         "R-Before_ICI", 
                                         "R-ICI_treat"))
## BAC
pICI_bac_df <- cbind(pICI_ds_list$k__Bacteria$sample_table,
                     bac_mn_res2$nm_df[rownames(pICI_ds_list$k__Bacteria$sample_table),] %>% 
                       select(ends_with(c("Coprococcus_comes", 
                                          "Hungatella_hathewayi")))) 
colnames(pICI_bac_df) <- lapply(colnames(pICI_bac_df), function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
colnames(pICI_bac_df)[(ncol(pICI_bac_df)-1):ncol(pICI_bac_df)] <- c("CC", "HH")
pICI_bac_df$Period2 <- factor(paste0(pICI_bac_df$Response, "-", 
                                     ifelse(pICI_bac_df$Period == "Before_ICI", 
                                            "Before_ICI", "ICI_treat")), 
                              levels = c("NR-Before_ICI", 
                                         "NR-ICI_treat",
                                         "R-Before_ICI", 
                                         "R-ICI_treat"))



pICI_comparisons <- list(c("NR-Before_ICI", "NR-ICI_treat"),
                         c("NR-Before_ICI", "R-Before_ICI"),
                         c("NR-Before_ICI", "R-ICI_treat"),
                         c("R-Before_ICI", "R-ICI_treat"))



pcomp_col <- c(`NR-Before_ICI` = "tomato",
               `NR-ICI_treat` = "red",
               `R-Before_ICI` = "steelblue1",
               `R-ICI_treat` = "royalblue")




g1_NS <- ggplot(pICI_euk_df, aes(x = Period2, y = NS)) +
  theme_Publication() +
  geom_violin(alpha = 0, aes(col = Period2)) +
  geom_boxplot(width = 0.1, outlier.colour = NA) +
  # scale_y_log10() +
  geom_jitter(aes(col = Period2), alpha = 0.5) +
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)+
  theme(legend.position = "none")
  
  
g1_FT <- ggplot(pICI_euk_df, aes(x = Period2, y = FT, col = Period2)) + 
  theme_bw() +
  geom_boxplot(alpha = 0) +
  scale_y_log10() +
  geom_jitter() + 
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)

g1_HP <- ggplot(pICI_euk_df, aes(x = Period2, y = HP, col = Period2)) + 
  theme_bw() +
  geom_boxplot(alpha = 0) +
  scale_y_log10() +
  geom_jitter() + 
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)
  
  
g1_CC <- ggplot(pICI_bac_df, aes(x = Period2, y = CC, col = Period2)) + 
  theme_bw() +
  geom_boxplot(alpha = 0) +
  scale_y_log10() +
  geom_jitter() + 
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)

g1_HH <- ggplot(pICI_bac_df, aes(x = Period2, y = HH, col = Period2)) + 
  theme_bw() +
  geom_boxplot(alpha = 0) +
  scale_y_log10() +
  geom_jitter() + 
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)


violin_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Boxplot"),
                               Prefix = "CombineAll-MultVar", Suffix = "pdf")
pdf(file = violin_pdf, width = 7, height = 4.5)

plot(g1_NS)
dev.off()







test_HH <- ggplot(pICI_bac_df %>% 
                    filter(Disease == "NSCLC"), 
                  aes(x = Period2, y = HH)) +
  theme_Publication() +
  geom_violin(alpha = 0, aes(col = Period2)) +
  geom_boxplot(width = 0.1, outlier.colour = NA) +
  # scale_y_log10() +
  geom_jitter(aes(col = Period2), alpha = 0.5) +
  stat_compare_means(comparisons = pICI_comparisons) +
  scale_color_manual(values = pcomp_col)+
  theme(legend.position = "none")



```





### scatte forest plot


```{r scatte forest plot}
# melt the Fold Change Wilcox res (cancer type)
# format: Full name     Disease     log2FC
#         cand1         MM          num.
scatt_df <- hm_col_df[, ncol(hm_col_df) - c(4:0)] %>% 
  rownames_to_column(., var = "Fullname") %>%
  melt(., id.vars = "Fullname", value.name = "log2FC", variable.name ="Disease")
# shorten the taxonomy name
scatt_df$Shortname <- gsub("^.+s__", "s__", scatt_df$Fullname)
scatt_df$Shortname <- factor(x = scatt_df$Shortname, 
                             levels = scatt_df %>% 
                               subset(Disease == "All") %>% 
                               select(c(log2FC, Shortname)) %>% 
                               arrange(log2FC) %>% 
                               pull(Shortname))
# add the kingdom
scatt_df$Kingdom <- lapply(scatt_df$Fullname, function(x) {
  k <- strsplit( x = x, split = "\\|") %>% unlist(); return(k[1])
}) %>% unlist()

# set response or non-response
scatt_df$Alter <- ifelse(scatt_df$log2FC > log2(1.5), "R", 
                         ifelse(scatt_df$log2FC <= -log2(1.5), "NR", "no_diff"))

scatt_df_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Kingdom-ScatterPlot"),
                          Prefix = "Scatter-matrix", 
                          Suffix = "csv")

write.csv(x = scatt_df, file = scatt_df_csv, row.names = F)

# sub-routines forest plot
Forest_plot <- function(sel_tax){
  sca_tax <- ggplot(data = scatt_df %>% 
                      subset(Kingdom == sel_tax) %>% 
                      subset(Disease != "All"),
                    aes(x = log2FC, y = Shortname, col = Disease,
                        alpha = Disease, size = Disease)) + 
    geom_point(aes(shape = Disease), fill = tax_col[sel_tax] %>% as.character()) + 
    geom_point(data = scatt_df %>% 
                 subset(Kingdom == sel_tax) %>% 
                 subset(Disease == "All"),
               aes(x = log2FC, y = Shortname, col = Alter,
                        alpha = Disease, size = Alter)) +
    theme_classic() +
    geom_vline(xintercept = c(-log2(1.5), log2(1.5)), linetype="dashed") +
    scale_color_manual(values = c(cancer_col, All = "black", respond_col, no_diff ="gray")) + 
    scale_alpha_manual(values = c(All = 1, MM = .5, NSCLC = .5, RCC = .5, HCC = .5)) +
    scale_size_manual(values = c(R = 4, NR = 4, no_diff = 3, MM = 2, NSCLC = 2, RCC = 2, HCC = 2)) +
    scale_shape_manual(values = c(All = 21, MM = 3, NSCLC = 4, RCC = 5, HCC = 7)) + 
    theme(legend.position = "none")
    
  return(sca_tax)
}

# sub-kingdom
sca_bac <- Forest_plot(sel_tax = "k__Bacteria")
sca_euk <- Forest_plot(sel_tax = "k__Eukaryota")
sca_arc <- Forest_plot(sel_tax = "k__Archaea")


sca_bac_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Kingdom-ScatterPlot"),
                          Prefix = paste0("ScatterPlot-", "Bacteria"), 
                          Suffix = "pdf")
pdf(file = sca_bac_pdf, width = 7, height = (nrow(scatt_df %>% subset(Kingdom == "k__Bacteria" & Disease == "All"))+1)*.3)
plot(sca_bac)
plot(sca_bac + theme(text = element_blank()))
dev.off()
sca_euk_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Kingdom-ScatterPlot"),
                          Prefix = paste0("ScatterPlot-", "Eukaryota"), 
                          Suffix = "pdf")
pdf(file = sca_euk_pdf, width = 7, height = (nrow(scatt_df %>% subset(Kingdom == "k__Eukaryota" & Disease == "All"))+1)*.3)
plot(sca_euk)
plot(sca_euk + theme(text = element_blank()))
dev.off()
sca_arc_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Kingdom-ScatterPlot"),
                          Prefix = paste0("ScatterPlot-", "Archaea"), 
                          Suffix = "pdf")
pdf(file = sca_arc_pdf, width = 7, height = (nrow(scatt_df %>% subset(Kingdom == "k__Archaea" & Disease == "All"))+1)*.3)
plot(sca_arc)
plot(sca_arc + theme(text = element_blank()))
dev.off()







```

### Heatmap (Wilcox MN)

```{r Heatmap Wilcox MN}

sel_cand_df <- label_df[label_df$Group_Num >= 2, ]

sel_cand_df$Kingdom <- lapply(sel_cand_df$Label, function(x) {
  k <- strsplit( x = x, split = "\\|") %>% unlist(); return(k[1])
}) %>% unlist()

sel_cand_df$ShortName <- lapply(sel_cand_df$Label, function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()
min_zero <- 1e-06
for (sub_kingdom in unique(sel_cand_df$Kingdom)) {
  tmp_cand <- sel_cand_df$Label[sel_cand_df$Kingdom == sub_kingdom]
  if (length(tmp_cand) <= 2) {
    next
  }
  short_name <- lapply(tmp_cand, function(x) {
    k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
    }) %>% as.character()
  sel_otu <- merge(bICI_ds_list[[sub_kingdom]]$taxa_abund$Species %>%
                     t() %>%
                     subset(select = tmp_cand) + min_zero,
                   bICI_ds_list[[sub_kingdom]]$sample_table, by = "row.names") %>%
    column_to_rownames("Row.names")
  
  colnames(sel_otu)[1:length(short_name)] <- short_name
  sel_otu[, 1:length(short_name)] <- sel_otu[, 1:length(short_name)] 
  for (sel_Dis in unique(sel_otu$Disease)) {
    if (sel_Dis == "HCC") {
      next
    }
    sel_otu_ <- subset(sel_otu, Disease == sel_Dis) 
    sel_otu_df <- subset(sel_otu_, select = short_name) %>%
       scale() %>% t()
    my_column_split <- sel_otu_$Response
    my_col_fun <- colorRamp2(quantile(sel_otu_df, c(.05, .25, .5, .75, .95)), 
                             c("#00A0B0", "#6A4A3C", "#EDC951", 
                               "#EB6841","#CC333F"))
    enr_dep <- apply(subset(sel_otu_, select = rownames(sel_otu_df)), 2, function(x){
      log2(mean(x[my_column_split == "R"])/mean(x[my_column_split != "R"]))
      }) %>% sort(decreasing = T)
    sel_otu_ <- sel_otu_[, c(names(enr_dep), colnames(bICI_ds_list[[sub_kingdom]]$sample_table))]
    sel_otu_df <- sel_otu_df[names(enr_dep), ]
    my_row_split <- apply(subset(sel_otu_, select = names(enr_dep)), 2, function(x){
      ifelse(mean(x[my_column_split == "R"])/mean(x[my_column_split != "R"]) > 1, "Enriched in R", "Depleted in R")
      }) %>% factor(x = ., levels = c("Enriched in R", "Depleted in R"))
    left_anno <- rowAnnotation(log2_FC = anno_barplot(enr_dep))
    
    my_hm <- Heatmap(sel_otu_df, show_column_names = F,
                     column_split = my_column_split, 
                     cluster_rows = F,
                     row_split = my_row_split,
                     col = my_col_fun, border = T,
                     show_column_dend = F,
                     show_row_dend = F, 
                     width = ncol(sel_otu_df)*unit(1, "mm"), 
                     height = nrow(sel_otu_df)*unit(5, "mm"), 
                     cluster_row_slices = F,
                     left_annotation = left_anno,
                     row_title_rot = ifelse(nrow(sel_otu_df) > 10, 90 , 0),
                     name = "z-score",
                     heatmap_legend_param = list(direction = "horizontal",
                                                 title_position  = "topcenter",
                                                 border = "black"))
    
    
    hm_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Heatmap"),
                         Prefix = paste("HM", sel_Dis, sub_kingdom, "z_score", sep = "-"), Suffix = "pdf")
    pdf(file = hm_pdf, width = ncol(sel_otu_df)/25.4 + 5, height = nrow(sel_otu_df)*5/25.4 + 2)
    draw(my_hm, heatmap_legend_side = "bottom")
    dev.off()
  }
}

### Plot all together
all_otu <- matrix(data = NA, nrow = nrow(bICI_ds_list$all$sample_table), ncol = 0)
rownames(all_otu) <- rownames(bICI_ds_list$all$sample_table)

for (sub_kingdom in unique(sel_cand_df$Kingdom)) {
  tmp_cand <- sel_cand_df$Label[sel_cand_df$Kingdom == sub_kingdom]
  all_otu <- merge(all_otu,
                   bICI_ds_list[[sub_kingdom]]$taxa_abund$Species %>%
                     t() %>%
                     subset(select = tmp_cand),
                   by = "row.names", all = T) %>%
    column_to_rownames("Row.names")
  colnames(all_otu)[(ncol(all_otu) - length(tmp_cand) + 1): ncol(all_otu)] <- sel_cand_df$ShortName[sel_cand_df$Kingdom == sub_kingdom]
}




```

-->


<!--

### Scatter Plot (Regression)

```{r Scatter Regression}
####################### NSCLC #######################
wil_bac_sp_mn_nsclc <- Wil_Matrix(sub_otu = t(mn_bac_df[, bICI_ds_list$k__Bacteria$sample_table$Disease == "NSCLC"]), 
                                  p_cutoff = 0.05, rarefy_or_not = F,
                                  sub_meta = bICI_ds_list$k__Bacteria$sample_table[bICI_ds_list$k__Bacteria$sample_table$Disease == "NSCLC",],
                                  sub_catolog = "Cohort_name", sub_comparison = "Response")

fc_bac_sp_mn_nsclc <- FoldChange(sub_otu = t(mn_bac_df[, bICI_ds_list$k__Bacteria$sample_table$Disease == "NSCLC"]), 
                                 sub_meta = bICI_ds_list$k__Bacteria$sample_table[bICI_ds_list$k__Bacteria$sample_table$Disease == "NSCLC",],
                                 sub_catolog = "Cohort_name", sub_comparison = "Response", comp_rank = c("R", "NR"))


tmp_wil_df <- wil_bac_sp_mn_nsclc$wil_df

NSCLC_wil_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN"),
                            Prefix = "wilcox_MN-NSCLC", Suffix = "csv")
write.csv(x = tmp_wil_df, file = NSCLC_wil_csv)

plot_Routy <- ggscatterstats(data = fc_bac_sp_mn_nsclc, 
                             x = `2018_Routy`,
                             y = All,
                             # type = "spearman",
                             xfill = cohort_col['2018_Routy'],
                             yfill = "gray",
                             marginal.type = "histogram",
                             # point.args = list(col = ifelse(tmp_wil_df$`2018_Routy`<.01|tmp_wil_df$`2022_Derosa`<.01,"red", "black")))
                             point.args = list(col = ifelse(tmp_wil_df$All<.05,"red", "black")))

plot_Derosa <- ggscatterstats(data = fc_bac_sp_mn_nsclc, 
                             x = `2022_Derosa`,
                             y = All,
                             xfill = cohort_col['2022_Derosa'],
                             yfill = "gray",
                             marginal.type = "histogram",
                             point.args = list(col = ifelse(tmp_wil_df$All<.05,"red", "black")))


NSCLC_reg_scatter_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN"),
                                    Prefix = "Regression_NSCLC_scatter", Suffix = "pdf")
pdf(NSCLC_reg_scatter_pdf, width = 6, height = 6)
plot(plot_Routy)
plot(plot_Derosa)
dev.off()


####################### MM #######################
wil_bac_sp_mn_mm <- Wil_Matrix(sub_otu = t(mn_bac_df[, bICI_ds_list$k__Bacteria$sample_table$Disease == "MM"]), 
                               p_cutoff = 0.05, rarefy_or_not = F,
                               sub_meta = bICI_ds_list$k__Bacteria$sample_table[bICI_ds_list$k__Bacteria$sample_table$Disease == "MM",],
                               sub_catolog = "Cohort_name", sub_comparison = "Response")

fc_bac_sp_mn_mm <- FoldChange(sub_otu = t(mn_bac_df[, bICI_ds_list$k__Bacteria$sample_table$Disease == "MM"]), 
                                 sub_meta = bICI_ds_list$k__Bacteria$sample_table[bICI_ds_list$k__Bacteria$sample_table$Disease == "MM",],
                                 sub_catolog = "Cohort_name", sub_comparison = "Response", comp_rank = c("R", "NR"))



tmp_wil_df2 <- wil_bac_sp_mn_mm$wil_df

MM_wil_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN"),
                         Prefix = "wilcox_MN-MM", Suffix = "csv")
write.csv(x = tmp_wil_df2, file = MM_wil_csv)

plot_Frankel <- ggscatterstats(data = fc_bac_sp_mn_mm, 
                               x = `2017_Frankel`,
                               y = All,
                               type = "spearman",
                               xfill = cohort_col['2017_Frankel'],
                               yfill = "gray",
                               marginal.type = "histogram",
                               point.args = list(col = ifelse(tmp_wil_df2$All<.05,"red", "black")))

plot_Gopalakrishnan <- ggscatterstats(data = fc_bac_sp_mn_mm, 
                                      x = `2018_Gopalakrishnan`,
                                      y = All,
                                      type = "spearman",
                                      xfill = cohort_col['2018_Gopalakrishnan'],
                                      yfill = "gray",
                                      marginal.type = "histogram",
                                      point.args = list(col = ifelse(tmp_wil_df2$All<.05,"red", "black")))

plot_Maston <- ggscatterstats(data = fc_bac_sp_mn_mm, 
                              x = `2018_Maston`,
                              y = All,
                              type = "spearman",
                              xfill = cohort_col['2018_Maston'],
                              yfill = "gray",
                              marginal.type = "histogram",
                              point.args = list(col = ifelse(tmp_wil_df2$All<.05,"red", "black")))

plot_Peters <- ggscatterstats(data = fc_bac_sp_mn_mm, 
                              x = `2019_Peters`,
                              y = All,
                              type = "spearman",
                              xfill = cohort_col['2019_Peters'],
                              yfill = "gray",
                              marginal.type = "histogram",
                              point.args = list(col = ifelse(tmp_wil_df2$All<.05,"red", "black")))

plot_Spencer <- ggscatterstats(data = fc_bac_sp_mn_mm, 
                               x = `2021_Spencer`,
                               y = All,
                               type = "spearman",
                               xfill = cohort_col['2021_Spencer'],
                               yfill = "gray",
                               marginal.type = "histogram",
                               point.args = list(col = ifelse(tmp_wil_df2$All<.05,"red", "black")))

plot_Lee <- ggscatterstats(data = fc_bac_sp_mn_mm, 
                           x = `2022_Lee`,
                           y = All,
                           type = "spearman",
                           xfill = cohort_col['2022_Lee'],
                           yfill = "gray",
                           marginal.type = "histogram",
                           point.args = list(col = ifelse(tmp_wil_df2$All<.05,"red", "black")))

plot_McCulloch <- ggscatterstats(data = fc_bac_sp_mn_mm, 
                                 x = `2022_McCulloch`,
                                 y = All,
                                 type = "spearman",
                                 xfill = cohort_col['2022_McCulloch'],
                                 yfill = "gray",
                                 marginal.type = "histogram",
                                 point.args = list(col = ifelse(tmp_wil_df2$All<.05,"red", "black")))

MM_reg_scatter_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN"),
                                 Prefix = "Regression_MM_scatter", Suffix = "pdf")
pdf(MM_reg_scatter_pdf, width = 6, height = 6)
plot(plot_Frankel)
plot(plot_Gopalakrishnan)
plot(plot_Maston)
plot(plot_Peters)
plot(plot_Spencer)
plot(plot_Lee)
plot(plot_McCulloch)
dev.off()
```
-->



<!--
```{r hm combine 2}
calculate_or_not <- T
save_or_not <- T


# filter the candidates, whom appreaded in at least 2 cancer types.
sel_cand_df <- label_df[label_df$Group_Num >= 2, ]
rownames(sel_cand_df) <- sel_cand_df$Label
sel_cand_df$Kingdom <- lapply(sel_cand_df$Label, function(x) {
  k <- strsplit( x = x, split = "\\|") %>% unlist(); return(k[1])
}) %>% unlist()

# shortern the taxonomy names
sel_cand_df$ShortName <- lapply(sel_cand_df$Label, function(x){
  k = strsplit(x, "\\|") %>% unlist; return(k[length(k)])
}) %>% as.character()

# save the selected union candidates.
sel_cand_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN"),
                          Prefix = "Union-Wilcox_MN-Kingdom-ShortName", Suffix = "csv")
write.csv(x = sel_cand_df, file = sel_cand_csv, row.names = F)



sel_mn_bac_df <- bICI_sel_list$k__Bacteria[rownames(bICI_sel_list$k__Bacteria) %in% sel_cand_df$Label, ]

if (calculate_or_not) {
  # selected the union candidates in median normalization abundance matrix
  sel_mn_bac_df <- bICI_sel_list$k__Bacteria[rownames(bICI_sel_list$k__Bacteria) %in% sel_cand_df$Label, ]
  # obtain the meta info
  tmp_meta <- bICI_ds_list$k__Bacteria$sample_table
  # paste the cohort name and cancer type 
  tmp_meta$CohDis <- paste0(tmp_meta$Cohort_name, "_", tmp_meta$Disease)
  # calculate the Wilcox res by new sub-group CohDis with selected candidates
  wil_bac_cohort <- Wil_Matrix(sub_otu = sel_mn_bac_df, p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = tmp_meta, comparison_rank = c("R", "NR"),
                              sub_catolog = "CohDis", sub_comparison = "Response")
  # calculate the Wilcox res by previous group cancer type with selected candidates
  wil_bac_disease <- Wil_Matrix(sub_otu = sel_mn_bac_df, p_cutoff = 0.05, rarefy_or_not = F,
                                sub_meta = bICI_ds_list$k__Bacteria$sample_table,
                                comparison_rank = c("R", "NR"),
                                sub_catolog = "Disease", sub_comparison = "Response")
  # save the rds
  mn_bac_cohort_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Wilcox-MN_Idx-Bacteria-Cohort", Suffix = "rds")
  saveRDS(object = wil_bac_cohort, file = mn_bac_cohort_rds)
  mn_bac_disease_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Wilcox-MN_Idx-Bacteria-disease", Suffix = "rds")
  saveRDS(object = wil_bac_disease, file = mn_bac_disease_rds)
  
  # as same as above (euk instead of bac)
  sel_mn_euk_df <- bICI_sel_list$k__Eukaryota[rownames(bICI_sel_list$k__Eukaryota) %in% sel_cand_df$Label, ]
  tmp_meta <- bICI_ds_list$k__Eukaryota$sample_table
  tmp_meta$CohDis <- paste0(tmp_meta$Cohort_name, "_", tmp_meta$Disease)
  wil_euk_cohort <- Wil_Matrix(sub_otu = sel_mn_euk_df, p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = tmp_meta, comparison_rank = c("R", "NR"),
                              sub_catolog = "CohDis", sub_comparison = "Response")
  wil_euk_disease <- Wil_Matrix(sub_otu = sel_mn_euk_df, p_cutoff = 0.05, rarefy_or_not = F,
                                sub_meta = bICI_ds_list$k__Eukaryota$sample_table,
                                comparison_rank = c("R", "NR"),
                                sub_catolog = "Disease", sub_comparison = "Response")
  mn_euk_cohort_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Wilcox-MN_Idx-Eukaryota-Cohort", Suffix = "rds")
  saveRDS(object = wil_euk_cohort, file = mn_euk_cohort_rds)
  mn_euk_disease_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Wilcox-MN_Idx-Eukaryota-disease", Suffix = "rds")
  saveRDS(object = wil_euk_disease, file = mn_euk_disease_rds)
  
  # as same as above (arc instead of bac)
  sel_mn_arc_df <- bICI_sel_list$k__Archaea[rownames(bICI_sel_list$k__Archaea) %in% sel_cand_df$Label, ]
  tmp_meta <- bICI_ds_list$k__Archaea$sample_table
  tmp_meta$CohDis <- paste0(tmp_meta$Cohort_name, "_", tmp_meta$Disease)
  wil_arc_cohort <- Wil_Matrix(sub_otu = sel_mn_arc_df, p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = tmp_meta, comparison_rank = c("R", "NR"),
                              sub_catolog = "CohDis", sub_comparison = "Response")
  wil_arc_disease <- Wil_Matrix(sub_otu = sel_mn_arc_df, p_cutoff = 0.05, rarefy_or_not = F,
                                sub_meta = bICI_ds_list$k__Archaea$sample_table,
                                comparison_rank = c("R", "NR"),
                                sub_catolog = "Disease", sub_comparison = "Response")
  mn_arc_cohort_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Wilcox-MN_Idx-Archaea-Cohort", Suffix = "rds")
  saveRDS(object = wil_arc_cohort, file = mn_arc_cohort_rds)
  mn_arc_disease_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                                  Prefix = "Wilcox-MN_Idx-Archaea-disease", Suffix = "rds")
  saveRDS(object = wil_arc_disease, file = mn_arc_disease_rds)
  
  
}else{
  wil_bac_cohort <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-02-22-Wilcox-MN_Idx-",
                                         "Bacteria", "-Cohort-v1.0.rds"))
  wil_euk_cohort <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-02-22-Wilcox-MN_Idx-",
                                         "Eukaryota", "-Cohort-v1.0.rds"))
  wil_arc_cohort <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-02-22-Wilcox-MN_Idx-",
                                         "Archaea", "-Cohort-v1.0.rds"))
  
  wil_bac_disease <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-02-22-Wilcox-MN_Idx-",
                                         "Bacteria", "-disease-v1.0.rds"))
  wil_euk_disease <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-02-22-Wilcox-MN_Idx-",
                                         "Eukaryota", "-disease-v1.0.rds"))
  wil_arc_disease <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2023-02-22-Wilcox-MN_Idx-",
                                         "Archaea", "-disease-v1.0.rds"))
}

hm_col_df <- log2(do.call("rbind",
                          list(cbind(wil_bac_cohort$FC_df[, -ncol(wil_bac_cohort$FC_df)], wil_bac_disease$FC_df),
                               cbind(wil_euk_cohort$FC_df[, -ncol(wil_euk_cohort$FC_df)], wil_euk_disease$FC_df),
                               cbind(wil_arc_cohort$FC_df[, -ncol(wil_arc_cohort$FC_df)], wil_arc_disease$FC_df))))
hm_col_df[hm_col_df >5] <- 5
hm_col_df[hm_col_df < -5] <- -5
tmp_lab_df <- do.call("rbind",
                      list(cbind(wil_bac_cohort$wil_df[, -ncol(wil_bac_cohort$wil_df)], wil_bac_disease$wil_df),
                           cbind(wil_euk_cohort$wil_df[, -ncol(wil_euk_cohort$wil_df)], wil_euk_disease$wil_df),
                           cbind(wil_arc_cohort$wil_df[, -ncol(wil_arc_cohort$wil_df)], wil_arc_disease$wil_df)))
hm_lab_df <- ifelse(tmp_lab_df > .05,
                    "",
                    ifelse(tmp_lab_df >.01, "*",
                           ifelse(tmp_lab_df >.01, "**", "***")))


col_fun = colorRamp2(c(-5, 0, 5), c("#ed0000cc", "#ffffff", "#00468bcc"))



top_anno <- HeatmapAnnotation(Disease = colnames(hm_col_df) %>% gsub("^.+_", "", .), 
                              col = list(Disease = c(cancer_col, All = "black")),
                              gp = gpar(col = 'black'))
right_anno <- rowAnnotation(Kingdom = sel_cand_df[rownames(hm_col_df), ]$Kingdom,
                            col = list(Kingdom = c(tax_col)), 
                            gp = gpar(col = 'black'))

left_col_fun <- function(x){
  ifelse(x == 0, "white", ifelse(x == 1, "#84bd007f", 
                                 ifelse(x == 2, "#8ca252cc",
                                        ifelse(x == 3, "#637939cc", "#20854ecc"))))
}

left_anno <- rowAnnotation(Num = anno_barplot(rowSums(hm_lab_df[, 1:11] != ""),
                                              axis_param = list(direction = "reverse"),
                                              gp = gpar(fill = do.call(left_col_fun, 
                                                                       list(rowSums(hm_lab_df[, 1:11] != "")))),
                                              width = unit(1.5, "cm")))


comb_hm <- Heatmap(hm_col_df %>% as.matrix(),
                   name = "log2(R/NR)",
                   width = unit(ncol(hm_col_df)/2, "cm"),
                   height = unit(nrow(hm_col_df)/2, "cm"),
                   column_split = factor(c(rep("Cohort", ncol(wil_bac_cohort$FC_df)-1), 
                                           rep("Disease", ncol(wil_bac_disease$FC_df)-1),
                                           "All"), levels = c("Cohort", "Disease", "All")),
                   cluster_column_slices = F,
                   cell_fun = function(j, i, x, y, width, height, fill) {
                     grid.text(hm_lab_df[i, j], x, y, gp = gpar(fontsize = 10))},
                   row_labels = sel_cand_df[rownames(hm_col_df) ,'ShortName'],
                   col = col_fun, top_annotation = top_anno,
                   right_annotation = right_anno, left_annotation = left_anno,
                   border = "grey", heatmap_legend_param = list(direction = "horizontal")) 

if (save_or_not) {
  tmp_path <- getwd()
  setwd(paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Comb-Heatmap"))
  
  comb_hm_pdf <- FileCreate(DirPath = './',
                            Prefix = "Heatmap-combine-cohort-disease-foldchange", Suffix = "pdf")
  pdf(file = comb_hm_pdf, width = ncol(hm_col_df)/25.4 + 15, height = nrow(hm_col_df)*5/25.4 + 5)
  draw(comb_hm, merge_legend = TRUE, 
       heatmap_legend_side = "left",
       annotation_legend_side = "left")
  dev.off()
  setwd(tmp_path)
}


```
-->

<!--
### Wilcoxon test (median normalization)

```{r wilcoxon MN}

calculate_or_not <- F

if (calculate_or_not) {

  
  # bacteria
  mn_bac_ls <- MedianNormal(raw_df = bICI_ds_list$k__Bacteria$taxa_abund$Species, 
                            meta_df = bICI_ds_list$k__Bacteria$sample_table, 
                            min_add = 0.0001, rarefy_cutoff = 0.01)
  
  mn_bac_df <- mn_bac_ls$nm_df
  mn_bac_csv <- FileCreate(DirPath = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/"),
                           Prefix = "Med_Normal-Bacteria", Suffix = "csv")
  write.csv(x = mn_bac_df, file = mn_bac_csv)
  
  mn_bac_index <- mn_bac_ls$mn_index_df
  mn_bac_idx_csv <- FileCreate(DirPath = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/"),
                               Prefix = "Med_Normal-Bacteria-index", Suffix = "csv")
  write.csv(x = mn_bac_index, file = mn_bac_idx_csv)
  
  # eukaryota
  mn_euk_ls <- MedianNormal(raw_df = bICI_ds_list$k__Eukaryota$taxa_abund$Species, 
                            meta_df = bICI_ds_list$k__Eukaryota$sample_table,
                            min_add = 0.0001, rarefy_cutoff = 0.01)
  
  mn_euk_df <- mn_euk_ls$nm_df
  mn_euk_csv <- FileCreate(DirPath = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/"),
                           Prefix = "Med_Normal-Eukaryota", Suffix = "csv")
  write.csv(x = mn_euk_df, file = mn_euk_csv)
  
  mn_euk_index <- mn_euk_ls$mn_index_df
  mn_euk_idx_csv <- FileCreate(DirPath = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/"),
                               Prefix = "Med_Normal-Eukaryota-index", Suffix = "csv")
  write.csv(x = mn_euk_index, file = mn_euk_idx_csv)
  
  # archaea
  mn_arc_ls <- MedianNormal(raw_df = bICI_ds_list$k__Archaea$taxa_abund$Species, 
                            meta_df = bICI_ds_list$k__Archaea$sample_table,
                            min_add = 0.0001, rarefy_cutoff = 0.01)
  
  mn_arc_df <- mn_arc_ls$nm_df
  mn_arc_csv <- FileCreate(DirPath = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/"),
                           Prefix = "Med_Normal-Archaea", Suffix = "csv")
  write.csv(x = mn_arc_df, file = mn_arc_csv)
  
  mn_arc_index <- mn_arc_ls$mn_index_df
  mn_arc_idx_csv <- FileCreate(DirPath = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/"),
                               Prefix = "Med_Normal-Archaea-index", Suffix = "csv")
  write.csv(x = mn_arc_index, file = mn_arc_idx_csv)
  
  mn_vir_ls <- MedianNormal(raw_df = bICI_ds_list$k__Viruses$taxa_abund$Species, 
                            meta_df = bICI_ds_list$k__Viruses$sample_table,
                            min_add = 0.0001, rarefy_cutoff = 0.01)
  
  mn_vir_df <- mn_vir_ls$nm_df
  mn_vir_csv <- FileCreate(DirPath = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/"),
                           Prefix = "Med_Normal-Viruses", Suffix = "csv")
  write.csv(x = mn_vir_df, file = mn_vir_csv)
  
  mn_vir_index <- mn_vir_ls$mn_index_df
  mn_vir_idx_csv <- FileCreate(DirPath = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/"),
                               Prefix = "Med_Normal-Viruses-index", Suffix = "csv")
  write.csv(x = mn_vir_index, file = mn_vir_idx_csv)
  
  
}else{
  mn_bac_df <- read.csv(file = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/2022-10-20-Med_Normal-",
                                      "Bacteria", "-v1.0.csv"), header = T, row.names = 1, check.names = F)
  mn_euk_df <- read.csv(file = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/2022-10-20-Med_Normal-",
                                      "Eukaryota", "-v1.0.csv"), header = T, row.names = 1, check.names = F)
  mn_arc_df <- read.csv(file = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/2022-10-20-Med_Normal-",
                                      "Archaea", "-v1.0.csv"), header = T, row.names = 1, check.names = F)
  mn_vir_df <- read.csv(file = paste0(save_path_prefix, "01.PreProcessing/MedianNormalization/2022-10-20-Med_Normal-",
                                      "Viruses", "-v1.0.csv"), header = T, row.names = 1, check.names = F)
}


if (calculate_or_not) {
  wil_bac_sp_mn <- Wil_Matrix(sub_otu = t(mn_bac_df), p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = bICI_ds_list$k__Bacteria$sample_table, comparison_rank = c("R", "NR"),
                              sub_catolog = "Disease", sub_comparison = "Response")
  mn_bac_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                           Prefix = "Wilcox-MN-Bacteria", Suffix = "rds")
  saveRDS(object = wil_bac_sp_mn, file = mn_bac_rds)
  
  wil_euk_sp_mn <- Wil_Matrix(sub_otu = t(mn_euk_df), p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = bICI_ds_list$k__Eukaryota$sample_table,comparison_rank = c("R", "NR"),
                              sub_catolog = "Disease", sub_comparison = "Response")
  mn_euk_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                           Prefix = "Wilcox-MN-Eukaryota", Suffix = "rds")
  saveRDS(object = wil_euk_sp_mn, file = mn_euk_rds)
  
  wil_arc_sp_mn <- Wil_Matrix(sub_otu = t(mn_arc_df), p_cutoff = 0.05, rarefy_or_not = F,
                           sub_meta = bICI_ds_list$k__Archaea$sample_table,comparison_rank = c("R", "NR"),
                           sub_catolog = "Disease", sub_comparison = "Response")
  mn_arc_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                           Prefix = "Wilcox-MN-Archaea", Suffix = "rds")
  saveRDS(object = wil_arc_sp_mn, file = mn_arc_rds)
  
  wil_vir_sp_mn <- Wil_Matrix(sub_otu = t(mn_vir_df), p_cutoff = 0.05, rarefy_or_not = F,
                              sub_meta = bICI_ds_list$k__Viruses$sample_table,comparison_rank = c("R", "NR"),
                              sub_catolog = "Disease", sub_comparison = "Response")
  mn_vir_rds <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds"),
                           Prefix = "Wilcox-MN-Viruses", Suffix = "rds")
  saveRDS(object = wil_vir_sp_mn, file = mn_vir_rds)
}else{
  wil_bac_sp_mn <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2022-11-23-Wilcox-MN-",
                                         "Bacteria", "-v1.0.rds"))
  wil_euk_sp_mn <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2022-11-23-Wilcox-MN-",
                                         "Eukaryota", "-v1.0.rds"))
  wil_arc_sp_mn <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2022-11-23-Wilcox-MN-",
                                         "Archaea", "-v1.0.rds"))
  wil_vir_sp_mn <- readRDS(file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/rds/2022-11-23-Wilcox-MN-",
                                         "Viruses", "-v1.0.rds"))
}

nrow(wil_bac_sp_mn$sel_wil_df) # 318
nrow(wil_euk_sp_mn$sel_wil_df) # 273
nrow(wil_arc_sp_mn$sel_wil_df) # 71
nrow(wil_vir_sp_mn$sel_wil_df) # 37


sum(rowSums(wil_bac_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 27
sum(rowSums(wil_euk_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 21
sum(rowSums(wil_arc_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 6
sum(rowSums(wil_vir_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 2

sum(rowSums(wil_bac_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 0
sum(rowSums(wil_euk_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 0
sum(rowSums(wil_arc_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 0
sum(rowSums(wil_vir_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 0

melt_wil_mn_list <- list()

melt_wil_mn_list$k__Bacteria <- melt(as.matrix(wil_bac_sp_mn$wil_sign_df[, 1:4]),
                                  varnames = c("Feature", "CancerType"), 
                                  value.name = "Select") %>% subset(Select == "sign")
melt_wil_mn_list$k__Eukaryota <- melt(as.matrix(wil_euk_sp_mn$wil_sign_df[, 1:4]),
                                   varnames = c("Feature", "CancerType"), 
                                   value.name = "Select") %>% subset(Select == "sign")
melt_wil_mn_list$k__Archaea <- melt(as.matrix(wil_arc_sp_mn$wil_sign_df[, 1:4]),
                                 varnames = c("Feature", "CancerType"), 
                                 value.name = "Select") %>% subset(Select == "sign")
melt_wil_mn_list$k__Viruses <- melt(as.matrix(wil_vir_sp_mn$wil_sign_df[, 1:4]),
                                 varnames = c("Feature", "CancerType"), 
                                 value.name = "Select") %>% subset(Select == "sign")

melt_wil_cohort_list <- list()

melt_wil_mn_list$k__Bacteria <- melt(as.matrix(wil_bac_sp_mn$wil_sign_df[, 1:4]),
                                  varnames = c("Feature", "CancerType"), 
                                  value.name = "Select") %>% subset(Select == "sign")
melt_wil_mn_list$k__Eukaryota <- melt(as.matrix(wil_euk_sp_mn$wil_sign_df[, 1:4]),
                                   varnames = c("Feature", "CancerType"), 
                                   value.name = "Select") %>% subset(Select == "sign")
melt_wil_mn_list$k__Archaea <- melt(as.matrix(wil_arc_sp_mn$wil_sign_df[, 1:4]),
                                 varnames = c("Feature", "CancerType"), 
                                 value.name = "Select") %>% subset(Select == "sign")
melt_wil_mn_list$k__Viruses <- melt(as.matrix(wil_vir_sp_mn$wil_sign_df[, 1:4]),
                                 varnames = c("Feature", "CancerType"), 
                                 value.name = "Select") %>% subset(Select == "sign")
if (F) { # FOR FENFEN
  tmp_df <- do.call("rbind",
                    list(wil_bac_sp_mn$sel_wil_df,
                         wil_euk_sp_mn$sel_wil_df,
                         wil_arc_sp_mn$sel_wil_df,
                         wil_vir_sp_mn$sel_wil_df))
  write.csv(x = tmp_df, file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Fenfen-CancerType-LYF.csv"), row.names = F)
  tmp_melt <- tmp_df %>% rownames_to_column("Candidates")  %>% melt(., id.vars = "Candidates")
  write.csv(x = tmp_melt, file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Fenfen-CancerType_melt-LYF.csv"), row.names = F)
  tmp_df <- do.call("rbind",
                    list(wil_bac_sp_mn$FC_df,
                         wil_euk_sp_mn$FC_df,
                         wil_arc_sp_mn$FC_df,
                         wil_vir_sp_mn$FC_df))
  write.csv(x = tmp_df, file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Fenfen-CancerType-FC-LYF.csv"), row.names = F)
  tmp_melt <- tmp_df %>% rownames_to_column("Candidates")  %>% melt(., id.vars = "Candidates")
  write.csv(x = tmp_melt, file = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Fenfen-CancerType_melt-FC-LYF.csv"), row.names = F)
}
```
-->



<!--
### test multivariate analysis 1

[link](https://cran.r-project.org/web/packages/survivalAnalysis/vignettes/multivariate.html)

```{r multivariate analysis test 1}

require(tidyverse)
require(tidytidbits)
require(survivalAnalysis)

covariate_names <- c(age="Age at Dx",
                     sex="Sex",
                     ph.ecog="ECOG Status",
                     wt.loss="Weight Loss (6 mo.)",
                     `sex:female`="Female",
                     `ph.ecog:0`="ECOG 0",
                     `ph.ecog:1`="ECOG 1",
                     `ph.ecog:2`="ECOG 2",
                     `ph.ecog:3`="ECOG 3")

survival::lung %>%
  mutate(sex=rename_factor(sex, `1` = "male", `2` = "female")) %>%
  analyse_multivariate(vars(time, status),
                       covariates = vars(age, sex, ph.ecog, wt.loss),
                       covariate_name_dict = covariate_names) ->
  result


forest_plot(result)

df <- survival::lung %>% mutate(sex=rename_factor(sex, `1` = "male", `2` = "female"))

map(vars(age, sex, ph.ecog, wt.loss), function(by)
{
  analyse_multivariate(df,
                       vars(time, status),
                       covariates = list(by), # covariates expects a list
                       covariate_name_dict = covariate_names)
}) %>%
  forest_plot(factor_labeller = covariate_names,
              endpoint_labeller = c(time="OS"),
              orderer = ~order(HR),
              labels_displayed = c("endpoint", "factor", "n"),
              ggtheme = ggplot2::theme_bw(base_size = 10))



```


### test multivariate analysis 2

[Univariate and Multivariate Survival Analysis](http://chenyuan.date/2017/08/Univariate-and-Multivariate-Survival-Analysis/)

```{r multivariate analysis test 2}
require(survival)
require(tidytidbits)
require(tidyverse)
require(survivalAnalysis)

### Univariate Survival Analysis

survival::lung %>% 
  count_by(ph.ecog)

# survival analysis comparing the subgroups formed by the ECOG status
survival::lung %>%
  mutate(ecog=recode_factor(ph.ecog, `0`="0", `1`="1", `2`="2-3", `3`="2-3")) %>% #group 2 und 3 together
  analyse_survival(vars(time, status), by=ecog) -> result
kaplan_meier_plot(result)

### Multivariate Survival Analysis

covariate_names <- c(age="Age at Dx", #provide readable labels for the covariates to allow easy interpretation.
                     sex="Sex",
                     ph.ecog="ECOG Status",
                     wt.loss="Weight Loss (6 mo.)",
                     `sex:female`="Female",
                     `ph.ecog:0`="ECOG 0",
                     `ph.ecog:1`="ECOG 1",
                     `ph.ecog:2`="ECOG 2",
                     `ph.ecog:3`="ECOG 3")

survival::lung %>%
  mutate(sex=rename_factor(sex, `1` = "male", `2` = "female")) %>% #make sex a factor.
  analyse_multivariate(vars(time, status),
                       covariates = vars(age, sex, ph.ecog, wt.loss),
                       covariate_name_dict = covariate_names) -> result #store the result object.
forest_plot(result,
            factor_labeller = covariate_names, #label the subgroups.
            endpoint_labeller = c(time="OS"), #label the endpoint.
            orderer = ~order(HR), #order by hazard ratio
            labels_displayed = c("endpoint", "factor", "n"),
            ggtheme = ggplot2::theme_bw(base_size = 10), #Adjust font size
            relative_widths = c(1, 1.5, 1), #more space for the plot, less space for the tables
            HR_x_breaks = c(0.25, 0.5, 0.75, 1, 1.5, 2))
```


### scatte forest plot (cancer type)


```{r scatte forest plot cancer type}

### MM ###

mm_meta <- bICI_ds_list$k__Bacteria$sample_table %>% subset(Disease == "MM")
mm_bac_Cohort <- Wil_Matrix(sub_otu = t(mn_bac_df[rownames(mm_meta), ]), p_cutoff = 0.05, rarefy_or_not = F,
                            sub_meta = mm_meta, comparison_rank = c("R", "NR"),
                            sub_catolog = "Cohort_name", sub_comparison = "Response")
mm_meta2 <- bICI_ds_list$k__Eukaryota$sample_table %>% subset(Disease == "MM")
mm_euk_Cohort <- Wil_Matrix(sub_otu = t(mn_euk_df[rownames(mm_meta2), ]), p_cutoff = 0.05, rarefy_or_not = F,
                            sub_meta = mm_meta2, comparison_rank = c("R", "NR"),
                            sub_catolog = "Cohort_name", sub_comparison = "Response")
sel_mm_cand <- c(rownames(mm_bac_Cohort$wil_df)[rowSums(mm_bac_Cohort$wil_df[, -ncol(mm_bac_Cohort$wil_df)] <.05) >= 3],
                 rownames(mm_euk_Cohort$wil_df)[rowSums(mm_euk_Cohort$wil_df[, -ncol(mm_euk_Cohort$wil_df)] <.05) >= 3])
sel_mm_FC_df <- rbind(mm_bac_Cohort$FC_df[rownames(mm_bac_Cohort$wil_df)[rowSums(mm_bac_Cohort$wil_df[, -ncol(mm_bac_Cohort$wil_df)] <.05) >= 3], ],
                      mm_euk_Cohort$FC_df[rownames(mm_euk_Cohort$wil_df)[rowSums(mm_euk_Cohort$wil_df[, -ncol(mm_euk_Cohort$wil_df)] <.05) >= 3], ]) %>%
  log2()

sel_mm_FC_df[sel_mm_FC_df > 5] <- 5
sel_mm_FC_df[sel_mm_FC_df < -5] <- -5

melt_mm_FC_df <- sel_mm_FC_df %>%
  rownames_to_column(., var = "Fullname") %>%
  melt(., id.vars = "Fullname", value.name = "log2FC", variable.name ="Cohort") %>%
  add_column(Shortname = gsub("^.+s__", "s__", .[["Fullname"]]),
             Kingdom = lapply( .[["Fullname"]], function(x) {
               k <- strsplit( x = x, split = "\\|") %>% unlist(); return(k[1])
               }) %>% unlist(),
             Alter = ifelse(.[["log2FC"]] > log2(1.5), "R", 
                         ifelse(.[["log2FC"]] <= -log2(1.5), "NR", "no_diff")))
melt_mm_FC_df$Shortname <- factor(x = melt_mm_FC_df$Shortname,
                                  levels = melt_mm_FC_df %>% 
                                    subset(Cohort == "All") %>% 
                                    select(c(log2FC, Shortname)) %>% 
                                    arrange(log2FC) %>% 
                                    pull(Shortname))

melt_mm_FC_plot <- ggplot(data = melt_mm_FC_df %>% 
                            subset(Cohort != "All"),
                          aes(x = log2FC, y = Shortname, col = Cohort,
                      alpha = Cohort, size = Cohort)) + 
  geom_point(aes(shape = Cohort), fill = tax_col[sel_tax] %>% as.character()) + 
  geom_point(data = melt_mm_FC_df %>% 
               subset(Cohort == "All"),
             aes(x = log2FC, y = Shortname, col = Alter,
                      alpha = Cohort, size = Alter)) +
  theme_classic() +
  geom_vline(xintercept = c(-log2(1.5), log2(1.5)), linetype="dashed") +
  scale_color_manual(values = c(cohort_col, All = "black", respond_col, no_diff ="gray")) + 
  scale_alpha_manual(values = c(All = 1, `2017_Frankel` = .5, `2018_Gopalakrishnan` = .5, 
                                `2018_Maston` = .5, `2019_Peters` = .5, `2021_Spencer` = .5,
                                `2022_Lee` =.5, `2022_McCulloch` = .5)) +
  scale_size_manual(values = c(R = 4, NR = 4, no_diff = 3, 
                               `2017_Frankel` = 2, `2018_Gopalakrishnan` = 2, 
                               `2018_Maston` = 2, `2019_Peters` = 2, `2021_Spencer` = 2,
                               `2022_Lee` =2, `2022_McCulloch` = 2)) +
  scale_shape_manual(values = c(All = 21, `2017_Frankel` = 3, `2018_Gopalakrishnan` = 4, 
                                `2018_Maston` = 5, `2019_Peters` = 7, `2021_Spencer` = 8,
                                `2022_Lee` =9, `2022_McCulloch` = 10)) +
  theme(legend.position = "none")


### NSCLC ###

nsclc_meta <- bICI_ds_list$k__Bacteria$sample_table %>% subset(Disease == "NSCLC")
nsclc_bac_Cohort <- Wil_Matrix(sub_otu = t(mn_bac_df[rownames(nsclc_meta), ]), p_cutoff = 0.05, rarefy_or_not = F,
                               sub_meta = nsclc_meta, comparison_rank = c("R", "NR"),
                               sub_catolog = "Cohort_name", sub_comparison = "Response")


nsclc_meta2 <- bICI_ds_list$k__Eukaryota$sample_table %>% subset(Disease == "NSCLC")
nsclc_euk_Cohort <- Wil_Matrix(sub_otu = t(mn_euk_df[rownames(nsclc_meta2), ]), p_cutoff = 0.05, rarefy_or_not = F,
                            sub_meta = nsclc_meta2, comparison_rank = c("R", "NR"),
                            sub_catolog = "Cohort_name", sub_comparison = "Response")
sel_nsclc_cand <- c(rownames(nsclc_bac_Cohort$wil_df)[rowSums(nsclc_bac_Cohort$wil_df[, -ncol(nsclc_bac_Cohort$wil_df)] <.05) >= 2],
                 rownames(nsclc_euk_Cohort$wil_df)[rowSums(nsclc_euk_Cohort$wil_df[, -ncol(nsclc_euk_Cohort$wil_df)] <.05) >= 2])
sel_nsclc_FC_df <- rbind(nsclc_bac_Cohort$FC_df[rownames(nsclc_bac_Cohort$wil_df)[rowSums(nsclc_bac_Cohort$wil_df[, -ncol(nsclc_bac_Cohort$wil_df)] <.05) >= 2], ],
                      nsclc_euk_Cohort$FC_df[rownames(nsclc_euk_Cohort$wil_df)[rowSums(nsclc_euk_Cohort$wil_df[, -ncol(nsclc_euk_Cohort$wil_df)] <.05) >= 2], ]) %>%
  log2()

sel_nsclc_FC_df[sel_nsclc_FC_df > 5] <- 5
sel_nsclc_FC_df[sel_nsclc_FC_df < -5] <- -5

melt_nsclc_FC_df <- sel_nsclc_FC_df %>%
  rownames_to_column(., var = "Fullname") %>%
  melt(., id.vars = "Fullname", value.name = "log2FC", variable.name ="Cohort") %>%
  add_column(Shortname = gsub("^.+s__", "s__", .[["Fullname"]]),
             Kingdom = lapply( .[["Fullname"]], function(x) {
               k <- strsplit( x = x, split = "\\|") %>% unlist(); return(k[1])
               }) %>% unlist(),
             Alter = ifelse(.[["log2FC"]] > log2(1.5), "R", 
                         ifelse(.[["log2FC"]] <= -log2(1.5), "NR", "no_diff")))
melt_nsclc_FC_df$Shortname <- factor(x = melt_nsclc_FC_df$Shortname,
                                  levels = melt_nsclc_FC_df %>% 
                                    subset(Cohort == "All") %>% 
                                    select(c(log2FC, Shortname)) %>% 
                                    arrange(log2FC) %>% 
                                    pull(Shortname))

melt_nsclc_FC_plot <- ggplot(data = melt_nsclc_FC_df %>% 
                            subset(Cohort != "All"),
                          aes(x = log2FC, y = Shortname, col = Cohort,
                      alpha = Cohort, size = Cohort)) + 
  geom_point(aes(shape = Cohort), fill = tax_col[sel_tax] %>% as.character()) + 
  geom_point(data = melt_nsclc_FC_df %>% 
               subset(Cohort == "All"),
             aes(x = log2FC, y = Shortname, col = Alter,
                      alpha = Cohort, size = Alter)) +
  theme_classic() +
  geom_vline(xintercept = c(-log2(1.5), log2(1.5)), linetype="dashed") +
  scale_color_manual(values = c(cohort_col, All = "black", respond_col, no_diff ="gray")) + 
  scale_alpha_manual(values = c(All = 1, `2018_Routy` = .5, `2022_Derosa` = .5)) +
  scale_size_manual(values = c(R = 4, NR = 4, no_diff = 3, 
                               `2018_Routy` = 2, `2022_Derosa` = 2)) +
  scale_shape_manual(values = c(All = 21, `2018_Routy` = 3, `2022_Derosa` = 4)) +
  theme(legend.position = "none")





melt_nsclc_FC_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Cohort-ScatterPlot"),
                          Prefix = paste0("ScatterPlot-", "NSCLC"), 
                          Suffix = "pdf")
pdf(file = melt_nsclc_FC_pdf, width = 7, height = (nrow(melt_nsclc_FC_df %>% subset( Cohort == "All"))+1)*.3)
plot(melt_nsclc_FC_plot)
plot(melt_nsclc_FC_plot + theme(text = element_blank()))
dev.off()

melt_mm_FC_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Cohort-ScatterPlot"),
                          Prefix = paste0("ScatterPlot-", "MM"), 
                          Suffix = "pdf")
pdf(file = melt_mm_FC_pdf, width = 7, height = (nrow(melt_mm_FC_df %>% subset( Cohort == "All"))+1)*.3)
plot(melt_mm_FC_plot)
plot(melt_mm_FC_plot + theme(text = element_blank()))
dev.off()



### RCC ###

rcc_meta <- bICI_ds_list$k__Bacteria$sample_table %>% subset(Disease == "RCC")
rcc_bac_Cohort <- Wil_Matrix(sub_otu = t(mn_bac_df[rownames(rcc_meta), ]), p_cutoff = 0.05, rarefy_or_not = F,
                             sub_meta = rcc_meta, comparison_rank = c("R", "NR"),
                             sub_catolog = "Cohort_name", sub_comparison = "Response")
rcc_euk_Cohort <- Wil_Matrix(sub_otu = t(mn_euk_df[rownames(rcc_meta), ]), p_cutoff = 0.05, rarefy_or_not = F,
                             sub_meta = rcc_meta, comparison_rank = c("R", "NR"),
                             sub_catolog = "Cohort_name", sub_comparison = "Response")
rcc_arc_Cohort <- Wil_Matrix(sub_otu = t(mn_arc_df[rownames(rcc_meta), ]), p_cutoff = 0.05, rarefy_or_not = F,
                             sub_meta = rcc_meta, comparison_rank = c("R", "NR"),
                             sub_catolog = "Cohort_name", sub_comparison = "Response")
rcc_vir_Cohort <- Wil_Matrix(sub_otu = t(mn_vir_df[rownames(rcc_meta), ]), p_cutoff = 0.05, rarefy_or_not = F,
                             sub_meta = rcc_meta, comparison_rank = c("R", "NR"),
                             sub_catolog = "Cohort_name", sub_comparison = "Response")

rcc_vol_df <- do.call('rbind',
                      list(data.frame(pvalue = rcc_bac_Cohort$wil_df$`2018_Routy`, 
                                      sign = ifelse(rcc_bac_Cohort$wil_sign_df$`2018_Routy` == 'sign', "k__Bacteria", "ns"),
                                      log2FC = log2(rcc_bac_Cohort$FC_df$`2018_Routy`),
                                      ShortName = gsub("^.+\\|", "", rownames(rcc_bac_Cohort$FC_df))),
                           data.frame(pvalue = rcc_euk_Cohort$wil_df$`2018_Routy`, 
                                      sign = ifelse(rcc_euk_Cohort$wil_sign_df$`2018_Routy` == 'sign', "k__Eukaryota", "ns"),
                                      log2FC = log2(rcc_euk_Cohort$FC_df$`2018_Routy`),
                                      ShortName = gsub("^.+\\|", "", rownames(rcc_euk_Cohort$FC_df))),
                           data.frame(pvalue = rcc_arc_Cohort$wil_df$`2018_Routy`, 
                                      sign = ifelse(rcc_arc_Cohort$wil_sign_df$`2018_Routy` == 'sign', "k__Archaea", "ns"),
                                      log2FC = log2(rcc_arc_Cohort$FC_df$`2018_Routy`),
                                      ShortName = gsub("^.+\\|", "", rownames(rcc_arc_Cohort$FC_df))),
                           data.frame(pvalue = rcc_vir_Cohort$wil_df$`2018_Routy`, 
                                      sign = ifelse(rcc_vir_Cohort$wil_sign_df$`2018_Routy` == 'sign', "k__Viruses", "ns"),
                                      log2FC = log2(rcc_vir_Cohort$FC_df$`2018_Routy`),
                                      ShortName = gsub("^.+\\|", "", rownames(rcc_vir_Cohort$FC_df))))) %>%
  as.data.frame()

rcc_vol_df$group <- ifelse(rcc_vol_df$log2FC %>% abs() >log2(1.5), rcc_vol_df$sign, "ns")
rcc_vol_df$EnrDep <- ifelse(rcc_vol_df$group == "ns", "ns", ifelse(rcc_vol_df$log2FC >0, "R", "NR"))
rcc_vol_df$NlogP <- -log10(rcc_vol_df$pvalue)

rcc_vol_plot <- rcc_vol_df %>%
  ggplot(., aes(x = log2FC, y = NlogP)) + 
  geom_point(aes(alpha = group, fill = group, size = group), shape = 21, col = "gray")+
  scale_alpha_manual(values = c(k__Bacteria = 1, k__Eukaryota = 1, k__Viruses = 1, ns = .1))+
  scale_fill_manual(values = c(tax_col, ns = "black")) + 
  scale_size_manual(values = c(k__Bacteria = 2, k__Eukaryota = 2, k__Viruses = 2, ns = 1)) +
  geom_vline(xintercept = c(-1, 1) * log2(1.5), linetype = 2, color = respond_col) +
  geom_hline(yintercept = -log10(.05), linetype = 2) + 
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme_classic() +
  theme(legend.position = 'none')


rcc_vol_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Cohort-ScatterPlot"),
                          Prefix = paste0("VolcanoPlot-", "RCC"), 
                          Suffix = "csv")
write.csv(x = rcc_vol_df, file = rcc_vol_csv)


rcc_vol_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Cohort-ScatterPlot"),
                          Prefix = paste0("VolcanoPlot-", "RCC"), 
                          Suffix = "pdf")
pdf(file = rcc_vol_pdf, width = 10, height = 7)
plot(rcc_vol_plot)
plot(rcc_vol_plot + theme(text = element_blank()))
plot(rcc_vol_plot + theme(text = element_blank()) + 
       geom_text_repel(data = rcc_vol_df %>% subset(group != "ns"), 
                       aes(label = ShortName),  size = 1,
                       box.padding = 0.5, max.overlaps = Inf))
dev.off()


### HCC ###

hcc_meta <- bICI_ds_list$k__Bacteria$sample_table %>% subset(Disease == "HCC")
hcc_bac_Cohort <- Wil_Matrix(sub_otu = t(mn_bac_df[rownames(hcc_meta), ]), p_cutoff = 0.05, rarefy_or_not = F,
                             sub_meta = hcc_meta, comparison_rank = c("R", "NR"),
                             sub_catolog = "Cohort_name", sub_comparison = "Response")
hcc_euk_Cohort <- Wil_Matrix(sub_otu = t(mn_euk_df[rownames(hcc_meta), ]), p_cutoff = 0.05, rarefy_or_not = F,
                             sub_meta = hcc_meta, comparison_rank = c("R", "NR"),
                             sub_catolog = "Cohort_name", sub_comparison = "Response")
hcc_arc_Cohort <- Wil_Matrix(sub_otu = t(mn_arc_df[rownames(hcc_meta), ]), p_cutoff = 0.05, rarefy_or_not = F,
                             sub_meta = hcc_meta, comparison_rank = c("R", "NR"),
                             sub_catolog = "Cohort_name", sub_comparison = "Response")
hcc_vir_Cohort <- Wil_Matrix(sub_otu = t(mn_vir_df[rownames(hcc_meta), ]), p_cutoff = 0.05, rarefy_or_not = F,
                             sub_meta = hcc_meta, comparison_rank = c("R", "NR"),
                             sub_catolog = "Cohort_name", sub_comparison = "Response")

hcc_vol_df <- do.call('rbind',
                      list(data.frame(pvalue = hcc_bac_Cohort$wil_df$`2019_Zheng`, 
                                      sign = ifelse(hcc_bac_Cohort$wil_sign_df$`2019_Zheng` == 'sign', "k__Bacteria", "ns"),
                                      log2FC = log2(hcc_bac_Cohort$FC_df$`2019_Zheng`),
                                      ShortName = gsub("^.+\\|", "", rownames(hcc_bac_Cohort$FC_df))),
                           data.frame(pvalue = hcc_euk_Cohort$wil_df$`2019_Zheng`, 
                                      sign = ifelse(hcc_euk_Cohort$wil_sign_df$`2019_Zheng` == 'sign', "k__Eukaryota", "ns"),
                                      log2FC = log2(hcc_euk_Cohort$FC_df$`2019_Zheng`),
                                      ShortName = gsub("^.+\\|", "", rownames(hcc_euk_Cohort$FC_df))),
                           data.frame(pvalue = hcc_arc_Cohort$wil_df$`2019_Zheng`, 
                                      sign = ifelse(hcc_arc_Cohort$wil_sign_df$`2019_Zheng` == 'sign', "k__Archaea", "ns"),
                                      log2FC = log2(hcc_arc_Cohort$FC_df$`2019_Zheng`),
                                      ShortName = gsub("^.+\\|", "", rownames(hcc_arc_Cohort$FC_df))),
                           data.frame(pvalue = hcc_vir_Cohort$wil_df$`2019_Zheng`, 
                                      sign = ifelse(hcc_vir_Cohort$wil_sign_df$`2019_Zheng` == 'sign', "k__Viruses", "ns"),
                                      log2FC = log2(hcc_vir_Cohort$FC_df$`2019_Zheng`),
                                      ShortName = gsub("^.+\\|", "", rownames(hcc_vir_Cohort$FC_df))))) %>%
  as.data.frame()

hcc_vol_df$group <- ifelse(hcc_vol_df$log2FC %>% abs() >log2(1.5), hcc_vol_df$sign, "ns")
hcc_vol_df$EnrDep <- ifelse(hcc_vol_df$group == "ns", "ns", ifelse(hcc_vol_df$log2FC >0, "R", "NR"))
hcc_vol_df$NlogP <- -log10(hcc_vol_df$pvalue)

hcc_vol_plot <- hcc_vol_df %>%
  ggplot(., aes(x = log2FC, y = NlogP)) + 
  geom_point(aes(alpha = group, fill = group, size = group), shape = 21, col = "gray")+
  scale_alpha_manual(values = c(k__Bacteria = 1, k__Eukaryota = 1, k__Viruses = 1, ns = .1))+
  scale_fill_manual(values = c(tax_col, ns = "black")) + 
  scale_size_manual(values = c(k__Bacteria = 2, k__Eukaryota = 2, k__Viruses = 2, ns = 1)) +
  geom_vline(xintercept = c(-1, 1) * log2(1.5), linetype = 2, color = respond_col) +
  geom_hline(yintercept = -log10(.05), linetype = 2) + 
  scale_x_continuous(expand = expansion(add = .1)) +
  scale_y_continuous(expand = expansion(add = c(0,.1))) +
  theme_classic() +
  theme(legend.position = 'none')

hcc_vol_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Cohort-ScatterPlot"),
                          Prefix = paste0("VolcanoPlot-", "HCC"), 
                          Suffix = "csv")
write.csv(x = hcc_vol_df, file = hcc_vol_csv)


hcc_vol_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/Wilcox_MN/Cohort-ScatterPlot"),
                          Prefix = paste0("VolcanoPlot-", "HCC"), 
                          Suffix = "pdf")
pdf(file = hcc_vol_pdf, width = 10, height = 7)
plot(hcc_vol_plot)
plot(hcc_vol_plot + theme(text = element_blank()))
plot(hcc_vol_plot + theme(text = element_blank()) + 
       geom_text_repel(data = hcc_vol_df %>% subset(group != "ns"), 
                       aes(label = ShortName),  size = 1,
                       box.padding = 0.5, max.overlaps = Inf))
dev.off()


```

-->









<!--

### segment plot

```{r}

melt_mm_bac_coh <- do.call("rbind",
                           list(mm_bac_Cohort$wil_df %>% 
                                  subset(All < .05) %>% 
                                  rownames_to_column(var = "Candidate") %>% 
                                  melt(.,id.vars = "Candidate", variable.name = "Cohort", value.name = "Pvalue"),
                                mm_bac_Cohort$wil_df %>% 
                                  subset(All >= .05) %>%
                                  rownames_to_column(var = "Candidate") %>% 
                                  select(Candidate, All) %>%
                                  add_column(Cohort = "All") %>%
                                  rename(Pvalue = All) %>%
                                  select(Candidate, Cohort, Pvalue))) %>%
  add_column(log2FC = apply(., 1,  function(x) mm_bac_Cohort$FC_df[x[1], x[2]]) %>% log2() %>% unlist(),
             Nlog10P = -log10(.[["Pvalue"]])) %>%
  add_column(sign = ifelse(.[["Pvalue"]] >.05 , "Non-Sign", "Sign"))
  
  

tmp_DT <- melt_mm_bac_coh
tmp_DT$ShortName <- gsub("^.+\\|", "", tmp_DT$Candidate)

setDT(tmp_DT)
tmp_plot1 <- ggplot(data = tmp_DT, aes(log2FC, Nlog10P, color = Cohort)) +
  geom_point(aes(shape = Cohort, 
                 size = Cohort,
                 col = Cohort, alpha = sign)) +
  geom_segment(data = tmp_DT[Cohort == "All"][tmp_DT[Cohort != "All"], on = "Candidate"],
               aes(xend = i.log2FC, yend = i.Nlog10P), alpha = .05) + 
  scale_shape_manual(values = c(`2017_Frankel` = 0, `2018_Gopalakrishnan` = 1, `2018_Maston` = 2,
                                `2019_Peters` = 3, `2021_Spencer` = 4, `2022_Lee` = 5,
                                `2022_McCulloch` = 6, All = 20)) +
  scale_size_manual(values = c(`2017_Frankel` = 1, `2018_Gopalakrishnan` = 1, `2018_Maston` = 1,
                               `2019_Peters` = 1, `2021_Spencer` = 1, `2022_Lee` = 1,
                               `2022_McCulloch` = 1, All = 3)) +
  scale_color_manual(values = c(cohort_col, All = "black")) + 
  scale_alpha_manual(values = c(`Non-Sign` = .1, `Sign` = 1)) + theme_bw() + 
  geom_text_repel(data = tmp_DT %>% subset(Cohort == "All" & sign == "Sign"), aes(label = ShortName))
tmp_plot1
```



### Wilcoxon test (relative Abundance)

```{r wilcoxon test}
wil_bac_sp <- Wil_Matrix(sub_otu = bICI_ds_list$k__Bacteria$taxa_abund$Species, 
                         sub_meta = bICI_ds_list$k__Bacteria$sample_table,
                         sub_catolog = "Disease", sub_comparison = "Response",
                         rarefy_cutoff = 0.01, p_cutoff = 0.05)

wil_euk_sp <- Wil_Matrix(sub_otu = bICI_ds_list$k__Eukaryota$taxa_abund$Species, 
                         sub_meta = bICI_ds_list$k__Eukaryota$sample_table,
                         sub_catolog = "Disease", sub_comparison = "Response",
                         rarefy_cutoff = 0.01, p_cutoff = 0.05)

wil_arc_sp <- Wil_Matrix(sub_otu = bICI_ds_list$k__Archaea$taxa_abund$Species, 
                         sub_meta = bICI_ds_list$k__Archaea$sample_table,
                         sub_catolog = "Disease", sub_comparison = "Response",
                         rarefy_cutoff = 0.01, p_cutoff = 0.05)

wil_vir_sp <- Wil_Matrix(sub_otu = bICI_ds_list$k__Viruses$taxa_abund$Species, 
                         sub_meta = bICI_ds_list$k__Viruses$sample_table,
                         sub_catolog = "Disease", sub_comparison = "Response",
                         rarefy_cutoff = 0.01, p_cutoff = 0.05)

nrow(wil_bac_sp$sel_wil_df) # 246
nrow(wil_euk_sp$sel_wil_df) # 215
nrow(wil_arc_sp$sel_wil_df) # 44
nrow(wil_vir_sp$sel_wil_df) # 33


sum(rowSums(wil_bac_sp$sel_wil_df[, 1:4] <.05) >= 2) # 10
sum(rowSums(wil_euk_sp$sel_wil_df[, 1:4] <.05) >= 2) # 11
sum(rowSums(wil_arc_sp$sel_wil_df[, 1:4] <.05) >= 2) # 2
sum(rowSums(wil_vir_sp$sel_wil_df[, 1:4] <.05) >= 2) # 1

sum(rowSums(wil_bac_sp$sel_wil_df[, 1:4] <.05) >= 3) # 0
sum(rowSums(wil_euk_sp$sel_wil_df[, 1:4] <.05) >= 3) # 0
sum(rowSums(wil_arc_sp$sel_wil_df[, 1:4] <.05) >= 3) # 0
sum(rowSums(wil_vir_sp$sel_wil_df[, 1:4] <.05) >= 3) # 0

melt_wil_list <- list()

melt_wil_list$k__Bacteria <- melt(as.matrix(wil_bac_sp$wil_sign_df[, 1:4]),
                                  varnames = c("Feature", "CancerType"), 
                                  value.name = "Select") %>% subset(Select == "sign")
melt_wil_list$k__Eukaryota <- melt(as.matrix(wil_euk_sp$wil_sign_df[, 1:4]),
                                   varnames = c("Feature", "CancerType"), 
                                   value.name = "Select") %>% subset(Select == "sign")
melt_wil_list$k__Archaea <- melt(as.matrix(wil_arc_sp$wil_sign_df[, 1:4]),
                                 varnames = c("Feature", "CancerType"), 
                                 value.name = "Select") %>% subset(Select == "sign")
melt_wil_list$k__Viruses <- melt(as.matrix(wil_vir_sp$wil_sign_df[, 1:4]),
                                 varnames = c("Feature", "CancerType"), 
                                 value.name = "Select") %>% subset(Select == "sign")
```



#### LEfSe_Cal

```{r LEfSe_Cal}
LEfSe_Cal <- function(my_ds, sub_comparison, sub_catolog, taxa_level = "all"){
  lefse_list <- list()
  sel_mt <- matrix(NA, nrow = 0, ncol = 4, dimnames = list(NULL, c('Feature', "Disease", "LDA", "unadj_P")))
  for (sub_set in unique(my_ds$sample_table[[sub_catolog]])) {
    sub_ds <- my_ds$clone()
    sub_ds$sample_table <- sub_ds$sample_table[sub_ds$sample_table[[sub_catolog]] == sub_set, ]
    sub_ds$tidy_dataset()
    lefse_ds <- trans_diff$new(dataset = sub_ds, method = "lefse",
                               group = sub_comparison, filter_thres = .0001, 
                               p_adjust_method = NULL, 
                               taxa_level = taxa_level)
    lefse_list[[sub_set]] <- lefse_ds
    sel_mt <- rbind(sel_mt, data.frame(Feature = lefse_ds$res_diff$Taxa, 
                                       Disease = rep(sub_set, nrow(lefse_ds$res_diff)),
                                       LDA = lefse_ds$res_diff$LDA,
                                       unadj_P = lefse_ds$res_diff$P.unadj))
  }
  cast_mt <- dcast(sel_mt, Feature ~ Disease, value.var = 'LDA')
  cast_mt2 <- dcast(sel_mt, Feature ~ Disease, value.var = 'unadj_P')
  return_list <- list(lefse_list = lefse_list, 
                      cast_lda_mt = cast_mt,
                      cast_p_mt = cast_mt2)
  return(return_list)
}

```

## ALDEx2

```{r ALDEx2}
require(ALDEx2)

rare_otu <- bICI_ds_list$k__Bacteria$otu_table

rare_otu <- bICI_ds_list$k__Bacteria$otu_table[bICI_ds_list$k__Bacteria$taxa_abund$Species %>%
                                                  filter(rowMeans(.) > .01/100) %>% rownames(),]

conds <- bICI_ds_list$k__Bacteria$sample_table$Response

aldex.euk <- aldex(rare_otu, conds, mc.samples=16, 
                   test="t", effect=TRUE, 
                   include.sample.summary=FALSE, 
                   denom="all", verbose=FALSE)

aldex.plot(aldex.euk, type="MA", test="welch",
           xlab="Log-ratio abundance",ylab="Difference",
           cutoff.pval = 0.15, cutoff.effect = 5)

aldex.plot(aldex.euk, type="MW",test="wilcox", 
           cutoff=0.15, cutoff.pval = .1,
           all.cex=0.7, called.cex=1.1, 
           rare.col="black", called.col="red")


aldex_list <- list()

cutoff<- c(k__Bacteria = 500,
           k__Eukaryota = 5,
           k__Archaea = 5,
           k__Viruses = 5)
for (sub_kingdom in c("k__Bacteria", "k__Eukaryota", "k__Archaea", "k__Viruses")) {
  for (sub_cancer in c("NSCLC", "RCC", "HCC")) {
    tmp_meta <- bICI_ds_list[[sub_kingdom]]$sample_table %>% subset(Disease == sub_cancer)
    rare_otu <- bICI_ds_list[[sub_kingdom]]$otu_table %>% 
      filter(rowMeans(.) > cutoff[sub_kingdom]) %>% 
      select(rownames(tmp_meta))
    conds <- tmp_meta$Response
    aldex_res <- aldex(rare_otu, conds, mc.samples=16, 
                       test="t", effect=TRUE, 
                       include.sample.summary=FALSE, 
                       denom="all", verbose=FALSE)
    aldex_list[[sub_kingdom]][[sub_cancer]] <- aldex_res
  }
  
  # aldex.plot(aldex_res, type="MA", test="welch",
  #            xlab="Log-ratio abundance",ylab="Difference",
  #            cutoff.pval = 0.15, cutoff.effect = 5)
  # 
  # aldex.plot(aldex_res, type="MW",test="wilcox", 
  #            cutoff=0.15, cutoff.pval = .1,
  #            all.cex=0.7, called.cex=1.1, 
  #            rare.col="black", called.col="red")
}

aldex.plot(aldex_list[[sub_kingdom]][[sub_cancer]],
           type="MA", test="welch",
           xlab="Log-ratio abundance",ylab="Difference",
           cutoff.pval = 0.15, cutoff.effect = 5)

aldex.plot(aldex_list[[sub_kingdom]][[sub_cancer]],
           type="MW",test="wilcox",
           cutoff=0.15, cutoff.pval = .5,
           all.cex=0.7, called.cex=1.1,
           rare.col="black", called.col="red")


```


### LEfSe

```{r LEfSe}
lefse_bac_res <- LEfSe_Cal(my_ds = bICI_ds_list$k__Bacteria$clone(), 
                           sub_comparison = "Response", 
                           sub_catolog = "Disease", 
                           taxa_level = "Species")
lefse_euk_res <- LEfSe_Cal(my_ds = bICI_ds_list$k__Eukaryota$clone(), 
                           sub_comparison = "Response", 
                           sub_catolog = "Disease", 
                           taxa_level = "Species")
lefse_arc_res <- LEfSe_Cal(my_ds = bICI_ds_list$k__Archaea$clone(), 
                           sub_comparison = "Response", 
                           sub_catolog = "Disease", 
                           taxa_level = "Species")
lefse_vir_res <- LEfSe_Cal(my_ds = bICI_ds_list$k__Viruses$clone(), 
                           sub_comparison = "Response", 
                           sub_catolog = "Disease", 
                           taxa_level = "Species")

sum(rowSums(is.na(lefse_bac_res$cast_lda_mt)) < 3) # 10/219
sum(rowSums(is.na(lefse_euk_res$cast_lda_mt)) < 3) # 3/225
sum(rowSums(is.na(lefse_arc_res$cast_lda_mt)) < 3) # 7/90
sum(rowSums(is.na(lefse_vir_res$cast_lda_mt)) < 3) # 6/70

sum(rowSums(is.na(lefse_bac_res$cast_lda_mt)) < 2) # 0
sum(rowSums(is.na(lefse_euk_res$cast_lda_mt)) < 2) # 0
sum(rowSums(is.na(lefse_arc_res$cast_lda_mt)) < 2) # 1
sum(rowSums(is.na(lefse_vir_res$cast_lda_mt)) < 2) # 0


melt_lefse_list <- list()

melt_lefse_list$k__Bacteria <- melt(lefse_bac_res$cast_lda_mt,
                                    id = "Feature",
                                    varnames = c("Feature", "CancerType"), 
                                    value.name = "Select") %>% na.omit()
melt_lefse_list$k__Eukaryota <- melt(lefse_euk_res$cast_lda_mt,
                                     id = "Feature",
                                     varnames = c("Feature", "CancerType"), 
                                     value.name = "Select") %>% na.omit()
melt_lefse_list$k__Archaea <- melt(lefse_arc_res$cast_lda_mt,
                                   id = "Feature",
                                   varnames = c("Feature", "CancerType"), 
                                   value.name = "Select") %>% na.omit()
melt_lefse_list$k__Viruses <- melt(lefse_vir_res$cast_lda_mt,
                                   id = "Feature",
                                   varnames = c("Feature", "CancerType"), 
                                   value.name = "Select") %>% na.omit()

```

## Feature Plot 

### UpSet Plot (LEfSe)

```{r upset lefse}
lt_list <- list()
m_list <- list()
lt <- list()


for (sel_kingdoms in names(melt_lefse_list)) {
  lt_list[[sel_kingdoms]] <- list()
  for (sel_disease in unique(melt_lefse_list[[sel_kingdoms]]$variable)) {
    lt_list[[sel_kingdoms]][[sel_disease]] <- melt_lefse_list[[sel_kingdoms]] %>%
      subset(variable == sel_disease, select = "Feature") %>% unlist() %>% as.character()
    lt[[sel_disease]] <- c(lt[[sel_disease]], lt_list[[sel_kingdoms]][[sel_disease]])
  }
  m_list[[sel_kingdoms]] <- make_comb_mat(lt_list[[sel_kingdoms]])
}
m = make_comb_mat(lt)

rowAnn_mt <- data.frame(k__Bacteria = set_size(m_list$k__Bacteria),
                        k__Eukaryota = set_size(m_list$k__Eukaryota),
                        k__Archaea = set_size(m_list$k__Archaea),
                        k__Viruses = set_size(m_list$k__Viruses)) %>%
  as.matrix()

colAnn_mt <- matrix(data = 0, 
                    nrow = length(comb_name(m)), 
                    ncol = 4,
                    dimnames = list(comb_name(m),
                                    names(tax_col)))

for (sub_k in names(m_list)) {
  for (InSize in names(comb_size(m_list[[sub_k]]))) {
    colAnn_mt[InSize, sub_k] <- comb_size(m_list[[sub_k]])[InSize]
  }
}


tmp_ht <- UpSet(m, set_order = order(set_size(m), decreasing = T), 
                heatmap_height = unit(length(set_name(m))*1.5, "cm"),
                heatmap_width = unit(length(comb_name(m))*1.5, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#F0F0FF", "#FFF0F0"),
                comb_order = order(lapply(comb_name(m), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m), 
                                   decreasing = T),
                top_annotation = HeatmapAnnotation(
                  "Intersection\nsize" = anno_barplot(colAnn_mt, annotation_name_side = "left",
                                                     gp = gpar(fill = tax_col, 
                                                               col = NA),
                                                     height = unit(3, "cm"),
                                                     border = F),
                  annotation_name_rot = 90,
                  annotation_name_side = "left"),
                right_annotation = rowAnnotation(
                  "Set size" = anno_barplot(rowAnn_mt,
                                            gp = gpar(fill = tax_col,
                                                      col = NA),
                                            # add_numbers = TRUE,
                                            width = unit(3, "cm"),
                                            border = F)),
                row_names_gp = gpar(fontface ='bold'))


sim_ht <- UpSet(m, set_order = order(set_size(m), decreasing = T), 
                heatmap_height = unit(length(set_name(m))*1.5, "cm"),
                heatmap_width = unit(length(comb_name(m))*1.5, "cm"),
                pt_size = unit(5, "mm"), lwd = 3,
                bg_col = c("#F0F0FF", "#FFF0F0"),
                comb_order = order(lapply(comb_name(m), function(x) {
                                     strsplit(x = x, split = "") %>% 
                                       unlist() %>%
                                       as.numeric() %>% 
                                       sum()
                                     }) %>% unlist() * -1, 
                                   comb_size(m), 
                                   decreasing = T),
                top_annotation = upset_top_annotation(m, add_numbers = TRUE,
                                                      numbers_gp = gpar(fontsize = 8),
                                                      numbers_rot = 0,
                                                      annotation_name_rot = 90,
                                                      height = unit(3, "cm")),
                right_annotation = upset_right_annotation(m, add_numbers = TRUE,
                                                          width = unit(3, "cm")), 
                row_names_gp = gpar(fontface ='bold'))


if (save_or_not) {
  
  rowAnn_order_mt <- rowAnn_mt[order(rowSums(rowAnn_mt), decreasing = T),]
  
  colAnn_order_mt <- colAnn_mt[comb_name(m)[order(lapply(comb_name(m),
                                                         function(x) {
                                                           strsplit(x = x, split = "") %>%
                                                             unlist() %>%
                                                             as.numeric() %>%
                                                             sum()
                                                           }) %>% unlist() * -1,
                                                  omb_size(m),
                                                  decreasing = T)],] %>% t()

  bICI_hm_pdf <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/LEfSE"),
                            Prefix = "Heatmap-bICI_lefse", Suffix = "pdf")
    
  pdf(file = bICI_hm_pdf, width = 10, height = 5)
  draw(tmp_ht)
  draw(sim_ht)
  dev.off()
  
  colAnn_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/LEfSE"),
                           Prefix = "Matrix-ColumnAnnotation", Suffix = "csv")
  write.csv(x = colAnn_order_mt, file = colAnn_csv)
  rowAnn_csv <- FileCreate(DirPath = paste0(save_path_prefix, "05.FeatureSelection/LEfSE"),
                           Prefix = "Matrix-RowAnnotation", Suffix = "csv")
  write.csv(x = rowAnn_order_mt, file = rowAnn_csv)

}

```

### Wilcoxon test (MN, cohort)

```{r MN cohort}
wil_bac_sp_mn_cohort <- Wil_Matrix(sub_otu = t(mn_bac_df), p_cutoff = 0.05, rarefy_or_not = F,
                                   sub_meta = bICI_ds_list$k__Bacteria$sample_table,
                                   sub_catolog = "Cohort_name", sub_comparison = "Response")

wil_euk_sp_mn_cohort <- Wil_Matrix(sub_otu = t(mn_euk_df), p_cutoff = 0.05, rarefy_or_not = F,
                                   sub_meta = bICI_ds_list$k__Eukaryota$sample_table,
                                   sub_catolog = "Cohort_name", sub_comparison = "Response")

wil_arc_sp_mn_cohort <- Wil_Matrix(sub_otu = t(mn_arc_df), p_cutoff = 0.05, rarefy_or_not = F,
                                   sub_meta = bICI_ds_list$k__Archaea$sample_table,
                                   sub_catolog = "Cohort_name", sub_comparison = "Response")

wil_vir_sp_mn_cohort <- Wil_Matrix(sub_otu = t(mn_vir_df), p_cutoff = 0.05, rarefy_or_not = F,
                                   sub_meta = bICI_ds_list$k__Viruses$sample_table,
                                   sub_catolog = "Cohort_name", sub_comparison = "Response")

```

### Wilcoxon test (MN Genus)

```{r MN Genus}

mn_genus_bac_df <- MedianNormal(raw_df = bICI_ds_list$k__Bacteria$taxa_abund$Genus, 
                                meta_df = bICI_ds_list$k__Bacteria$sample_table, 
                                min_add = 0.0001, rarefy_cutoff = 0.01)

mn_genus_euk_df <- MedianNormal(raw_df = bICI_ds_list$k__Eukaryota$taxa_abund$Genus, 
                                meta_df = bICI_ds_list$k__Eukaryota$sample_table,
                                min_add = 0.0001, rarefy_cutoff = 0.01)

mn_genus_arc_df <- MedianNormal(raw_df = bICI_ds_list$k__Archaea$taxa_abund$Genus, 
                                meta_df = bICI_ds_list$k__Archaea$sample_table,
                                min_add = 0.0001, rarefy_cutoff = 0.01)

mn_genus_vir_df <- MedianNormal(raw_df = bICI_ds_list$k__Viruses$taxa_abund$Genus, 
                                meta_df = bICI_ds_list$k__Viruses$sample_table,
                                min_add = 0.0001, rarefy_cutoff = 0.01)



# raw_meta <- paste0(bICI_ds_list$k__Bacteria$sample_table$Cohort_name, "-", 
#                    bICI_ds_list$k__Bacteria$sample_table$Disease)
# names(raw_meta) <- rownames(bICI_ds_list$k__Bacteria$sample_table)



wil_genus_bac_sp_mn <- Wil_Matrix(sub_otu = t(mn_genus_bac_df), p_cutoff = 0.05, rarefy_or_not = F,
                                  sub_meta = bICI_ds_list$k__Bacteria$sample_table,
                                  sub_catolog = "Disease", sub_comparison = "Response")

wil_genus_euk_sp_mn <- Wil_Matrix(sub_otu = t(mn_genus_euk_df), p_cutoff = 0.05, rarefy_or_not = F,
                                  sub_meta = bICI_ds_list$k__Eukaryota$sample_table,
                                  sub_catolog = "Disease", sub_comparison = "Response")

wil_genus_arc_sp_mn <- Wil_Matrix(sub_otu = t(mn_genus_arc_df), p_cutoff = 0.05, rarefy_or_not = F,
                                  sub_meta = bICI_ds_list$k__Archaea$sample_table,
                                  sub_catolog = "Disease", sub_comparison = "Response")

wil_genus_vir_sp_mn <- Wil_Matrix(sub_otu = t(mn_genus_vir_df), p_cutoff = 0.05, rarefy_or_not = F,
                                  sub_meta = bICI_ds_list$k__Viruses$sample_table,
                                  sub_catolog = "Disease", sub_comparison = "Response")

nrow(wil_genus_bac_sp_mn$sel_wil_df) 
nrow(wil_genus_euk_sp_mn$sel_wil_df) 
nrow(wil_genus_arc_sp_mn$sel_wil_df) 
nrow(wil_genus_vir_sp_mn$sel_wil_df) 

sum(rowSums(wil_genus_bac_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 2/44
sum(rowSums(wil_genus_euk_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 10/173
sum(rowSums(wil_genus_arc_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 4/44
sum(rowSums(wil_genus_vir_sp_mn$sel_wil_df[, 1:4] <.05) >= 2) # 1/14

sum(rowSums(wil_genus_bac_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 0
sum(rowSums(wil_genus_euk_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 0
sum(rowSums(wil_genus_arc_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 1
sum(rowSums(wil_genus_vir_sp_mn$sel_wil_df[, 1:4] <.05) >= 3) # 0

melt_wil_genus_list <- list()

melt_wil_genus_list$k__Bacteria <- melt(as.matrix(wil_genus_bac_sp_mn$wil_sign_df[, 1:4]),
                                  varnames = c("Feature", "CancerType"), 
                                  value.name = "Select") %>% subset(Select == "sign")
melt_wil_genus_list$k__Eukaryota <- melt(as.matrix(wil_genus_euk_sp_mn$wil_sign_df[, 1:4]),
                                   varnames = c("Feature", "CancerType"), 
                                   value.name = "Select") %>% subset(Select == "sign")
melt_wil_genus_list$k__Archaea <- melt(as.matrix(wil_genus_arc_sp_mn$wil_sign_df[, 1:4]),
                                 varnames = c("Feature", "CancerType"), 
                                 value.name = "Select") %>% subset(Select == "sign")
melt_wil_genus_list$k__Viruses <- melt(as.matrix(wil_genus_vir_sp_mn$wil_sign_df[, 1:4]),
                                 varnames = c("Feature", "CancerType"), 
                                 value.name = "Select") %>% subset(Select == "sign")

```

### test (ALDEX2)

```{r aldex2}
relAbun_sp <- bICI_ds_list$k__Bacteria$taxa_abund$Species
otu_sp <- bICI_ds_list$k__Bacteria$otu_table
aldex2_meta <- bICI_ds_list$k__Bacteria$sample_table

sel_cancer <- "HCC"

aldex_mn_bac_res <- aldex(otu_sp[rowMeans(relAbun_sp) >.01/100, aldex2_meta$Disease == sel_cancer], 
                          aldex2_meta$Response[aldex2_meta$Disease == sel_cancer], 
                          mc.samples=16, 
                          test="t", 
                          effect=TRUE,
                          include.sample.summary=FALSE, 
                          denom="all", 
                          verbose=FALSE)

sig_by_both <- which(aldex_mn_bac_res$we.ep < 0.1 & aldex_mn_bac_res$wi.ep < 0.1)

rownames(aldex_mn_bac_res)[sig_by_both]



relAbun_sp <- bICI_ds_list$k__Bacteria$taxa_abund$Species
otu_sp <- bICI_ds_list$k__Bacteria$otu_table
aldex2_meta <- bICI_ds_list$k__Bacteria$sample_table


aldex_mn_bac_res <- aldex(otu_sp[rowMeans(relAbun_sp) >.01/100,], 
                          aldex2_meta$Response, 
                          mc.samples=16,  
                          test="t", 
                          effect=TRUE, 
                          include.sample.summary=FALSE, 
                          denom="all", 
                          verbose=FALSE)

sig_by_both <- which(aldex_mn_bac_res$we.ep < 0.1 & aldex_mn_bac_res$wi.ep < 0.1)

rownames(aldex_mn_bac_res)[sig_by_both]

```

-->

